---
title: "Set analysis parameters"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{Set analysis parameters}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

This vignette explicates the "set analysis parameters" step of the `sceptre` pipeline. The purpose of this step is to select the parameters that govern how the statistical analysis is to be carried out.

```{r, out.width = "650px", fig.align="center", echo = FALSE}
knitr::include_graphics("pipeline_schematic_step_2.png")
```

We begin by loading the `sceptre` package.

```{r}
library(sceptre)
```

Throughout, we examine two example datasets. First, we examine the example high-MOI CRISPRi screen of candidate enhancers in K562 cells. This is the example dataset that we used in the Get Started (`vignette("sceptre")`) and Import Data (`vignette("import-data")`) vignettes. We initialize a `sceptre_object` called `sceptre_object_highmoi` to represent these data.

```{r}
data(highmoi_example_data)
data(grna_target_data_frame_highmoi)
sceptre_object_highmoi <- import_data(response_matrix = highmoi_example_data$response_matrix,
                                      grna_matrix = highmoi_example_data$grna_matrix,
                                      grna_target_data_frame = grna_target_data_frame_highmoi,
                                      moi = "high",
                                      extra_covariates = highmoi_example_data$extra_covariates,
                                      response_names = highmoi_example_data$gene_names)
```

Next, we examine an example low-MOI CRISPRko screen of gene transcription start sites in immune cells. We initialize a `sceptre_object` called `sceptre_object_lowmoi` to store this latter dataset.

```{r}
sceptre_object_lowmoi <- import_data(response_matrix = lowmoi_example_data$response_matrix,
                                     grna_matrix = lowmoi_example_data$grna_matrix,
                                     extra_covariates = lowmoi_example_data$extra_covariates,
                                     grna_target_data_frame = lowmoi_example_data$grna_target_data_frame,
                                     moi = "low")
```

We print the details of the CRISPRko dataset below.

```{r}
sceptre_object_lowmoi
```

The CRISPR perturbations target gene transcription start sites. Each gene is targeted by three or four gRNAs, and nine gRNAs in the dataset are non-targeting. The "response features" are genes. We print a few rows of the gRNA target data frame corresponding to this dataset below.

```{r}
lowmoi_example_data$grna_target_data_frame[1:8,]
```

Our goal in analyzing this dataset is to dissect gene regulatory networks by linking perturbations of genes to changes in the expression of other genes.

We set the analysis parameters by calling the function `set_analysis_parameters()` on the `sceptre_object`. `set_analysis_parameters()` takes several arguments: `positive_control_pairs`, `discovery_pairs`, `side`, `grna_grouping_strategy`, `control_group`, `resampling_mechanism`, `formula_object`, `fit_parametric_curve`, `multiple_testing_method`, and `multiple_testing_alpha`. The only required argument among these is `discovery_pairs`. We describe these arguments below.

## Positive control pairs

Positive control pairs are target-response pairs for which we know (or have strong reason to believe) that there is a regulatory relationship between the target and the response. We can use positive control pairs to verify that `sceptre` (or any association testing method for that matter) is sensitive (i.e., capable of detecting true associations) on the dataset under analysis. We use the function `construct_positive_control_pairs()` to construct the positive control pairs. `construct_positive_control_pairs()` takes as an argument a `sceptre_object` and returns a data frame with columns `target` and `response_id`, where gRNA targets that coincide with response IDs are paired. We call `construct_positive_control_pairs()` on the high-MOI CRISPRi dataset the low-MOI CRISPRko dataset. In both cases the positive control set consists of transcription start sites paired to the gene regulated by those transcription start sites.

```{r}
positive_control_pairs_highmoi <- construct_positive_control_pairs(sceptre_object_highmoi)
head(positive_control_pairs_highmoi) # high MOI CRISPRi dataset
```

```{r}
positive_control_pairs_lowmoi <- construct_positive_control_pairs(sceptre_object_lowmoi)
head(positive_control_pairs_lowmoi) # low MOI CRISPRko dataset
```

Positive control pairs need not consist exclusively of gene transcription start sites paired to target genes. For example, enhancer-gene links that have been validated previously (through, e.g., an arrayed CRISPR screen) also can serve as positive control pairs. Users manually can append additional positive control pairs to the positive control pair data frame via a call to `rbind()`. For example, suppose we know on the high-MOI CRISPRi data that `candidate_enh_1` regulates gene `ENSG00000069482` and that `candidate_enh_10` regulates gene `ENSG00000135821`. We could add these enhancer-gene links to the positive control pair data frame as follows.

```{r}
# construct data frame of additional positive control pairs
additional_positive_control_pairs <-
  data.frame(grna_target = c("candidate_enh_1", "candidate_enh_10"),
             response_id = c("ENSG00000069482", "ENSG00000135821"))

# append additional pairs to positive control data frame
positive_control_pairs_highmoi_updated <- rbind(positive_control_pairs_highmoi,
                                                additional_positive_control_pairs)
positive_control_pairs_highmoi_updated
```

We then could supply `positive_control_pairs_highmoi_updated` instead of `positive_control_pairs_highmoi` to `set_analysis_parameters()`.

## Discovery pairs

Discovery pairs are target-response pairs whose association status we do not know but seek to learn. Unlike positive control and negative control pairs, which serve a mainly technical purpose, discovery pairs are of primary scientific interest. Discovery pairs must be supplied to `set_analysis_parameters()`; all other arguments (aside from `sceptre_object`) are optional. `sceptre` provides two helper functions for constructing discovery pairs: `construct_cis_pairs()` and `construct_trans_pairs()`.

#### Construct *cis* pairs

`construct_cis_pairs()` returns the set of target-response pairs for which the target and response are located on the same chromosome and are in close physical proximity to one another. `construct_cis_pairs()` assumes that the columns `chr`, `start`, and `stop` are present within the `grna_target_data_frame`. `construct_cis_pairs()` also assumes that the responses are genes as opposed to, say, proteins. `construct_cis_pairs()` is a useful pair constructor function for screens that aim to map noncoding regulatory elements (e.g., enhancers, silencers, or noncoding GWAS variants) to target genes in *cis*. `construct_cis_pairs()` takes several arguments: `sceptre_object` (required) `distance_threshold` (optional), `positive_control_pairs` (optional), and `ref_genome` (optional). Each target is paired to the set of genes within `distance_threshold` bases of that target. (The default value of `distance_threshold` is 500,000 bases, or half a megabase.) The `positive_control_pairs` data frame optionally can be passed to `construct_cis_pairs()`, in which case the positive control targets (i.e., the entries within the `grna_target` column of `positive_control_pairs`) are excluded from the *cis* pairs. Finally, `ref_genome` is a string indicating the reference genome that `chr`, `start`, and `stop` are defined with respect to. (The only reference genome currently available is "10X_GRCh38_2020," which is the GRCh38 reference genome that has shipped with CellRanger since 2020.)

We use `construct_cis_pairs()` to construct the discovery pairs for the high-MOI CRISPRi dataset. We set `distance_threshold` to `5e6` (i.e., 5 megabases) so as to increase the number of pairs in the discovery set for illustration purposes.

```{r}
discovery_pairs_highmoi <- construct_cis_pairs(sceptre_object = sceptre_object_highmoi,
                                               positive_control_pairs = positive_control_pairs_highmoi,
                                               distance_threshold = 5e6)
```

`discovery_pairs_highmoi` is a data frame with columns `grna_target` and `response_id`; each candidate enhancer is mapped to the set of genes in close proximity to that candidate enhancer.

```{r}
discovery_pairs_highmoi[c(1, 10, 20, 30, 40, 50),]
```

#### Construct *trans* pairs

`construct_trans_pairs()` returns the entire set of possible target-response pairs. `construct_trans_pairs()` is a useful pair constructor function for analyses in which we seek to conduct a *trans* analysis, testing each target against each response. `construct_trans_pairs()` takes as arguments `sceptre_object` (required), `positive_control_pairs` (optional), `exclude_positive_control_pairs` (optional), and `exclude_positive_control_targets` (optional). By default `construct_trans_pairs()` returns a data frame with columns `grna_target` and `response_id`, where each gRNA target is mapped to each response ID. If the `positive_control_pairs` data frame is passed and `exclude_positive_control_pairs` is set to `TRUE`, then the positive control target-response pairs are excluded from the *trans* pairs. Next, if `positive_control_pairs` is passed and `exclude_positive_control_targets` is set to `TRUE`, then *all* pairs containing a positive control gRNA target are excluded from the *trans* pairs. (In this sense setting `exclude_positive_control_targets` to `TRUE` is stronger than setting `exclude_positive_control_pairs` to `TRUE`.)

We use `construct_trans_pairs()` to construct the discovery set for the low-MOI CRISPRko dataset. We exclude the positive control pairs. The discovery set consists of the set of gene targets paired to *all other* genes.

```{r}
discovery_pairs_lowmoi <- construct_trans_pairs(sceptre_object = sceptre_object_lowmoi,
                                                positive_control_pairs = positive_control_pairs_lowmoi,
                                                exclude_positive_control_pairs = TRUE)
head(discovery_pairs_lowmoi)
```

We also can use `construct_trans_pairs()` to construct a *trans* discovery set for the high-MOI, enhancer-targeting CRISPRi dataset. To this end we call `construct_trans_pairs()`, setting `exclude_positive_control_targets` to `TRUE` so as to exclude all positive control gRNA targets from the *trans* pairs.

```{r}
discovery_pairs_highmoi_trans <- construct_trans_pairs(sceptre_object = sceptre_object_highmoi,
                                                       positive_control_pairs = positive_control_pairs_highmoi,
                                                       exclude_positive_control_targets = TRUE)
```

The resulting data frame, `discovery_pairs_highmoi_trans`, maps each candidate enhancer to the entire set of genes.

```{r}
head(discovery_pairs_highmoi_trans)
```

Some users may wish to run both a *cis* analysis and a *trans* analysis on their data. We recommend that such users carry out the `sceptre` pipeline twice: once using the *cis* discovery set and once using the *trans* discovery set.

## Side

`sceptre` can run left-tailed, right-tailed, and two-tailed tests of association. Left-tailed tests test for a *decrease* in expression, right-tailed tests test for an *increase* in expression, and two-tailed tests test for an increase *or* decrease in expression. The parameter `side` controls the sidedness of the test and can take values `"left"`, `"right"`, or `"both"`. (`"both"` indicates a two-tailed test.) Two-tailed tests typically are the best choice for *trans* analyses, as the direction of *trans* relationships generally is uncertain. For *cis* analyses, on the other hand, the sidedness of the test should be chosen on the basis of the type of genomic element being targeted and the CRISPR perturbation modality. The following table summarizes whether a left- or right-tailed test is appropriate for a *cis* analysis as a function of these variables.

| Target element | CRISPR modality     | Testing for            | Sidedness |
|----------------|---------------------|------------------------|-----------|
| Enhancer       | CRISPRi or CRISPRko | Decrease in expression | Left      |
| Enhancer       | CRISPRa             | Increase in expression | Right     |
| Silencer       | CRISPRi or CRISPRko | Increase in expression | Right     |
| Silencer       | CRISPRa             | Decrease in expression | Left      |

## gRNA grouping strategy

Multiple gRNAs typically target a given genomic element.

are designed to target a given genomic element. The parameter `grna_grouping_strategy` controls if and how gRNAs that target the same element are "combined" or "grouped."

Thus, if `grna_1` is present in cells 2, 5, 10, 11, and 19, and if `grna_2` is present in cells 1, 5, 10, and 13, then the combined gRNA is defined to be present in cells 1, 2, 5, 10, 11, 13, and 19.
