<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="`sceptre` is an R package for statistically rigorous, massively scalable, and user-friendly single-cell CRISPR screen data analysis. The `sceptre` pipeline consists of several distinct steps: (1) import data from 10X CellRanger, Parse CRISPR Detect, or a set of R objects; (2) set the analysis parameters, which are the parameters that govern how the statistical analysis is to be conducted; (3) assign gRNAs to cells using one of three principled methods; (4) run quality control; (5) run the calibration check, which is an analysis that verifies that `sceptre` controls the rate of false positives on the dataset under analysis; (6) run the power check, which is an analysis that verifies that `sceptre` is capable of discovering true associations on the dataset under analysis; (7) run the discovery analysis to identify high-confidence perturbation-gene links among the set of perturbations and genes whose association status we do not know but seek to learn; (8) write outputs to a specified directory, including results, plots, and a textual summary of the analysis. `sceptre` leverages several novel statistical and computational algorithms to achieve high statistical accuracy and computational speed.">
<title>Analysis of Single-Cell CRISPR Screen Data • sceptre</title>
<script src="deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="pkgdown.js"></script><meta property="og:title" content="Analysis of Single-Cell CRISPR Screen Data">
<meta property="og:description" content="`sceptre` is an R package for statistically rigorous, massively scalable, and user-friendly single-cell CRISPR screen data analysis. The `sceptre` pipeline consists of several distinct steps: (1) import data from 10X CellRanger, Parse CRISPR Detect, or a set of R objects; (2) set the analysis parameters, which are the parameters that govern how the statistical analysis is to be conducted; (3) assign gRNAs to cells using one of three principled methods; (4) run quality control; (5) run the calibration check, which is an analysis that verifies that `sceptre` controls the rate of false positives on the dataset under analysis; (6) run the power check, which is an analysis that verifies that `sceptre` is capable of discovering true associations on the dataset under analysis; (7) run the discovery analysis to identify high-confidence perturbation-gene links among the set of perturbations and genes whose association status we do not know but seek to learn; (8) write outputs to a specified directory, including results, plots, and a textual summary of the analysis. `sceptre` leverages several novel statistical and computational algorithms to achieve high statistical accuracy and computational speed.">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="index.html">sceptre</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.9.1</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="articles/sceptre.html">
    <span class="fa fa-play"></span>
     
    Get started
  </a>
</li>
<li class="nav-item">
  <a class="external-link nav-link" href="https://timothy-barry.github.io/sceptre-book/">
    <span class="fa fa-book"></span>
     
    Manual
  </a>
</li>
<li class="nav-item">
  <a class="external-link nav-link" href="https://timothy-barry.github.io/sceptre-book/faq.html">
    <span class="fa fa-question-circle-o"></span>
     
    FAQ
  </a>
</li>
<li class="nav-item">
  <a class="nav-link" href="reference/index.html">
    <span class="fa fa-code"></span>
     
    Functions
  </a>
</li>
<li class="nav-item">
  <a class="nav-link" href="news/index.html">
    <span class="fa fa-history"></span>
     
    Changelog
  </a>
</li>
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/Katsevich-Lab/sceptre/discussions">
    <span class="fa fa-users"></span>
     
    Community
  </a>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav">
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/Katsevich-Lab/sceptre/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>

    
  </div>
</nav><div class="container template-home">
<div class="row">
  <main id="main" class="col-md-9"><div style="margin-top: 5px;">
<p><img src="reference/figures/hex.jpg" align="right" width="150"></p>
</div>
<div class="section level2">
<h2 id="sceptre">
<span style="font-size:60px;"><code>sceptre</code></span><a class="anchor" aria-label="anchor" href="#sceptre"></a>
</h2>
<p>An R package for single-cell CRISPR screen data analysis, emphasizing statistical rigor, computational efficiency, and ease of use.</p>
<!-- badges: start -->

<!-- badges: end -->
</div>
<div class="section level2">
<h2 id="release-of-sceptre-v091">Release of <code>sceptre</code> v0.9.1<a class="anchor" aria-label="anchor" href="#release-of-sceptre-v091"></a>
</h2>
<p>We are excited to announce the release of <code>sceptre</code> v0.9.1, a substantial update to the package. This milestone includes the following developments:</p>
<ul>
<li>A reimagined user experience based on a modular, object-oriented workflow.</li>
<li>Further improvements in speed and memory efficiency.</li>
<li>A unified interface for low- and high-MOI analyses.</li>
<li>A suite of plotting functions facilitating visualization of each step in the pipeline.</li>
<li>Expanded support for gRNA assignment and quality control.</li>
<li>Interoperability with output from 10X Cell Ranger and Parse Biosciences CRISPR Detect.</li>
<li>An <a href="https://timothy-barry.github.io/sceptre-book/" class="external-link">e-book</a> guiding users through the entire process of analyzing their data using <code>sceptre</code>.</li>
</ul>
<p><code>sceptre</code> v0.9.1 facilitates an entire analysis pipeline for single-cell CRISPR screens, starting from UMI count data obtained from tools like 10X Cell Ranger.</p>
<!--
&#10;## About `sceptre`
&#10;Single-cell CRISPR screens (e.g., Perturb-seq, TAP-seq) combine CRISPR and single-cell sequencing to survey the effects of genetic perturbations on individual cells. Despite their promise, single-cell CRISPR screens present considerable statistical and computational challenges. `sceptre` is an R package for single-cell CRISPR screen data analysis, emphasizing statistical rigor, computational efficiency, and ease of use.
&#10;### Key features
&#10;-   Import data from 10X CellRanger or a set of R matrices
-   Assign gRNAs to cells using one of three principled methods
-   Perform extensive quality control
-   Run gRNA-to-gene differential expression analyses
-   Verify false discovery rate control using negative control gRNAs
-   Verify adequate power using positive control gRNAs
-   Visualize each step in the pipeline by rendering an informative plot
&#10;### Compatible experimental designs
&#10;-   low multiplicity-of-infection and high multiplicity-of-infection
-   Gene-targeting and noncoding-regulatory-element-targeting
-   CRISPRko, CRISPRi, CRISPRa, CRISPR base editing, and CRISPR prime editing
-   Gene and protein expression readout
&#10;## Get started
&#10;The fastest way to get started with `sceptre` is to work through the [Get Started vignette](articles/sceptre.html) (30 minutes). Users also can read the `sceptre` [manual](https://timothy-barry.github.io/sceptre-book/).
-->
</div>
<div class="section level2">
<h2 id="featured-publications">Featured publications<a class="anchor" aria-label="anchor" href="#featured-publications"></a>
</h2>
<ul>
<li>
<a href="https://www.science.org/doi/10.1126/science.adh7699" class="external-link">Morris et al., 2023</a>. “Discovery of target genes and pathways…”. <em>Science</em>.</li>
<li>
<a href="https://genomebiology.biomedcentral.com/articles/10.1186/s13059-021-02545-2" class="external-link">Barry et al., 2021</a>. “SCEPTRE improves calibration and sensitivity…”. <em>Genome Biology</em>.</li>
</ul>
</div>
<div class="section level2">
<h2 id="bug-reports-feature-requests-and-software-questions">Bug reports, feature requests, and software questions<a class="anchor" aria-label="anchor" href="#bug-reports-feature-requests-and-software-questions"></a>
</h2>
<p>For bug reports, please open a <a href="https://github.com/Katsevich-Lab/sceptre/issues" class="external-link">GitHub issue</a>. For feature requests, please start a discussion under <a href="https://github.com/Katsevich-Lab/sceptre/discussions/categories/feature-requests" class="external-link">feature requests</a>. For questions about <code>sceptre</code> functionality, documentation, or how to apply it to your data, please start a discussion under <a href="https://github.com/Katsevich-Lab/sceptre/discussions/categories/q-a" class="external-link">Q&amp;A</a>.</p>
</div>

  </main><aside class="col-md-3"><div class="links">
<h2 data-toc-skip>Links</h2>
<ul class="list-unstyled">
<li><a href="https://github.com/Katsevich-Lab/sceptre/" class="external-link">Browse source code</a></li>
<li><a href="https://github.com/Katsevich-Lab/sceptre/issues" class="external-link">Report a bug</a></li>
<li><a href="https://github.com/Katsevich-Lab/sceptre/issues" class="external-link">Report a bug</a></li>
</ul>
</div>

<div class="license">
<h2 data-toc-skip>License</h2>
<ul class="list-unstyled">
<li><a href="LICENSE.html">Full license</a></li>
<li><small><a href="https://www.r-project.org/Licenses/GPL-3" class="external-link">GPL-3</a></small></li>
</ul>
</div>

<div class="citation">
<h2 data-toc-skip>Citation</h2>
<ul class="list-unstyled">
<li><a href="authors.html#citation">Citing sceptre</a></li>
</ul>
</div>

<div class="developers">
<h2 data-toc-skip>Developers</h2>
<ul class="list-unstyled">
<li>Timothy Barry <br><small class="roles"> Author, maintainer </small> <a href="https://orcid.org/0000-0002-4356-627X" target="orcid.widget" aria-label="ORCID" class="external-link"><span class="fab fa-orcid orcid" aria-hidden="true"></span></a> </li>
<li>Joseph Deutsch <br><small class="roles"> Contributor </small>  </li>
<li>Eugene Katsevich <br><small class="roles"> Author, thesis advisor </small>  </li>
</ul>
</div>

<div class="funders">
<h2 data-toc-skip>Funders</h2>
<ul class="list-unstyled">
<li>
<p><a href="https://analytics.wharton.upenn.edu/" class="external-link">Analytics at Wharton</a></p>
<p><a href="https://www.nsf.gov/" class="external-link">National Science Foundation</a></p>
<p><a href="https://www.nih.gov" class="external-link">National Institutes of Health</a></p>
</li>
</ul>
</div>

<div class="dev-status">
<h2 data-toc-skip>Dev status</h2>
<ul class="list-unstyled">
<li><a href="https://github.com/Katsevich-Lab/sceptre/actions" class="external-link"><img src="https://github.com/Katsevich-Lab/sceptre/workflows/R-CMD-check/badge.svg" alt="R-CMD-check"></a></li>
</ul>
</div>

  </aside>
</div>


    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>Developed by Timothy Barry, Eugene Katsevich.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
