<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="`sceptre` is an R package for statistically rigorous, massively scalable, and user-friendly single-cell CRISPR screen data analysis. The `sceptre` pipeline consists of several distinct steps: (1) import data from 10X CellRanger, Parse CRISPR Detect, or a set of R objects; (2) set the analysis parameters, which are the parameters that govern how the statistical analysis is to be conducted; (3) assign gRNAs to cells using one of three principled methods; (4) run quality control; (5) run the calibration check, which is an analysis that verifies that `sceptre` controls the rate of false positives on the dataset under analysis; (6) run the power check, which is an analysis that verifies that `sceptre` is capable of discovering true associations on the dataset under analysis; (7) run the discovery analysis to identify high-confidence perturbation-gene links among the set of perturbations and genes whose association status we do not know but seek to learn; (8) write outputs to a specified directory, including results, plots, and a textual summary of the analysis. `sceptre` leverages several novel statistical and computational algorithms to achieve high statistical accuracy and computational speed.">
<title>Analysis of Single-Cell CRISPR Screen Data • sceptre</title>
<script src="deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.11/clipboard.min.js" integrity="sha512-7O5pXpc0oCRrxk8RUfDYFgn0nO1t+jLuIOQdOMRp4APB7uZ4vSjspzp5y6YDtDs4VzUSTbWzBFZ/LKJhnyFOKw==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="pkgdown.js"></script><meta property="og:title" content="Analysis of Single-Cell CRISPR Screen Data">
<meta property="og:description" content="`sceptre` is an R package for statistically rigorous, massively scalable, and user-friendly single-cell CRISPR screen data analysis. The `sceptre` pipeline consists of several distinct steps: (1) import data from 10X CellRanger, Parse CRISPR Detect, or a set of R objects; (2) set the analysis parameters, which are the parameters that govern how the statistical analysis is to be conducted; (3) assign gRNAs to cells using one of three principled methods; (4) run quality control; (5) run the calibration check, which is an analysis that verifies that `sceptre` controls the rate of false positives on the dataset under analysis; (6) run the power check, which is an analysis that verifies that `sceptre` is capable of discovering true associations on the dataset under analysis; (7) run the discovery analysis to identify high-confidence perturbation-gene links among the set of perturbations and genes whose association status we do not know but seek to learn; (8) write outputs to a specified directory, including results, plots, and a textual summary of the analysis. `sceptre` leverages several novel statistical and computational algorithms to achieve high statistical accuracy and computational speed.">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light" data-bs-theme="light"><div class="container">
    
    <a class="navbar-brand me-2" href="index.html">sceptre</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.10.3</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="articles/sceptre.html">
    <span class="fa fa-play"></span>
     
    Get started
  </a>
</li>
<li class="nav-item">
  <a class="external-link nav-link" href="https://timothy-barry.github.io/sceptre-book/">
    <span class="fa fa-book"></span>
     
    Manual
  </a>
</li>
<li class="nav-item">
  <a class="external-link nav-link" href="https://timothy-barry.github.io/sceptre-book/faq.html">
    <span class="fa fa-question-circle-o"></span>
     
    FAQ
  </a>
</li>
<li class="nav-item">
  <a class="nav-link" href="reference/index.html">
    <span class="fa fa-code"></span>
     
    Functions
  </a>
</li>
<li class="nav-item">
  <a class="nav-link" href="news/index.html">
    <span class="fa fa-history"></span>
     
    Changelog
  </a>
</li>
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/Katsevich-Lab/sceptre/discussions">
    <span class="fa fa-users"></span>
     
    Community
  </a>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav">
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/Katsevich-Lab/sceptre/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>

    
  </div>
</nav><div class="container template-home">
<div class="row">
  <main id="main" class="col-md-9"><div style="margin-top: 5px;">
<p><img src="reference/figures/hex.jpg" align="right" width="150"></p>
</div>
<div class="section level2">
<h2 id="sceptre">
<span style="font-size:60px;"><code>sceptre</code></span><a class="anchor" aria-label="anchor" href="#sceptre"></a>
</h2>
<p><code>sceptre</code> is an R package for single-cell CRISPR screen data analysis, emphasizing statistical rigor, massive scalability, and ease of use.</p>
<!-- badges: start -->

<!-- badges: end -->
</div>
<div class="section level2">
<h2 id="v0100-sceptre-at-massive-scale">v0.10.0: <code>sceptre</code> at massive scale<a class="anchor" aria-label="anchor" href="#v0100-sceptre-at-massive-scale"></a>
</h2>
<p><code>sceptre</code> v0.10.0 represents another a major upgrade to the <code>sceptre</code> software. We have integrated <code>sceptre</code> with <a href="https://timothy-barry.github.io/ondisc/" class="external-link"><code>ondisc</code></a>, a companion R package that facilitates large-scale computing on single-cell data. <code>sceptre</code> now supports the analysis of single-cell CRISPR screen data out-of-core on a laptop or distributed across hundreds of processors on a computing cluster or cloud.</p>
<p><code>sceptre</code> v0.10.0 includes the following updates:</p>
<ul>
<li>Support for disk-backed <code>sceptre</code> objects, enabling the analysis of data too large to fit in memory</li>
<li>A <code>sceptre</code> <a href="https://github.com/timothy-barry/sceptre-pipeline" class="external-link">Nextflow pipeline</a>, enabling the deployment of <code>sceptre</code> across hundreds of processors on a cluster or cloud</li>
<li>An updated <a href="https://timothy-barry.github.io/sceptre-book/" class="external-link">e-book</a> containing new sections about at-scale <code>sceptre</code> and the statistical methodology underlying <code>sceptre</code>
</li>
<li>New functionality for plotting and inspecting gRNA-to-cell assignments</li>
<li>A comprehensive suite of unit tests, which verify the correctness of the code</li>
<li>More detailed man pages (including runnable examples) and minor bug fixes</li>
</ul>
<p>You can see our <a href="https://timothy-barry.github.io/poster_recomb_2024.pdf" class="external-link">RECOMB poster</a> for more information about this update.</p>
</div>
<div class="section level2">
<h2 id="featured-publications">Featured publications<a class="anchor" aria-label="anchor" href="#featured-publications"></a>
</h2>
<ul>
<li>
<a href="https://genomebiology.biomedcentral.com/articles/10.1186/s13059-024-03254-2" class="external-link">Barry et al., 2024</a>. “Robust differential expression testing…”. <em>Genome Biology</em>.</li>
<li>
<a href="https://timothy-barry.github.io/biostatistics_2024.pdf" class="external-link">Barry et al., 2024</a>. “Exponential family measurement error models…”. <em>Biostatistics.</em>
</li>
<li>
<a href="http://sanjanalab.org/reprints/Morris_Science_2023.pdf" class="external-link">Morris et al., 2023</a>. “Discovery of target genes and pathways…”. <em>Science</em>.</li>
<li>
<a href="https://genomebiology.biomedcentral.com/articles/10.1186/s13059-021-02545-2" class="external-link">Barry et al., 2021</a>. “SCEPTRE improves calibration and sensitivity…”. <em>Genome Biology</em>.</li>
</ul>
<p><code>sceptre</code> also recently was featured in a <a href="https://www.10xgenomics.com/analysis-guides/single-cell-crispr-screen-analysis-with-sceptre" class="external-link">10x Genomics analysis guide</a>.</p>
</div>
<div class="section level2">
<h2 id="bug-reports-feature-requests-and-software-questions">Bug reports, feature requests, and software questions<a class="anchor" aria-label="anchor" href="#bug-reports-feature-requests-and-software-questions"></a>
</h2>
<p>For bug reports, please open a <a href="https://github.com/Katsevich-Lab/sceptre/issues" class="external-link">GitHub issue</a>. For questions about <code>sceptre</code> functionality, documentation, or how to apply it to your data, please start a discussion under <a href="https://github.com/Katsevich-Lab/sceptre/discussions/categories/q-a" class="external-link">Q&amp;A</a>.</p>
</div>

  </main><aside class="col-md-3"><div class="links">
<h2 data-toc-skip>Links</h2>
<ul class="list-unstyled">
<li><a href="https://github.com/Katsevich-Lab/sceptre/" class="external-link">Browse source code</a></li>
<li><a href="https://github.com/Katsevich-Lab/sceptre/issues" class="external-link">Report a bug</a></li>
</ul>
</div>

<div class="license">
<h2 data-toc-skip>License</h2>
<ul class="list-unstyled">
<li><a href="LICENSE.html">Full license</a></li>
<li><small><a href="https://www.r-project.org/Licenses/GPL-3" class="external-link">GPL-3</a></small></li>
</ul>
</div>

<div class="citation">
<h2 data-toc-skip>Citation</h2>
<ul class="list-unstyled">
<li><a href="authors.html#citation">Citing sceptre</a></li>
</ul>
</div>

<div class="developers">
<h2 data-toc-skip>Developers</h2>
<ul class="list-unstyled">
<li>Timothy Barry <br><small class="roles"> Author, maintainer </small> <a href="https://orcid.org/0000-0002-4356-627X" target="orcid.widget" aria-label="ORCID" class="external-link"><span class="fab fa-orcid orcid" aria-hidden="true"></span></a> </li>
<li>Louis Deutsch <br><small class="roles"> Author </small>  </li>
<li>Eugene Katsevich <br><small class="roles"> Author </small>  </li>
</ul>
</div>

<div class="funders">
<h2 data-toc-skip>Funders</h2>
<ul class="list-unstyled">
<li>
<p><a href="https://analytics.wharton.upenn.edu/" class="external-link">Analytics at Wharton</a></p>
<p><a href="https://www.nsf.gov/" class="external-link">National Science Foundation</a></p>
<p><a href="https://www.nih.gov" class="external-link">National Institutes of Health</a></p>
</li>
</ul>
</div>

<div class="dev-status">
<h2 data-toc-skip>Dev status</h2>
<ul class="list-unstyled">
<li><a href="https://github.com/Katsevich-Lab/sceptre/actions?query=workflow%3AR-CMD-check+branch%3Amain" class="external-link"><img src="https://img.shields.io/github/actions/workflow/status/Katsevich-Lab/sceptre/check-standard.yaml?branch=main&amp;cacheSeconds=30" alt="R-CMD-check"></a></li>
</ul>
</div>

  </aside>
</div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Timothy Barry, Louis Deutsch, Eugene Katsevich.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.9.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
