<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><meta name="description" content="sceptre"><title>Assign gRNAs • sceptre</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.2.2/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.2.2/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Assign gRNAs"><meta property="og:description" content="sceptre"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">sceptre</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.0.4</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item">
  <a class="external-link nav-link" href="https://timothy-barry.github.io/sceptre-book/sceptre.html">
    <span class="fa fa-play"></span>
     
    Get started
  </a>
</li>
<li class="nav-item">
  <a class="external-link nav-link" href="https://timothy-barry.github.io/sceptre-book/">
    <span class="fa fa-book"></span>
     
    Manual
  </a>
</li>
<li class="nav-item">
  <a class="external-link nav-link" href="https://timothy-barry.github.io/sceptre-book/faq.html">
    <span class="fa fa-question-circle-o"></span>
     
    FAQ
  </a>
</li>
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">
    <span class="fa fa-code"></span>
     
    Functions
  </a>
</li>
<li class="nav-item">
  <a class="nav-link" href="../news/index.html">
    <span class="fa fa-history"></span>
     
    Changelog
  </a>
</li>
      </ul><form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off"></form>

      <ul class="navbar-nav"><li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/Katsevich-Lab/sceptre/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div>

    
  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>Assign gRNAs</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/Katsevich-Lab/sceptre/blob/HEAD/vignettes/assign-grnas.Rmd" class="external-link"><code>vignettes/assign-grnas.Rmd</code></a></small>
      <div class="d-none name"><code>assign-grnas.Rmd</code></div>
    </div>

    
    
<p>The third step of the pipeline is to assign gRNAs to cells. This step
entails using the gRNA UMI counts to determine which cells contain which
gRNAs. <code>sceptre</code> provides three gRNA-to-cell assignment
methods: the “mixture model method,” the “maximum method,” and “the
thresholding method.”</p>
<p><img src="/Users/ekatsevi/code/research/sceptre/vignettes/pipeline_schematic_step_3.png" width="650px" style="display: block; margin: auto;" /></p>
<p>We begin by loading <code>sceptre</code>.</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" tabindex="-1"></a><span class="fu">library</span>(sceptre)</span></code></pre></div>
<div id="initialize-the-crispri-and-crisprko-sceptre_objects"
class="section level2">
<h2>Initialize the CRISPRi and CRISPRko
<code>sceptre_object</code>s</h2>
<p>We use the high-MOI CRISPRi and low-MOI CRISPRko data as running
examples. We call <code>import_data()</code> and
<code>set_analysis_parameters()</code> on the low-MOI CRISPRko data to
initialize a <code>sceptre_object</code> called
<code>sceptre_object_lowmoi</code>.</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" tabindex="-1"></a><span class="co"># low-MOI CRISPRko data setup</span></span>
<span id="cb2-2"><a href="#cb2-2" tabindex="-1"></a><span class="co"># 1. import data</span></span>
<span id="cb2-3"><a href="#cb2-3" tabindex="-1"></a>sceptre_object_lowmoi <span class="ot">&lt;-</span> <span class="fu">import_data</span>(<span class="at">response_matrix =</span> lowmoi_example_data<span class="sc">$</span>response_matrix,</span>
<span id="cb2-4"><a href="#cb2-4" tabindex="-1"></a>                                     <span class="at">grna_matrix =</span> lowmoi_example_data<span class="sc">$</span>grna_matrix,</span>
<span id="cb2-5"><a href="#cb2-5" tabindex="-1"></a>                                     <span class="at">extra_covariates =</span> lowmoi_example_data<span class="sc">$</span>extra_covariates,</span>
<span id="cb2-6"><a href="#cb2-6" tabindex="-1"></a>                                     <span class="at">grna_target_data_frame =</span> lowmoi_example_data<span class="sc">$</span>grna_target_data_frame,</span>
<span id="cb2-7"><a href="#cb2-7" tabindex="-1"></a>                                     <span class="at">moi =</span> <span class="st">&quot;low&quot;</span>)</span>
<span id="cb2-8"><a href="#cb2-8" tabindex="-1"></a></span>
<span id="cb2-9"><a href="#cb2-9" tabindex="-1"></a><span class="co"># 2. set analysis parameters</span></span>
<span id="cb2-10"><a href="#cb2-10" tabindex="-1"></a>positive_control_pairs <span class="ot">&lt;-</span> <span class="fu">construct_positive_control_pairs</span>(sceptre_object_lowmoi)</span>
<span id="cb2-11"><a href="#cb2-11" tabindex="-1"></a>discovery_pairs <span class="ot">&lt;-</span> <span class="fu">construct_trans_pairs</span>(<span class="at">sceptre_object =</span> sceptre_object_lowmoi,</span>
<span id="cb2-12"><a href="#cb2-12" tabindex="-1"></a>                                         <span class="at">positive_control_pairs =</span> positive_control_pairs)</span>
<span id="cb2-13"><a href="#cb2-13" tabindex="-1"></a>sceptre_object_lowmoi <span class="ot">&lt;-</span> <span class="fu">set_analysis_parameters</span>(<span class="at">sceptre_object =</span> sceptre_object_lowmoi,</span>
<span id="cb2-14"><a href="#cb2-14" tabindex="-1"></a>                                                 <span class="at">discovery_pairs =</span> discovery_pairs,</span>
<span id="cb2-15"><a href="#cb2-15" tabindex="-1"></a>                                                 <span class="at">positive_control_pairs =</span> positive_control_pairs)</span></code></pre></div>
<p>We do the same for the high-MOI CRISPRi data, creating
<code>sceptre_object_highmoi</code>.</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" tabindex="-1"></a><span class="co"># high-MOI CRISPRi setup</span></span>
<span id="cb3-2"><a href="#cb3-2" tabindex="-1"></a><span class="co"># 1. import data</span></span>
<span id="cb3-3"><a href="#cb3-3" tabindex="-1"></a>sceptre_object_highmoi <span class="ot">&lt;-</span> <span class="fu">import_data</span>(<span class="at">response_matrix =</span> highmoi_example_data<span class="sc">$</span>response_matrix,</span>
<span id="cb3-4"><a href="#cb3-4" tabindex="-1"></a>                                      <span class="at">grna_matrix =</span> highmoi_example_data<span class="sc">$</span>grna_matrix,</span>
<span id="cb3-5"><a href="#cb3-5" tabindex="-1"></a>                                      <span class="at">grna_target_data_frame =</span> grna_target_data_frame_highmoi,</span>
<span id="cb3-6"><a href="#cb3-6" tabindex="-1"></a>                                      <span class="at">moi =</span> <span class="st">&quot;high&quot;</span>,</span>
<span id="cb3-7"><a href="#cb3-7" tabindex="-1"></a>                                      <span class="at">extra_covariates =</span> highmoi_example_data<span class="sc">$</span>extra_covariates,</span>
<span id="cb3-8"><a href="#cb3-8" tabindex="-1"></a>                                      <span class="at">response_names =</span> highmoi_example_data<span class="sc">$</span>gene_names)</span>
<span id="cb3-9"><a href="#cb3-9" tabindex="-1"></a><span class="co"># 2. set analysis parameters</span></span>
<span id="cb3-10"><a href="#cb3-10" tabindex="-1"></a>positive_control_pairs <span class="ot">&lt;-</span> <span class="fu">construct_positive_control_pairs</span>(sceptre_object_highmoi)</span>
<span id="cb3-11"><a href="#cb3-11" tabindex="-1"></a>discovery_pairs <span class="ot">&lt;-</span> <span class="fu">construct_cis_pairs</span>(sceptre_object_highmoi,</span>
<span id="cb3-12"><a href="#cb3-12" tabindex="-1"></a>                                       <span class="at">positive_control_pairs =</span> positive_control_pairs,</span>
<span id="cb3-13"><a href="#cb3-13" tabindex="-1"></a>                                       <span class="at">distance_threshold =</span> <span class="fl">5e6</span>)</span>
<span id="cb3-14"><a href="#cb3-14" tabindex="-1"></a>sceptre_object_highmoi <span class="ot">&lt;-</span> <span class="fu">set_analysis_parameters</span>(<span class="at">sceptre_object =</span> sceptre_object_highmoi,</span>
<span id="cb3-15"><a href="#cb3-15" tabindex="-1"></a>                                                  <span class="at">discovery_pairs =</span> discovery_pairs,</span>
<span id="cb3-16"><a href="#cb3-16" tabindex="-1"></a>                                                  <span class="at">positive_control_pairs =</span> positive_control_pairs,</span>
<span id="cb3-17"><a href="#cb3-17" tabindex="-1"></a>                                                  <span class="at">side =</span> <span class="st">&quot;left&quot;</span>)</span></code></pre></div>
<p>We are now ready to carry out the gRNA assignment step on both
datasets.</p>
</div>
<div id="visualize-the-grna-count-distributions" class="section level2">
<h2>Visualize the gRNA count distributions</h2>
<p>A helpful first step in assigning gRNAs to cells is to visualize the
gRNA UMI count distributions. As discussed in the Get Started vignette
(<code>vignette("sceptre")</code>), we can use the function
<code>plot_grna_count_distributions()</code> to plot the empirical UMI
count distribution of one or more gRNAs.
<code>plot_grna_count_distributions()</code> takes several arguments:
<code>sceptre_object</code> (required), <code>n_grnas_to_plot</code>
(optional), <code>grnas_to_plot</code> (optional), and
<code>threshold</code> (optional). <code>n_grnas_to_plot</code> is an
integer specifying the number of randomly-selected gRNAs to plot.
<code>grnas_to_plot</code> is a character vector specifying (by name)
one or more specific gRNAs to plot. Finally, <code>threshold</code> is
an integer specifying the location at which to draw a dotted vertical
line. We call <code>plot_grna_count_distributions()</code> on the
low-MOI CRISPRko data, plotting nine random gRNAs.</p>
<div class="sourceCode" id="cb4"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" tabindex="-1"></a><span class="fu">plot_grna_count_distributions</span>(<span class="at">sceptre_object =</span> sceptre_object_lowmoi,</span>
<span id="cb4-2"><a href="#cb4-2" tabindex="-1"></a>                              <span class="at">n_grnas_to_plot =</span> <span class="dv">9</span>)</span></code></pre></div>
<div class="figure" style="text-align: center">
<img src="/Users/ekatsevi/code/research/sceptre/vignettes/grna_count_distribution_step3.png" alt="Histograms of the gRNA count distributions for the low-MOI CRISPRko data." width="750px" />
<p class="caption">
Histograms of the gRNA count distributions for the low-MOI CRISPRko
data.
</p>
</div>
<p>We see that the gRNA count distributions exhibit generally bimodal
behavior. As discussed in Get Started, this bimodality is due to the
phenomenon of background contamination: gRNA reads sometimes map to
cells that do not contain the corresponding gRNA. <code>sceptre</code>
provides three methods for assigning gRNAs to cells, all of which
account for background contamination.</p>
</div>
<div id="assign-grnas-to-cells" class="section level2">
<h2>Assign gRNAs to cells</h2>
<p>We assign gRNAs to cells to cells by calling the function
<code>assign_grnas()</code>. <code>assign_grnas()</code> takes the
arguments <code>sceptre_object</code> (required), <code>method</code>
(optional), <code>print_progress</code> (optional), and
<code>parallel</code> (optional). <code>method</code> is the gRNA
assignment method and should be set to <code>"mixture"</code>,
<code>"maximum"</code>, or <code>"thresholding"</code>.
<code>print_progress</code> (default <code>TRUE</code>) and
<code>parallel</code> (default <code>FALSE</code>) are logical values
specifying whether to print progress updates and run the function in
parallel, respectively. Note that parallel computation is not yet
configured for Windows; thus, Windows users either should omit the
<code>parallel</code> argument or set <code>parallel</code> to
<code>FALSE</code>. Below, we describe the three gRNA assignment
strategies in detail.</p>
<div id="mixture-method" class="section level3">
<h3>Mixture method</h3>
<p>The first gRNA assignment strategy is the <code>"mixture"</code>
method, which involves assigning gRNAs to cells using a mixture model.
The mixture method is the default method for high-MOI screens and is an
optional method for low-MOI screens. The method works as follows. First,
we fit a latent variable Poisson GLM to the data, regressing the gRNA
UMI count vector onto the (latent) gRNA indicator vector and
cell-specific covariate matrix. (A given entry of the gRNA indicator
vector is defined to be “1” if the gRNA is present in the corresponding
cell and “0” otherwise.) We fit the latent variable Poisson GLM using a
novel variant of the EM algorithm. The fitted model yields the
probability that each cell contains the gRNA; we threshold these
probabilities to assign the gRNA to cells. An advantage of the GLM-based
mixture modeling approach is that it accounts for cell-specific
covariates, such as sequencing depth and batch. For example, cells that
are sequenced deeply (and have a large value for
<code>grna_n_umis</code> or <code>response_n_umis</code>) generally
exhibit higher levels of ambient background contamination;
<code>sceptre</code> controls for heterogeneity across cells due to
sequencing depth and other factors.</p>
<p>We use the mixture assignment method to assign gRNAs to cells on both
the high-MOI CRISPRi data and low-MOI CRISPRko data, saving the
resulting outputs as <code>sceptre_object_lowmoi_mixture</code> and
<code>sceptre_object_highmoi_mixture</code>, respectively.</p>
<div class="sourceCode" id="cb5"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" tabindex="-1"></a><span class="co"># high-MOI CRISPRi data</span></span>
<span id="cb5-2"><a href="#cb5-2" tabindex="-1"></a>sceptre_object_highmoi_mixture <span class="ot">&lt;-</span> <span class="fu">assign_grnas</span>(<span class="at">sceptre_object =</span> sceptre_object_highmoi,</span>
<span id="cb5-3"><a href="#cb5-3" tabindex="-1"></a>                                               <span class="at">method =</span> <span class="st">&quot;mixture&quot;</span>, <span class="at">parallel =</span> <span class="cn">TRUE</span>)</span></code></pre></div>
<div class="sourceCode" id="cb6"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" tabindex="-1"></a><span class="co"># low-MOI CRISPRko data</span></span>
<span id="cb6-2"><a href="#cb6-2" tabindex="-1"></a>sceptre_object_lowmoi_mixture <span class="ot">&lt;-</span> <span class="fu">assign_grnas</span>(<span class="at">sceptre_object =</span> sceptre_object_lowmoi,</span>
<span id="cb6-3"><a href="#cb6-3" tabindex="-1"></a>                                              <span class="at">method =</span> <span class="st">&quot;mixture&quot;</span>, <span class="at">parallel =</span> <span class="cn">TRUE</span>)</span></code></pre></div>
<p>We can pass optional, method-specific parameters to
<code>assign_grnas()</code> to control precisely how the assignment
method is deployed. The relevant parameters for the mixture method
include <code>formula_object</code>, <code>n_em_rep</code>,
<code>n_nonzero_cells_cutoff</code>, <code>backup_threshold</code>, and
<code>probability_threshold</code>. <code>formula_object</code> is a
formula object that specifies how <code>sceptre</code> is to adjust for
the cell-specific covariates in the latent-variable gRNA count model.
The default formula object is constructed by summing over all covariates
and then log-transforming the count-based covariates. It is sometimes
helpful to exclude covariates from the formula object so as to improve
the speed of the gRNA assignment step. For example, we define a reduced
formula object <code>formula_object_reduced</code> that includes only
<code>grna_n_nonzero</code> and <code>grna_n_umis</code>, the two most
important covariates for assigning gRNAs to cells.</p>
<div class="sourceCode" id="cb7"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1" tabindex="-1"></a>formula_object_reduced <span class="ot">&lt;-</span> <span class="fu">formula</span>(<span class="sc">~</span><span class="fu">log</span>(grna_n_nonzero) <span class="sc">+</span> <span class="fu">log</span>(grna_n_umis))</span></code></pre></div>
<p><code>n_em_rep</code> is the number of times to run the EM algorithm
using random starting estimates (default: <code>n_em_rep = 5</code>).
Setting <code>n_em_rep</code> to a larger integer can improve the
accuracy of the gRNA assignments at the cost of increasing compute.
<code>n_nonzero_cells_cutoff</code> is the minimum number of cells that
must exhibit nonzero expression of the gRNA to attempt fitting a mixture
model to the gRNA count distribution (default:
<code>n_nonzero_cells_cutoff = 10</code>). If the gRNA is expressed in
fewer than <code>n_nonzero_cells_cutoff</code> cells, the gRNA instead
is assigned to cells via the thresholding method, where the threshold
used is <code>backup_threshold</code> (default:
<code>backup_threshold = 5</code>). (See below for a detailed discussion
of the thresholding method.) Finally, cells whose estimated probability
of having received a gRNA exceeds <code>probability_threshold</code> are
called as containing the gRNA (default:
<code>probability_threshold = 0.8</code>). In practice
<code>probability_threshold</code> typically does not impact the results
considerably.</p>
<p>To illustrate use of these method-specific parameters, we again call
<code>assign_grnas()</code> to assign gRNAs to cells on the high-MOI
CRISPRi data, this time setting <code>formula_object</code> to
<code>formula_object_reduced</code>.</p>
<div class="sourceCode" id="cb8"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1" tabindex="-1"></a>sceptre_object_highmoi_mixture <span class="ot">&lt;-</span> <span class="fu">assign_grnas</span>(<span class="at">sceptre_object =</span> sceptre_object_highmoi,</span>
<span id="cb8-2"><a href="#cb8-2" tabindex="-1"></a>                                               <span class="at">formula_object =</span> formula_object_reduced,</span>
<span id="cb8-3"><a href="#cb8-3" tabindex="-1"></a>                                               <span class="at">parallel =</span> <span class="cn">TRUE</span>)</span></code></pre></div>
</div>
<div id="maximum-method" class="section level3">
<h3>Maximum method</h3>
<p>The second gRNA assignment strategy is the <code>"maximum"</code>
method. The maximum method is the default assignment method for low-MOI
screens and is not available as an option for high-MOI screens. The
maximum method assigns the gRNA that accounts for the greatest number of
UMIs in a given cell to that cell. We apply the maximum method to the
low-MOI CRISPRko data below.</p>
<div class="sourceCode" id="cb9"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1" tabindex="-1"></a>sceptre_object_lowmoi_maximum <span class="ot">&lt;-</span> <span class="fu">assign_grnas</span>(sceptre_object_lowmoi, <span class="at">method =</span> <span class="st">&quot;maximum&quot;</span>)</span></code></pre></div>
<p>The maximum method allows for one optional argument:
<code>umi_fraction_threshold</code> (default:
<code>umi_fraction_threshold = 0.8</code>). Cells for which the
maximally expressed gRNA constitutes fewer than
<code>umi_fraction_threshold</code> of the UMIs in that cell are flagged
as containing multiple gRNAs. (Cells containing multiple gRNAs are
removed as part of low-MOI quality control step, as discussed later).
For example, if the maximally expressed gRNA in a given cell makes up
0.64 (or 64<span class="math inline">\(\%\)</span>) of the UMIs in that
cell, then that cell is flagged as containing multiple gRNAs and is
removed during quality control (assuming that
<code>umi_fraction_threshold = 0.8</code>).</p>
</div>
<div id="thresholding-method" class="section level3">
<h3>Thresholding method</h3>
<p>The third method for assigning gRNAs to cells is the
<code>"thresholding"</code> method; this method is available in both
low- and high-MOI settings. The thresholding method assigns a gRNA to a
cell if the UMI count of the gRNA in the cell is greater than or equal
to some integer threshold (by default 5). We apply the
<code>"thresholding"</code> method to both CRISPRi and CRISPRko datasets
as follows.</p>
<div class="sourceCode" id="cb10"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1" tabindex="-1"></a><span class="co"># high-MOI CRISPRi data</span></span>
<span id="cb10-2"><a href="#cb10-2" tabindex="-1"></a>sceptre_object_highmoi_thresholding <span class="ot">&lt;-</span> <span class="fu">assign_grnas</span>(<span class="at">sceptre_object =</span> sceptre_object_highmoi,</span>
<span id="cb10-3"><a href="#cb10-3" tabindex="-1"></a>                                                    <span class="at">method =</span> <span class="st">&quot;thresholding&quot;</span>)</span>
<span id="cb10-4"><a href="#cb10-4" tabindex="-1"></a><span class="co"># low-MOI CRISPRko data</span></span>
<span id="cb10-5"><a href="#cb10-5" tabindex="-1"></a>sceptre_object_lowmoi_thresholding <span class="ot">&lt;-</span> <span class="fu">assign_grnas</span>(<span class="at">sceptre_object =</span> sceptre_object_lowmoi,</span>
<span id="cb10-6"><a href="#cb10-6" tabindex="-1"></a>                                                   <span class="at">method =</span> <span class="st">&quot;thresholding&quot;</span>)</span></code></pre></div>
<p>The thresholding method allows for one optional argument, namely the
integer threshold <code>threshold</code>. An important special case is
to set <code>threshold</code> to <code>1</code>, which causes any gRNA
expressed in a given cell to be assigned to that cell. We assign gRNAs
to cells on the CRISPRi data using a threshold of 1 below.</p>
<div class="sourceCode" id="cb11"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1" tabindex="-1"></a>sceptre_object_highmoi_thresholding <span class="ot">&lt;-</span> <span class="fu">assign_grnas</span>(<span class="at">sceptre_object =</span> sceptre_object_highmoi,</span>
<span id="cb11-2"><a href="#cb11-2" tabindex="-1"></a>                                                    <span class="at">method =</span> <span class="st">&quot;thresholding&quot;</span>,</span>
<span id="cb11-3"><a href="#cb11-3" tabindex="-1"></a>                                                    <span class="at">threshold =</span> <span class="dv">1</span>)</span></code></pre></div>
<p>Note that setting <code>threshold</code> to <code>1</code> causes
background contamination to be ignored.</p>
</div>
</div>
<div
id="plotting-and-printing-the-outcome-of-the-grna-to-cell-assignment-step"
class="section level2">
<h2>Plotting and printing the outcome of the gRNA-to-cell assignment
step</h2>
<p>We can call <code>plot()</code> on the <code>sceptre_object</code> to
render a plot summarizing the outcome of the gRNA-to-cell assignment
step. <code>plot()</code> takes the arguments
<code>sceptre_object</code> (required) <code>n_grnas_to_plot</code>
(optional), <code>grnas_to_plot</code> (optional), and
<code>return_indiv_plots</code> (optional). <code>n_grnas_to_plot</code>
is the number of (randomly-selected) gRNAs to plot;
<code>grnas_to_plot</code> is a vector of names of one or more gRNAs to
plot; and <code>return_indiv_plots</code> is a logical value indicating
whether to return the constituent panels of the plot individually
(<code>TRUE</code>) or combined into a single figure
(<code>FALSE</code>; default).</p>
<p>We visualize the outcome of the gRNA-to-cell assignment step on the
low-MOI CRISPRko data based on the maximum assignment strategy. We
specifically plot the gRNAs “IFNGR2g2”, “STAT1g2”, and “STAT5Ag4”.
(These gRNAs were selected arbitrarily for the purpose of illustrating
the <code>grnas_to_plot</code> argument.)</p>
<div class="sourceCode" id="cb12"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb12-1"><a href="#cb12-1" tabindex="-1"></a>grnas_to_plot <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">&quot;IFNGR2g2&quot;</span>, <span class="st">&quot;STAT1g2&quot;</span>, <span class="st">&quot;STAT5Ag4&quot;</span>)</span>
<span id="cb12-2"><a href="#cb12-2" tabindex="-1"></a><span class="fu">plot</span>(sceptre_object_lowmoi_maximum, <span class="at">grnas_to_plot =</span> grnas_to_plot)</span></code></pre></div>
<div class="figure" style="text-align: center">
<img src="/Users/ekatsevi/code/research/sceptre/vignettes/lowmoi_maximum_assignment_step3.png" alt="Maximum assignment method on the low-MOI CRISPRko data." width="650px" />
<p class="caption">
Maximum assignment method on the low-MOI CRISPRko data.
</p>
</div>
<p>The Get Started vignette (<code>vignette("sceptre")</code>) describes
how to interpret this plot. Briefly, the top panel displays the
gRNA-to-cell assignments of several individual gRNAS, and the bottom
left panel shows the number of cells per gRNA. Typically, the bottom
right panel plots the number of gRNAs per cell, but the bottom right
panel is not meaningful under the maximum assignment strategy and is
thus omitted.</p>
<p>It is informative to compare the output of multiple gRNA assignment
methods and check for consistency across methods. To this end we render
a visualization of the gRNA-to-cell assignments on the same dataset,
this time based on the mixture method.</p>
<div class="sourceCode" id="cb13"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb13-1"><a href="#cb13-1" tabindex="-1"></a><span class="fu">plot</span>(sceptre_object_lowmoi_mixture, <span class="at">grnas_to_plot =</span> grnas_to_plot)</span></code></pre></div>
<div class="figure" style="text-align: center">
<img src="/Users/ekatsevi/code/research/sceptre/vignettes/lowmoi_mixture_assignment_step3.png" alt="Mixture assignment method on the low-MOI CRISPRko data." width="650px" />
<p class="caption">
Mixture assignment method on the low-MOI CRISPRko data.
</p>
</div>
<p>Encouragingly, the gRNA assignments appear to be highly concordant
across the maximum and mixture methods. Suppose that we ultimately
choose to use the mixture strategy to assign gRNAs to cells on the
low-MOI CRISPRko dataset. We update <code>sceptre_object_lowmoi</code>
as follows.</p>
<div class="sourceCode" id="cb14"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb14-1"><a href="#cb14-1" tabindex="-1"></a>sceptre_object_lowmoi <span class="ot">&lt;-</span> sceptre_object_lowmoi_mixture</span></code></pre></div>
<p>We can call <code>print()</code> on
<code>sceptre_object_lowmoi</code> to print a summary tracking the
progress of the analysis. The field “gRNA-to-cell assignment
information” presents information about the gRNA assignment step,
including the selected assignment method, the mean number of cells per
gRNA, and the mean number of gRNAs per cell.</p>
<div class="sourceCode" id="cb15"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb15-1"><a href="#cb15-1" tabindex="-1"></a><span class="fu">print</span>(sceptre_object_lowmoi_mixture)</span>
<span id="cb15-2"><a href="#cb15-2" tabindex="-1"></a><span class="co">#&gt; An object of class [34msceptre_object[39m.</span></span>
<span id="cb15-3"><a href="#cb15-3" tabindex="-1"></a><span class="co">#&gt; </span></span>
<span id="cb15-4"><a href="#cb15-4" tabindex="-1"></a><span class="co">#&gt; Attributes of the data:</span></span>
<span id="cb15-5"><a href="#cb15-5" tabindex="-1"></a><span class="co">#&gt;  • [34m20729[39m cells</span></span>
<span id="cb15-6"><a href="#cb15-6" tabindex="-1"></a><span class="co">#&gt;  • [34m299[39m responses</span></span>
<span id="cb15-7"><a href="#cb15-7" tabindex="-1"></a><span class="co">#&gt;  • [34mLow[39m multiplicity-of-infection </span></span>
<span id="cb15-8"><a href="#cb15-8" tabindex="-1"></a><span class="co">#&gt;  • [34m101[39m targeting gRNAs (distributed across [34m26[39m targets) </span></span>
<span id="cb15-9"><a href="#cb15-9" tabindex="-1"></a><span class="co">#&gt;  • [34m9[39m non-targeting gRNAs </span></span>
<span id="cb15-10"><a href="#cb15-10" tabindex="-1"></a><span class="co">#&gt;  • [34m6[39m covariates (bio_rep, grna_n_nonzero, grna_n_umis, response_n_nonzero, response_n_umis, response_p_mito)</span></span>
<span id="cb15-11"><a href="#cb15-11" tabindex="-1"></a><span class="co">#&gt; </span></span>
<span id="cb15-12"><a href="#cb15-12" tabindex="-1"></a><span class="co">#&gt; Analysis status:</span></span>
<span id="cb15-13"><a href="#cb15-13" tabindex="-1"></a><span class="co">#&gt;  [32m✓[39m import_data()</span></span>
<span id="cb15-14"><a href="#cb15-14" tabindex="-1"></a><span class="co">#&gt;  [32m✓[39m set_analysis_parameters()</span></span>
<span id="cb15-15"><a href="#cb15-15" tabindex="-1"></a><span class="co">#&gt;  [32m✓[39m assign_grnas()</span></span>
<span id="cb15-16"><a href="#cb15-16" tabindex="-1"></a><span class="co">#&gt;  [31m✗[39m run_qc()</span></span>
<span id="cb15-17"><a href="#cb15-17" tabindex="-1"></a><span class="co">#&gt;  [31m✗[39m run_calibration_check()</span></span>
<span id="cb15-18"><a href="#cb15-18" tabindex="-1"></a><span class="co">#&gt;  [31m✗[39m run_power_check()</span></span>
<span id="cb15-19"><a href="#cb15-19" tabindex="-1"></a><span class="co">#&gt;  [31m✗[39m run_discovery_analysis()</span></span>
<span id="cb15-20"><a href="#cb15-20" tabindex="-1"></a><span class="co">#&gt; </span></span>
<span id="cb15-21"><a href="#cb15-21" tabindex="-1"></a><span class="co">#&gt; Analysis parameters: </span></span>
<span id="cb15-22"><a href="#cb15-22" tabindex="-1"></a><span class="co">#&gt;  • Discovery pairs: data frame with [34m7765[39m pairs</span></span>
<span id="cb15-23"><a href="#cb15-23" tabindex="-1"></a><span class="co">#&gt;  • Positive control pairs: data frame with [34m9[39m pairs</span></span>
<span id="cb15-24"><a href="#cb15-24" tabindex="-1"></a><span class="co">#&gt;  • Sidedness of test: [34mboth[39m</span></span>
<span id="cb15-25"><a href="#cb15-25" tabindex="-1"></a><span class="co">#&gt;  • N nonzero treatment cells threshold: not specified</span></span>
<span id="cb15-26"><a href="#cb15-26" tabindex="-1"></a><span class="co">#&gt;  • N nonzero control cells threshold: not specified</span></span>
<span id="cb15-27"><a href="#cb15-27" tabindex="-1"></a><span class="co">#&gt;  • Control group: [34mnon-targeting cells[39m</span></span>
<span id="cb15-28"><a href="#cb15-28" tabindex="-1"></a><span class="co">#&gt;  • Resampling mechanism: [34mpermutations[39m</span></span>
<span id="cb15-29"><a href="#cb15-29" tabindex="-1"></a><span class="co">#&gt;  • gRNA grouping strategy: [34munion[39m</span></span>
<span id="cb15-30"><a href="#cb15-30" tabindex="-1"></a><span class="co">#&gt;  • Formula object: [34mlog(response_n_nonzero) + log(response_n_umis) + bio_rep[39m</span></span>
<span id="cb15-31"><a href="#cb15-31" tabindex="-1"></a><span class="co">#&gt; </span></span>
<span id="cb15-32"><a href="#cb15-32" tabindex="-1"></a><span class="co">#&gt; gRNA-to-cell assignment information:</span></span>
<span id="cb15-33"><a href="#cb15-33" tabindex="-1"></a><span class="co">#&gt;  • Assignment method: [34mmixture[39m</span></span>
<span id="cb15-34"><a href="#cb15-34" tabindex="-1"></a><span class="co">#&gt;  • Mean N cells per gRNA: [34m205.07[39m</span></span>
<span id="cb15-35"><a href="#cb15-35" tabindex="-1"></a><span class="co">#&gt;  • Mean N gRNAs per cell (MOI): [34m1.09[39m</span></span></code></pre></div>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside></div>



    <footer><div class="pkgdown-footer-left">
  <p></p><p>Developed by Timothy Barry, Eugene Katsevich.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer></div>

  

  

  </body></html>
