<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="sceptre">
<title>High MOI sceptre tutorial • sceptre</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.2.2/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.2.2/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.rawgit.com/afeld/bootstrap-toc/v1.0.1/dist/bootstrap-toc.min.js"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="High MOI sceptre tutorial">
<meta property="og:description" content="sceptre">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-default navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">sceptre</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="Released version">0.2.0</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="active nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="../articles/highmoi_tutorial.html">High MOI sceptre tutorial</a>
    <a class="dropdown-item" href="../articles/lowmoi_tutorial.html">Low MOI sceptre tutorial</a>
    <a class="dropdown-item" href="../articles/quality_control_in_low_moi.html">Low MOI Quality Control</a>
  </div>
</li>
<li class="nav-item">
  <a class="nav-link" href="../news/index.html">Changelog</a>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav"></ul>
</div>

    
  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>High MOI sceptre tutorial</h1>
                        <h4 data-toc-skip class="author">Tim Barry &amp;
Gene Katsevich</h4>
            
            <h4 data-toc-skip class="date">April 2023</h4>
      
      
      <div class="d-none name"><code>highmoi_tutorial.Rmd</code></div>
    </div>

    
    
<p>This vignette illustrates application of <code>sceptre</code> to an
example high-multiplicity-of-infection (MOI) single-cell CRISPR screen
dataset. We begin by installing and loading all required packages,
including the up-to-date version of <code>sceptre</code>.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"devtools"</span><span class="op">)</span></span>
<span><span class="fu">devtools</span><span class="fu">::</span><span class="fu"><a href="https://remotes.r-lib.org/reference/install_github.html" class="external-link">install_github</a></span><span class="op">(</span><span class="st">"katsevich-lab/sceptre"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">packages</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"doParallel"</span>, <span class="st">"foreach"</span>, <span class="st">"fst"</span>, <span class="st">"purrr"</span>,</span>
<span>              <span class="st">"R.utils"</span>, <span class="st">"scales"</span>, <span class="st">"sn"</span>, <span class="st">"tidyr"</span>, <span class="st">"VGAM"</span><span class="op">)</span></span>
<span><span class="kw">for</span> <span class="op">(</span><span class="va">pack</span> <span class="kw">in</span> <span class="va">packages</span><span class="op">)</span> <span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="va">pack</span><span class="op">)</span></span></code></pre></div>
<p>We <code>library</code> the packages whose functions we will be
calling in the tutorial.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://dplyr.tidyverse.org" class="external-link">dplyr</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org" class="external-link">ggplot2</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://wilkelab.org/cowplot/" class="external-link">cowplot</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://Matrix.R-forge.R-project.org/" class="external-link">Matrix</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">sceptre</span><span class="op">)</span></span></code></pre></div>
<p>We proceed in seven steps.</p>
<div class="section level2">
<h2 id="step-1-prepare-the-data">Step 1: Prepare the data<a class="anchor" aria-label="anchor" href="#step-1-prepare-the-data"></a>
</h2>
<p>The first step is to prepare the data to pass to
<code>sceptre</code>. We must prepare three separate data objects: the
gene-by-cell expression matrix, the gRNA-by-cell expression matrix, and
the cell-specific matrix of covariates.</p>
<div class="section level3">
<h3 id="gene-and-grna-expression-matrices">Gene and gRNA expression matrices<a class="anchor" aria-label="anchor" href="#gene-and-grna-expression-matrices"></a>
</h3>
<p>We load the example gene-by-cell and gRNA-by-cell expression matrices
that are included in the <code>sceptre</code> package.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">gene_matrix_highmoi</span><span class="op">)</span></span>
<span><span class="va">gene_matrix</span> <span class="op">&lt;-</span> <span class="va">gene_matrix_highmoi</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">gRNA_matrix_highmoi</span><span class="op">)</span></span>
<span><span class="va">gRNA_matrix</span> <span class="op">&lt;-</span> <span class="va">gRNA_matrix_highmoi</span></span></code></pre></div>
<p>Briefly, <code>gene_matrix</code> (respectively,
<code>gRNA_matrix</code>) is a 20 x 40,000 (respectively, 50 x 40,000)
matrix of gene (respectively, gRNA) unique molecular identifier (UMI)
counts. The data are taken from the <a href="https://www.sciencedirect.com/science/article/pii/S009286741831554X" class="external-link">paper</a>
“A genome-wide framework for mapping gene regulation via cellular
genetic screens” by Gasperini et al., 2019. The authors used a
CRISPRi-based assay to target 5,000+ putative enhancers in a population
of K562 cells. The authors additionally targeted 200+ gene transcription
start sites (TSSs) and designed a library of 50 non-targeting gRNAs to
serve as negative controls. Genes, gRNAs, and cells are all down-sampled
to reduce the size of the data. One can use the commands
<code>?gene_matrix</code> and <code>?gRNA_matrix</code> to read more
about these data.</p>
<p>The row names of <code>gene_matrix</code> and
<code>gRNA_matrix</code> are the gene IDs and gRNA IDs, respectively.
The gRNA IDs are gRNA-specific oligonucleotide barcodes. In general the
gRNA IDs must be strings that uniquely identify each gRNA. The column
names, meanwhile, are the cell barcodes. Next, <code>gene_matrix</code>
and <code>gRNA_matrix</code> are sparse matrices (as implemented by the
<code>Matrix</code> package). In general these matrices can be either
sparse matrices or standard (dense) R matrices. We print a few rows and
columns of <code>gene_matrix</code> and <code>gRNA_matrix</code> to get
a sense of what the data look like.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/options.html" class="external-link">options</a></span><span class="op">(</span>width <span class="op">=</span> <span class="fl">300</span><span class="op">)</span></span>
<span><span class="co"># gene_matrix; rows are genes, columns are cells</span></span>
<span><span class="va">gene_matrix</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">10</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">3</span><span class="op">]</span></span>
<span><span class="co">#&gt; 10 x 3 sparse Matrix of class "dgTMatrix"</span></span>
<span><span class="co">#&gt;                 CTCGAAAGTCCATGAT-1_2B_7 GACACGCTCAAAGACA-1_2B_4 GGCTCGATCTGTGCAA-1_1B_1</span></span>
<span><span class="co">#&gt; ENSG00000111057                       1                       5                       1</span></span>
<span><span class="co">#&gt; ENSG00000121211                       .                       1                       .</span></span>
<span><span class="co">#&gt; ENSG00000134809                       3                      10                       1</span></span>
<span><span class="co">#&gt; ENSG00000166902                       1                       3                       .</span></span>
<span><span class="co">#&gt; ENSG00000172809                      12                      19                      33</span></span>
<span><span class="co">#&gt; ENSG00000105298                       .                       .                       .</span></span>
<span><span class="co">#&gt; ENSG00000108389                       .                       1                       .</span></span>
<span><span class="co">#&gt; ENSG00000157985                       .                       .                       .</span></span>
<span><span class="co">#&gt; ENSG00000107731                       .                       .                       .</span></span>
<span><span class="co">#&gt; ENSG00000184357                       .                       1                       .</span></span>
<span></span>
<span><span class="co"># gRNA matrix; rows are gRNAs, columns are cells</span></span>
<span><span class="va">gRNA_matrix</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">10</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">3</span><span class="op">]</span></span>
<span><span class="co">#&gt; 10 x 3 sparse Matrix of class "dgTMatrix"</span></span>
<span><span class="co">#&gt;                      CTCGAAAGTCCATGAT-1_2B_7 GACACGCTCAAAGACA-1_2B_4 GGCTCGATCTGTGCAA-1_1B_1</span></span>
<span><span class="co">#&gt; TGTCAGTCCTCCCTCCCCCA                       .                       .                       .</span></span>
<span><span class="co">#&gt; CTAAACAGTGAAAGTCACAG                       .                       .                       .</span></span>
<span><span class="co">#&gt; GCTCCAATCATATTCTAGAG                       .                       .                       .</span></span>
<span><span class="co">#&gt; GATGAAACCTAAGGGCACAC                       .                       .                       .</span></span>
<span><span class="co">#&gt; TGCAGACGAGTGTCTCAGAG                       .                       .                       .</span></span>
<span><span class="co">#&gt; TCATCTTGAAGTCAGCTCCA                       .                       .                       3</span></span>
<span><span class="co">#&gt; GAGAATGGCATGGAGCTCAA                       .                       .                       .</span></span>
<span><span class="co">#&gt; AGAGCAAGAGAGCAACTCCG                       .                       .                       .</span></span>
<span><span class="co">#&gt; CAGCAATTCTACCTTCAGGT                       .                       .                       .</span></span>
<span><span class="co">#&gt; TTTGGTTGTTGCAAATGAGG                       .                       .                       .</span></span></code></pre></div>
<p>We also plot a histogram of the counts of an arbitrarily selected
gene (“ENSG00000111057”) and gRNA (“TGTCAGTCCTCCCTCCCCCA”) to visualize
the data.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">example_gene</span> <span class="op">&lt;-</span> <span class="va">gene_matrix</span><span class="op">[</span><span class="st">"ENSG00000111057"</span>,<span class="op">]</span></span>
<span><span class="va">example_gRNA</span> <span class="op">&lt;-</span> <span class="va">gRNA_matrix</span><span class="op">[</span><span class="st">"TGTCAGTCCTCCCTCCCCCA"</span>,<span class="op">]</span></span>
<span></span>
<span><span class="va">hist_gene</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span>data <span class="op">=</span> <span class="fu"><a href="https://tibble.tidyverse.org/reference/tibble.html" class="external-link">tibble</a></span><span class="op">(</span>count <span class="op">=</span> <span class="va">example_gene</span><span class="op">)</span> <span class="op">|&gt;</span> <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">count</span> <span class="op">&gt;=</span> <span class="fl">1</span>, <span class="va">count</span> <span class="op">&lt;=</span> <span class="fl">40</span><span class="op">)</span>, mapping <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">count</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_histogram.html" class="external-link">geom_histogram</a></span><span class="op">(</span>binwidth <span class="op">=</span> <span class="fl">3</span>, col <span class="op">=</span> <span class="st">"black"</span>, fill <span class="op">=</span> <span class="st">"dodgerblue3"</span>, alpha <span class="op">=</span> <span class="fl">0.7</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html" class="external-link">scale_y_continuous</a></span><span class="op">(</span>trans<span class="op">=</span><span class="st">'log10'</span>, expand <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0</span>, <span class="cn">NA</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">xlab</a></span><span class="op">(</span><span class="st">"Gene expressions"</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ylab</a></span><span class="op">(</span><span class="st">""</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_bw</a></span><span class="op">(</span>base_size <span class="op">=</span> <span class="fl">10</span><span class="op">)</span></span>
<span></span>
<span><span class="va">hist_gRNA</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span>data <span class="op">=</span> <span class="fu"><a href="https://tibble.tidyverse.org/reference/tibble.html" class="external-link">tibble</a></span><span class="op">(</span>count <span class="op">=</span> <span class="va">example_gRNA</span><span class="op">)</span> <span class="op">|&gt;</span> <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">count</span> <span class="op">&gt;=</span> <span class="fl">1</span>, <span class="va">count</span> <span class="op">&lt;=</span> <span class="fl">40</span><span class="op">)</span>, mapping <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">count</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_histogram.html" class="external-link">geom_histogram</a></span><span class="op">(</span>binwidth <span class="op">=</span> <span class="fl">3</span>, col <span class="op">=</span> <span class="st">"black"</span>, fill <span class="op">=</span> <span class="st">"orchid4"</span>, alpha <span class="op">=</span> <span class="fl">0.7</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html" class="external-link">scale_y_continuous</a></span><span class="op">(</span>trans<span class="op">=</span><span class="st">'log10'</span>, expand <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0</span>, <span class="cn">NA</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">xlab</a></span><span class="op">(</span><span class="st">"gRNA expressions"</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ylab</a></span><span class="op">(</span><span class="st">""</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_bw</a></span><span class="op">(</span>base_size <span class="op">=</span> <span class="fl">10</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://wilkelab.org/cowplot/reference/plot_grid.html" class="external-link">plot_grid</a></span><span class="op">(</span><span class="va">hist_gene</span>, <span class="va">hist_gRNA</span>, labels <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"a"</span>, <span class="st">"b"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="highmoi_tutorial_files/figure-html/plot_hist-1.png" width="672"></p>
<p>As expected, the data are highly discrete counts. Note that we do
<em>not</em> normalize either the gene or gRNA expression matrices,
opting instead to work directly with the raw counts.</p>
</div>
<div class="section level3">
<h3 id="cell-wise-covariate-matrix">Cell-wise covariate matrix<a class="anchor" aria-label="anchor" href="#cell-wise-covariate-matrix"></a>
</h3>
<p>Next, we load the cell-wise covariate matrix,
<code>covariate_matrix</code>.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">covariate_matrix_highmoi</span><span class="op">)</span></span>
<span><span class="va">covariate_matrix</span> <span class="op">&lt;-</span> <span class="va">covariate_matrix_highmoi</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">covariate_matrix</span><span class="op">)</span></span>
<span><span class="co">#&gt;                         lg_gRNA_lib_size lg_gene_lib_size     p_mito        batch</span></span>
<span><span class="co">#&gt; CTCGAAAGTCCATGAT-1_2B_7         6.023448         9.276783 0.03861979 prep_batch_2</span></span>
<span><span class="co">#&gt; GACACGCTCAAAGACA-1_2B_4         6.561031        10.274982 0.04499379 prep_batch_2</span></span>
<span><span class="co">#&gt; GGCTCGATCTGTGCAA-1_1B_1         8.196988         9.744375 0.02460745 prep_batch_1</span></span>
<span><span class="co">#&gt; CTGGTCTGTGGGTATG-1_2B_7         5.929589         9.260368 0.03326996 prep_batch_2</span></span>
<span><span class="co">#&gt; TCTGGAAGTACATGTC-1_2A_5         7.134094        10.002880 0.07319942 prep_batch_2</span></span>
<span><span class="co">#&gt; AACTGGTGTAGAGCTG-1_1A_4         5.655992         8.875567 0.01327186 prep_batch_1</span></span></code></pre></div>
<p><code>covariate_matrix</code> is a 40,000 x 4 data frame of
“technical factors,” or covariates. The row names of this data frame are
the cell barcodes. The covariates are as follows:</p>
<ul>
<li><p>Log-transformed gene library size
(<code>lg_gene_lib_size</code>); this vector can be computed via
<code>log(colSums(gene_matrix))</code></p></li>
<li><p>Log-transformed gRNA library size
(<code>lg_gRNA_lib_size</code>); this vector can be computed via
<code>log(colSums(gRNA_matrix))</code></p></li>
<li><p>Sequencing batch (<code>batch</code>)</p></li>
<li><p>Percentage of gene transcripts that map to mitochondrial genes
(<code>p_mito</code>)</p></li>
</ul>
<p>We strongly recommend that users include the same four covariates
(i.e., <code>lg_gene_lib_size</code>, <code>lg_gRNA_lib_size</code>,
<code>batch</code>, and <code>p_mito</code>) in their own cell-wise
covariate matrix.</p>
</div>
</div>
<div class="section level2">
<h2 id="step-2-assign-perturbation-identities-to-cells">Step 2: Assign perturbation identities to cells<a class="anchor" aria-label="anchor" href="#step-2-assign-perturbation-identities-to-cells"></a>
</h2>
<p>The second step is to impute perturbation assignments onto the cells.
For a given gRNA and cell, we label the cell as having been
<em>perturbed</em> by the gRNA if the UMI count of the gRNA within the
cell exceeds a user-specified, integer-valued threshold; otherwise, we
label the cell has having been <em>unperturbed</em> by the gRNA. The
default threshold that we use for this purpose is 3, which we have found
to work well in practice. This simple thresholding routine is
implemented by the function <code>threshold_gRNA_matrix</code>.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">perturbation_matrix</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/threshold_gRNA_matrix.html">threshold_gRNA_matrix</a></span><span class="op">(</span><span class="va">gRNA_matrix</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># `perturbation_matrix` is a thresholded version of `gRNA_matrix`</span></span>
<span><span class="va">perturbation_matrix</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">10</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">3</span><span class="op">]</span></span>
<span><span class="co">#&gt; 10 x 3 sparse Matrix of class "lgTMatrix"</span></span>
<span><span class="co">#&gt;                      CTCGAAAGTCCATGAT-1_2B_7 GACACGCTCAAAGACA-1_2B_4 GGCTCGATCTGTGCAA-1_1B_1</span></span>
<span><span class="co">#&gt; TGTCAGTCCTCCCTCCCCCA                       .                       .                       .</span></span>
<span><span class="co">#&gt; CTAAACAGTGAAAGTCACAG                       .                       .                       .</span></span>
<span><span class="co">#&gt; GCTCCAATCATATTCTAGAG                       .                       .                       .</span></span>
<span><span class="co">#&gt; GATGAAACCTAAGGGCACAC                       .                       .                       .</span></span>
<span><span class="co">#&gt; TGCAGACGAGTGTCTCAGAG                       .                       .                       .</span></span>
<span><span class="co">#&gt; TCATCTTGAAGTCAGCTCCA                       .                       .                       |</span></span>
<span><span class="co">#&gt; GAGAATGGCATGGAGCTCAA                       .                       .                       .</span></span>
<span><span class="co">#&gt; AGAGCAAGAGAGCAACTCCG                       .                       .                       .</span></span>
<span><span class="co">#&gt; CAGCAATTCTACCTTCAGGT                       .                       .                       .</span></span>
<span><span class="co">#&gt; TTTGGTTGTTGCAAATGAGG                       .                       .                       .</span></span></code></pre></div>
<p><strong>Note</strong>: We draw a terminological distinction between a
“gRNA count” and a “perturbation.” A “gRNA count” is the number of
transcripts (in UMIs) that we observe for a given gRNA in a given cell;
a “perturbation” is an assignment of a gRNA to a cell, obtained by
thresholding a gRNA count. A “perturbation matrix” is a binary matrix of
gRNA-to-cell assignments.</p>
<p><strong>Note</strong>: Step 2 is optional; users instead can pass the
raw gRNA count matrix to the <code>sceptre</code> function, in which
case the matrix is thresholded internally to produce a perturbation
matrix. (This option, though convenient, precludes the possibility of
<em>combining</em> gRNAs and thus is not recommended; see below.)</p>
<p><strong>Note</strong>: Users optionally can implement their own
strategy for converting the gRNA count matrix into a perturbation
matrix.</p>
</div>
<div class="section level2">
<h2 id="step-3-combine-perturbations">Step 3: Combine perturbations<a class="anchor" aria-label="anchor" href="#step-3-combine-perturbations"></a>
</h2>
<p>The third step is to combine gRNAs that target the same chromosomal
location. Information about the genomic position that each gRNA targets
is available in the <code>gRNA_groups_table</code>, which we explore
below.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">gRNA_groups_table_highmoi</span><span class="op">)</span></span>
<span><span class="va">gRNA_groups_table</span> <span class="op">&lt;-</span> <span class="va">gRNA_groups_table_highmoi</span></span>
<span><span class="va">gRNA_groups_table</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">2</span>, <span class="fl">31</span>, <span class="fl">32</span>, <span class="fl">41</span>, <span class="fl">42</span><span class="op">)</span>,<span class="op">]</span> </span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 6 × 3</span></span></span>
<span><span class="co">#&gt;   gRNA_id              gRNA_group gRNA_type </span></span>
<span><span class="co">#&gt;   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>                <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>      <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>     </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">1</span> TGTCAGTCCTCCCTCCCCCA chr10.2250 enh_target</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">2</span> CTAAACAGTGAAAGTCACAG chr10.2250 enh_target</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">3</span> AGAGGTAACCAAAATAGCAA NTC_2      non_target</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">4</span> ATATGTAACCTCCAGAATGA NTC_2      non_target</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">5</span> CAGGCTTTGCGGACGACGGG KRT18_TSS  tss_target</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">6</span> GCTTTGCGGACGACGGTGGG KRT18_TSS  tss_target</span></span></code></pre></div>
<p>The <code>gRNA_groups_table</code> data frame has three columns:</p>
<ul>
<li><p><code>gRNA_id</code>: the ID of an individual gRNA</p></li>
<li>
<p><code>gRNA_group</code>: the “group” to which a given gRNA
belongs. Gasperini et al. designed exactly two gRNAs to target each
putative enhancer or TSS; gRNAs that target the same putative enhancer
or TSS are paired to form groups of size two. Nontargeting gRNAs,
meanwhile, are paired randomly, also forming groups of size two.</p>
<p>Note that in general, a gRNA group can contain any nonzero number of
gRNAs.</p>
</li>
<li><p><code>target_type</code>: one of <code>enh_target</code>,
<code>tss_target</code>, and <code>non_target</code>, indicating whether
a given gRNA is enhancer-targeting, TSS-targeting, and
non-targeting.</p></li>
</ul>
<p>For example, the gRNAs “GCTCTTGGCTGGAGAATGCA” and
“GTTGCAGATGAGGCAACCGA” target the putative enhancer located at
<em>chr10.1318</em> and are grouped.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">gRNA_groups_table</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">2</span>,<span class="op">]</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 2 × 3</span></span></span>
<span><span class="co">#&gt;   gRNA_id              gRNA_group gRNA_type </span></span>
<span><span class="co">#&gt;   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>                <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>      <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>     </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">1</span> TGTCAGTCCTCCCTCCCCCA chr10.2250 enh_target</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">2</span> CTAAACAGTGAAAGTCACAG chr10.2250 enh_target</span></span></code></pre></div>
<p>Similarly, the gRNAs “CCCGCGCGCCGCACGGACGG” and
“GCGGATCGGGGCAAGGCTCG” target the TSS of gene <em>HDGF</em> and are
grouped.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">gRNA_groups_table</span><span class="op">[</span><span class="fl">41</span><span class="op">:</span><span class="fl">42</span>,<span class="op">]</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 2 × 3</span></span></span>
<span><span class="co">#&gt;   gRNA_id              gRNA_group gRNA_type </span></span>
<span><span class="co">#&gt;   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>                <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>      <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>     </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">1</span> CAGGCTTTGCGGACGACGGG KRT18_TSS  tss_target</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">2</span> GCTTTGCGGACGACGGTGGG KRT18_TSS  tss_target</span></span></code></pre></div>
<p>Finally, the non-targeting gRNAs “AATATTCTCCCTCATTCTGG” and
“TTAAAATTGATTCTGCCACT” are paired randomly to form a group called
“NTC_2.”</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">gRNA_groups_table</span><span class="op">[</span><span class="fl">31</span><span class="op">:</span><span class="fl">32</span>,<span class="op">]</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 2 × 3</span></span></span>
<span><span class="co">#&gt;   gRNA_id              gRNA_group gRNA_type </span></span>
<span><span class="co">#&gt;   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>                <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>      <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>     </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">1</span> AGAGGTAACCAAAATAGCAA NTC_2      non_target</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">2</span> ATATGTAACCTCCAGAATGA NTC_2      non_target</span></span></code></pre></div>
<p>The function <code>combine_perturbations</code> takes as arguments a
<code>perturbation_matrix</code> (i.e., the thresholded
<code>gRNA_matrix</code>) and a <code>gRNA_groups_table</code> and
collapses gRNAs in the same group into a single “combined” gRNA. The
data frame <code>gRNA_groups_table</code> must have columns
<code>gRNA_id</code> and <code>gRNA_group</code>, as above.</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">combined_perturbation_matrix</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/combine_perturbations.html">combine_perturbations</a></span><span class="op">(</span>perturbation_matrix <span class="op">=</span> <span class="va">perturbation_matrix</span>,</span>
<span>                                                      gRNA_groups_table <span class="op">=</span> <span class="va">gRNA_groups_table</span><span class="op">)</span></span></code></pre></div>
<p>The ouput matrix, <code>combined_perturbation_matrix</code>, is a 25
x 40,000 binary matrix, with rows corresponding to the 25 “gRNA groups”
within <code>gRNA_groups_table</code>. The column names (i.e., the cell
barcodes) remain unchanged.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/dim.html" class="external-link">dim</a></span><span class="op">(</span><span class="va">combined_perturbation_matrix</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1]    25 40000</span></span>
<span><span class="va">combined_perturbation_matrix</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">10</span>,<span class="fl">1</span><span class="op">:</span><span class="fl">3</span><span class="op">]</span></span>
<span><span class="co">#&gt; 10 x 3 sparse Matrix of class "lgCMatrix"</span></span>
<span><span class="co">#&gt;            CTCGAAAGTCCATGAT-1_2B_7 GACACGCTCAAAGACA-1_2B_4 GGCTCGATCTGTGCAA-1_1B_1</span></span>
<span><span class="co">#&gt; chr10.2250                       .                       .                       .</span></span>
<span><span class="co">#&gt; chr14.2077                       .                       .                       .</span></span>
<span><span class="co">#&gt; chr1.6211                        .                       .                       |</span></span>
<span><span class="co">#&gt; chr17.1748                       .                       .                       .</span></span>
<span><span class="co">#&gt; chr17.3995                       .                       .                       .</span></span>
<span><span class="co">#&gt; chr17.5367                       .                       .                       .</span></span>
<span><span class="co">#&gt; chr17.875                        .                       .                       .</span></span>
<span><span class="co">#&gt; chr18.248                        .                       .                       .</span></span>
<span><span class="co">#&gt; chr19.2680                       .                       .                       .</span></span>
<span><span class="co">#&gt; chr19.633                        .                       .                       .</span></span></code></pre></div>
<p><strong>Note</strong>: Step 3 is optional but recommended;
<code>sceptre</code> works on either combined or uncombined perturbation
matrices.</p>
<p><strong>Note</strong>: The <code>gRNA_type</code> column in
<code>gRNA_groups_table</code> is optional but useful for bookkeeping
purposes.</p>
</div>
<div class="section level2">
<h2 id="step-4-determine-which-pairs-to-analyze">Step 4: Determine which pairs to analyze<a class="anchor" aria-label="anchor" href="#step-4-determine-which-pairs-to-analyze"></a>
</h2>
<p>Step 4 is to determine the pairs of genes and gRNA groups to analyze.
It is common to analyze pairs of genes and gRNA groups that are in close
physical proximity to uncover <em>cis</em>-regulatory relationships. The
data frame <code>gene_gRNA_group_pairs</code> contains the 120 pairs of
genes and gRNA groups that we will analyze in our analysis.</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">gene_gRNA_group_pairs_highmoi</span><span class="op">)</span></span>
<span><span class="va">gene_gRNA_group_pairs</span> <span class="op">&lt;-</span> <span class="va">gene_gRNA_group_pairs_highmoi</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">gene_gRNA_group_pairs</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] 120</span></span>
<span><span class="va">gene_gRNA_group_pairs</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fl">2</span>, <span class="fl">6</span><span class="op">:</span><span class="fl">7</span>, <span class="fl">21</span><span class="op">:</span><span class="fl">22</span><span class="op">)</span>,<span class="op">]</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 6 × 3</span></span></span>
<span><span class="co">#&gt;   gene_id         gRNA_group pair_type       </span></span>
<span><span class="co">#&gt;   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>           <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>      <span style="color: #949494; font-style: italic;">&lt;fct&gt;</span>           </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">1</span> ENSG00000111057 KRT18_TSS  positive_control</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">2</span> ENSG00000121211 MND1_TSS   positive_control</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">3</span> ENSG00000105298 chr19.633  candidate       </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">4</span> ENSG00000108389 chr17.3995 candidate       </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">5</span> ENSG00000111057 NTC_2      negative_control</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">6</span> ENSG00000121211 NTC_2      negative_control</span></span></code></pre></div>
<p>The <code>gene_gRNA_group_pairs</code> data frame contains three
columns: <code>gene_id</code>, <code>gRNA_group</code>, and
<code>pair_type</code>. The gene IDs (respectively, gRNA groups) in
<code>gene_gRNA_group_pairs</code> must be a subset of the gene IDs
(respectively, gRNA groups) in <code>gene_matrix</code> (respectively,
<code>combined_perturbation_matrix</code>). The <code>pair_type</code>
column gives the “type” of each pair, one of
<code>positive_control</code> (for TSS-targeting gRNA groups paired to
their target gene), <code>candidate</code> (for putative
enhancer-targeting gRNA groups paired to a nearby gene), and
<code>negative_control</code> (for pairs of genes and gRNA groups that
consist of a non-targeting gRNA group).</p>
<p>The negative control pairs were constructed by pairing each
non-targeting gRNA group to the entire set of genes, as is standard.
When a large number of genes and non-targeting gRNA groups is available,
it is common to analyze a random subset of the non-targeting gRNA
group-gene pairs to reduce computational cost. We recommend analyzing at
least as many negative control pairs as there are candidate pairs.</p>
<p><strong>Note</strong>: The <code>pair_type</code> column, while
helpful for bookkeeping purposes, is optional.</p>
<p><strong>Note</strong>: Negative control and positive control pairs
are <em>not</em> required to run <code>sceptre</code>; however, negative
and positive control pairs are extremely useful for checking the
calibration and power of <code>sceptre</code> (see Step 7, below) and
therefore are strongly recommended.</p>
</div>
<div class="section level2">
<h2 id="step-5-determine-the-sidedness-of-the-test">Step 5: Determine the sidedness of the test<a class="anchor" aria-label="anchor" href="#step-5-determine-the-sidedness-of-the-test"></a>
</h2>
<p>The fourth step is to determine the sideness of the statistical test.
If we are testing for an <em>increase</em> (respectively,
<em>decrease</em>) in gene expression in response to the perturbation,
then we should use a <em>right</em>-sided (respectively,
<em>left</em>-sided) test. On the other hand, if we are testing for an
increase <em>or</em> decrease in gene expression, then we should use a
<em>two</em>-sided test.</p>
<p>Whether we seek to test for an increase or decrease in gene
expression (or either and increase <em>or</em> decrease) depends on (i)
the genomic element being targeted, (ii) the CRISPR modality being used,
and (iii) whether we are testing an association in <em>cis</em> or in
<em>trans</em>. For example, if we perturb a putative enhancer via
CRISPRi and test for an association between that enhancer and a gene in
<em>cis</em>, then we should should test for a <em>decrease</em> in
expression and therefore use a <em>left</em>-sided test. The following
table summarizes whether a left-, right-, or two-tailed tailed test is
appropriate as a function of the aforementioned variables.</p>
<table class="table">
<colgroup>
<col width="20%">
<col width="20%">
<col width="20%">
<col width="20%">
<col width="20%">
</colgroup>
<thead><tr class="header">
<th align="center">Target element</th>
<th align="center">CRISPR modality</th>
<th align="center">
<em>Cis</em> or <em>trans</em>
</th>
<th align="center">Testing for…</th>
<th align="center">Sidedness</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="center">Enhancer or TSS</td>
<td align="center">CRISPRi or CRISPRko</td>
<td align="center"><em>Cis</em></td>
<td align="center">Decrease in expression</td>
<td align="center">Left</td>
</tr>
<tr class="even">
<td align="center">Silencer</td>
<td align="center">CRISPRi or CRISPRko</td>
<td align="center"><em>Cis</em></td>
<td align="center">Increase in expression</td>
<td align="center">Right</td>
</tr>
<tr class="odd">
<td align="center">Enhancer or TSS</td>
<td align="center">CRISPRa</td>
<td align="center"><em>Cis</em></td>
<td align="center">Increase in expression</td>
<td align="center">Right</td>
</tr>
<tr class="even">
<td align="center">Silencer</td>
<td align="center">CRISPRa</td>
<td align="center"><em>Cis</em></td>
<td align="center">Decrease in expression</td>
<td align="center">Left</td>
</tr>
<tr class="odd">
<td align="center">Any</td>
<td align="center">Any</td>
<td align="center"><em>Trans</em></td>
<td align="center">Increase <em>or</em> Decrease</td>
<td align="center">Both</td>
</tr>
</tbody>
</table>
<p>The direction of <em>trans</em> relationships is less certain than
that of <em>cis</em> relationships, as <em>trans</em> relationships
typically are mediated by one or more transcription factors. We
therefore recommend using a two-sided test for all tests of
<em>trans</em> relationships.</p>
<p>We use a left-tailed test for the example data, as suggested by the
first row of the table.</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">side</span> <span class="op">&lt;-</span> <span class="st">"left"</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="step-6-run-the-method">Step 6: Run the method<a class="anchor" aria-label="anchor" href="#step-6-run-the-method"></a>
</h2>
<p>The sixth step is to call the function
<code>run_sceptre_highmoi</code> on the data. The most important
arguments to this function are <code>gene_matrix</code>,
<code>combined_perturbation_matrix</code>,
<code>covariate_matrix</code>, <code>gene_gRNA_group_pairs</code>, and
<code>side</code>, all of which we prepared above.
<code>run_sceptre_highmoi</code> has several additional, optional
arguments, which are set to reasonable defaults. One can read more about
<code>run_sceptre_highmoi</code> by checking the documentation
(<code><a href="../reference/run_sceptre_highmoi.html">?run_sceptre_highmoi</a></code>). The function takes about 40 second
to run on the example data on an 8-core Macbook Pro.</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span> <span class="va">result</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/run_sceptre_highmoi.html">run_sceptre_highmoi</a></span><span class="op">(</span>gene_matrix <span class="op">=</span> <span class="va">gene_matrix</span>,</span>
<span>                               combined_perturbation_matrix <span class="op">=</span> <span class="va">combined_perturbation_matrix</span>,</span>
<span>                               covariate_matrix <span class="op">=</span> <span class="va">covariate_matrix</span>,</span>
<span>                               gene_gRNA_group_pairs <span class="op">=</span> <span class="va">gene_gRNA_group_pairs</span>,</span>
<span>                               side <span class="op">=</span> <span class="va">side</span><span class="op">)</span></span></code></pre></div>
<p><strong>Note</strong>: Users can supply the raw gRNA expression
matrix they obtained from Step 1 or the perturbation matrix they
obtained from Step 2 to the <code>combined_perturbation_matrix</code>
argument, in which case <code>sceptre</code> treats each gRNA as its own
group of one. We recommend against this usage unless the data consist of
only a single gRNA per target.</p>
</div>
<div class="section level2">
<h2 id="step-7--analyze-the-results">Step 7. Analyze the results<a class="anchor" aria-label="anchor" href="#step-7--analyze-the-results"></a>
</h2>
<p>The output of <code>run_sceptre_highmoi</code> is a data frame called
<code>result</code>, which is simply <code>gene_gRNA_pairs</code> with
two additional columns: <code>p_value</code>, and <code>z_value</code>.
<code>p_value</code> is the SCEPTRE <em>p</em>-value for a given pair,
and <code>z_value</code> is the z-value of a negative binomial GLM
fitted to the data for a given pair. Positive z-values indicate
increased expression, and negative z-values indicate decreased
expression. We examine several rows of <code>result</code> below.</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">result</span>, <span class="fl">10</span><span class="op">)</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 10 × 6</span></span></span>
<span><span class="co">#&gt;    gene_id         gRNA_id    pair_type         p_value z_value log_fold_change</span></span>
<span><span class="co">#&gt;    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>           <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>      <span style="color: #949494; font-style: italic;">&lt;fct&gt;</span>               <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>           <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 1</span> ENSG00000111057 KRT18_TSS  positive_control 2.22<span style="color: #949494;">e</span><span style="color: #BB0000;">-16</span> -<span style="color: #BB0000;">12.5</span>           -<span style="color: #BB0000;">1.21</span>  </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 2</span> ENSG00000121211 MND1_TSS   positive_control 2.22<span style="color: #949494;">e</span><span style="color: #BB0000;">-16</span>  -<span style="color: #BB0000;">9.74</span>          -<span style="color: #BB0000;">1.26</span>  </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 3</span> ENSG00000134809 TIMM10_TSS positive_control 2.22<span style="color: #949494;">e</span><span style="color: #BB0000;">-16</span> -<span style="color: #BB0000;">14.0</span>           -<span style="color: #BB0000;">0.980</span> </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 4</span> ENSG00000166902 MRPL16_TSS positive_control 2.22<span style="color: #949494;">e</span><span style="color: #BB0000;">-16</span>  -<span style="color: #BB0000;">9.80</span>          -<span style="color: #BB0000;">1.15</span>  </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 5</span> ENSG00000172809 RPL38_TSS  positive_control 2.23<span style="color: #949494;">e</span><span style="color: #BB0000;">-13</span> -<span style="color: #BB0000;">14.3</span>           -<span style="color: #BB0000;">0.641</span> </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 6</span> ENSG00000105298 chr19.633  candidate        4.29<span style="color: #949494;">e</span><span style="color: #BB0000;">- 1</span>  -<span style="color: #BB0000;">0.217</span>         -<span style="color: #BB0000;">0.045</span><span style="color: #BB0000; text-decoration: underline;">2</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 7</span> ENSG00000108389 chr17.3995 candidate        8.36<span style="color: #949494;">e</span><span style="color: #BB0000;">- 1</span>   0.936          0.131 </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 8</span> ENSG00000157985 chr2.7132  candidate        1.62<span style="color: #949494;">e</span><span style="color: #BB0000;">- 1</span>  -<span style="color: #BB0000;">0.984</span>         -<span style="color: #BB0000;">0.218</span> </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 9</span> ENSG00000107731 chr10.2250 candidate        4.39<span style="color: #949494;">e</span><span style="color: #BB0000;">- 2</span>  -<span style="color: #BB0000;">1.27</span>        -<span style="color: #BB0000;">Inf</span>     </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">10</span> ENSG00000184357 chr6.1291  candidate        8.49<span style="color: #949494;">e</span><span style="color: #BB0000;">- 1</span>   0.797          0.036<span style="text-decoration: underline;">4</span></span></span></code></pre></div>
<div class="section level3">
<h3 id="negative-control-pairs">Negative control pairs<a class="anchor" aria-label="anchor" href="#negative-control-pairs"></a>
</h3>
<p>We must verify that the test is correctly calibrated. To this end we
create a QQ-plot of the <em>p</em>-values corresponding to the negative
control pairs. We extract the negative control <em>p</em>-values from
the <code>result</code> data frame and pass them to the function
<code>make_qq_plot</code>.</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">neg_control_p_vals</span> <span class="op">&lt;-</span> <span class="va">result</span> <span class="op">|&gt;</span> <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">pair_type</span> <span class="op">==</span> <span class="st">"negative_control"</span><span class="op">)</span> <span class="op">|&gt;</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/pull.html" class="external-link">pull</a></span><span class="op">(</span><span class="va">p_value</span><span class="op">)</span></span>
<span><span class="va">qq_plot</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/make_qq_plot.html">make_qq_plot</a></span><span class="op">(</span><span class="va">neg_control_p_vals</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">qq_plot</span><span class="op">)</span></span></code></pre></div>
<p><img src="highmoi_tutorial_files/figure-html/unnamed-chunk-19-1.png" width="480"></p>
<p>The QQ-plot is a visual diagnostic that displays the extent to which
the negative control <em>p</em>-values deviate from the expected uniform
distribution. The points, each of which represents a <em>p</em>-value,
should lie roughly along the diagonal line. If the points do not lie
along the diagonal, then the test is poorly calibrated and the results
are unreliable. The gray region is a 95% pointwise confidence band.</p>
<p>A reasonable rule of thumb for interpreting the QQ-plot is as
follows: if all (or nearly all) of the points with an expected null
<em>p-</em>value of 0.1 or less fall within the confidence band, then we
can be confident that the test is correctly calibrated. In other words,
if nearly all the points to the <em>right</em> of 0.1 (dashed red line)
on the <em>x</em>-axis fall within the gray region, then we have good
reason to trust the results. This rule of thumb is satisfied on the
example data, as seen above.</p>
<p>A forthcoming article will give recommendations for what to do if the
QQ-plot reveals that the test is miscalibrated. We do not expect this to
happen commonly.</p>
</div>
<div class="section level3">
<h3 id="positive-control-pairs">Positive control pairs<a class="anchor" aria-label="anchor" href="#positive-control-pairs"></a>
</h3>
<p>Next, to verify that the test is powerful (i.e., that the test is
capable of detecting true relationships), we assess the power of the
test on the positive control pairs. We extract the positive control
<em>p</em>-values and examine their values.</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pos_control_p_vals</span> <span class="op">&lt;-</span> <span class="va">result</span> <span class="op">|&gt;</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">pair_type</span> <span class="op">==</span> <span class="st">"positive_control"</span><span class="op">)</span> <span class="op">|&gt;</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/pull.html" class="external-link">pull</a></span><span class="op">(</span><span class="va">p_value</span><span class="op">)</span></span>
<span><span class="va">pos_control_p_vals</span></span>
<span><span class="co">#&gt; [1] 2.220446e-16 2.220446e-16 2.220446e-16 2.220446e-16 2.232103e-13</span></span></code></pre></div>
<p>The positive control <em>p</em>-values are all small, indicating that
the test is powerful.</p>
</div>
<div class="section level3">
<h3 id="candidate-pairs">Candidate pairs<a class="anchor" aria-label="anchor" href="#candidate-pairs"></a>
</h3>
<p>Finally, we produce the discovery set. We extract the
<em>p</em>-values corresponding to the candidate pairs and apply a
Benjamini-Hochberg (BH) correction to adjust for multiple testing.</p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">candidate_pair_results</span> <span class="op">&lt;-</span> <span class="va">result</span> <span class="op">|&gt;</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">pair_type</span> <span class="op">==</span> <span class="st">"candidate"</span><span class="op">)</span></span>
<span><span class="va">candidate_pair_results_p_adj</span> <span class="op">&lt;-</span> <span class="va">candidate_pair_results</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span>p_val_adj <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/stats/p.adjust.html" class="external-link">p.adjust</a></span><span class="op">(</span><span class="va">p_value</span>, method <span class="op">=</span> <span class="st">"BH"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">candidate_pair_results_p_adj</span><span class="op">)</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 6 × 7</span></span></span>
<span><span class="co">#&gt;   gene_id         gRNA_id    pair_type p_value z_value log_fold_change p_val_adj</span></span>
<span><span class="co">#&gt;   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>           <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>      <span style="color: #949494; font-style: italic;">&lt;fct&gt;</span>       <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>           <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>     <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">1</span> ENSG00000105298 chr19.633  candidate  0.429   -<span style="color: #BB0000;">0.217</span>         -<span style="color: #BB0000;">0.045</span><span style="color: #BB0000; text-decoration: underline;">2</span>     0.737</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">2</span> ENSG00000108389 chr17.3995 candidate  0.836    0.936          0.131      0.906</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">3</span> ENSG00000157985 chr2.7132  candidate  0.162   -<span style="color: #BB0000;">0.984</span>         -<span style="color: #BB0000;">0.218</span>      0.537</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">4</span> ENSG00000107731 chr10.2250 candidate  0.043<span style="text-decoration: underline;">9</span>  -<span style="color: #BB0000;">1.27</span>        -<span style="color: #BB0000;">Inf</span>          0.431</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">5</span> ENSG00000184357 chr6.1291  candidate  0.849    0.797          0.036<span style="text-decoration: underline;">4</span>     0.906</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">6</span> ENSG00000090060 chr14.2077 candidate  0.179   -<span style="color: #BB0000;">1.02</span>          -<span style="color: #BB0000;">0.048</span><span style="color: #BB0000; text-decoration: underline;">7</span>     0.537</span></span></code></pre></div>
<p>We call pairs with an adjusted <em>p</em>-value of less or equal than
0.1 significant; the discovery set (i.e., the set of significant pairs)
has a false discovery rate (FDR) of 10%.</p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">discovery_set</span> <span class="op">&lt;-</span> <span class="va">candidate_pair_results_p_adj</span> <span class="op">|&gt;</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">p_val_adj</span> <span class="op">&lt;=</span> <span class="fl">0.1</span><span class="op">)</span></span></code></pre></div>
<p>The discovery set in this case is empty, as we tested only fifteen
candidate pairs.</p>
<p><strong>Note</strong>: It is crucial to apply BH <em>only</em> to the
pairs in the discovery set (i.e., the candidate pairs). Including
positive control or negative control pairs in the discovery set causes
BH to become excessively liberal or conservative, respectively.</p>
</div>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>Developed by <a href="https://timothy-barry.github.io/" class="external-link">Timothy Barry</a>, <a href="https://ekatsevi.github.io/" class="external-link">Eugene Katsevich</a>.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.5.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
