<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Low MOI sceptre tutorial • sceptre</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/lumen/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Low MOI sceptre tutorial">
<meta property="og:description" content="sceptre">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">sceptre</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.5.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/highmoi_tutorial.html">High MOI sceptre tutorial</a>
    </li>
    <li>
      <a href="../articles/lowmoi_tutorial.html">Low MOI sceptre tutorial</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Low MOI sceptre tutorial</h1>
                        <h4 data-toc-skip class="author">Tim Barry</h4>
            
      
      
      <div class="hidden name"><code>lowmoi_tutorial.Rmd</code></div>

    </div>

    
    
<p>This tutorial illustrates application of <code>sceptre</code> to an
example low multiplicity-of-infection (MOI) single-cell CRISPR screen
dataset. The example dataset that we examine is taken from the paper
“Characterizing the molecular regulation of inhibitory immune
checkpoints with multimodal single-cell screens” by Papalexi et al.,
2021. The authors of this study perturbed 26 genes hypothesized to play
a role in the expression of immune checkpoint molecules and measured the
effects of these perturbations via single-cell RNA sequencing.</p>
<div class="section level2">
<h2 id="initial-code-demonstration">Initial code demonstration<a class="anchor" aria-label="anchor" href="#initial-code-demonstration"></a>
</h2>
<p>The code below is a self-contained and complete application of
<code>sceptre</code> to the example data. Users can copy, paste, and
execute this code in Rstudio to get a sense for how <code>sceptre</code>
works. The entire script takes less than two minutes to run on a
standard laptop.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># load package</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">sceptre</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># load the data associated with the experiment</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">response_matrix_lowmoi</span><span class="op">)</span> <span class="co"># gene-by-cell expression matrix</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">grna_matrix_lowmoi</span><span class="op">)</span> <span class="co"># gRNA-by-cell expression matrix</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">covariate_data_frame_lowmoi</span><span class="op">)</span> <span class="co"># cell-by-covariate data frame</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">grna_group_data_frame_lowmoi</span><span class="op">)</span> <span class="co"># gRNA group information</span></span>
<span></span>
<span><span class="co"># obtain the set of pairs to analyze</span></span>
<span><span class="va">response_grna_group_pairs</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/generate_all_pairs.html">generate_all_pairs</a></span><span class="op">(</span><span class="va">response_matrix_lowmoi</span>,</span>
<span>                                                <span class="va">grna_group_data_frame_lowmoi</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># set the formula object</span></span>
<span><span class="va">formula_object</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/formula.html" class="external-link">formula</a></span><span class="op">(</span><span class="op">~</span><span class="fu"><a href="https://rdrr.io/r/base/Log.html" class="external-link">log</a></span><span class="op">(</span><span class="va">response_n_umis</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://rdrr.io/r/base/Log.html" class="external-link">log</a></span><span class="op">(</span><span class="va">response_n_nonzero</span><span class="op">)</span> <span class="op">+</span></span>
<span>                            <span class="va">bio_rep</span> <span class="op">+</span> <span class="va">p_mito</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># run the calibration check analysis (NOTE: `calibration_check` set to `TRUE`)</span></span>
<span><span class="va">calibration_result</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/run_sceptre_lowmoi.html">run_sceptre_lowmoi</a></span><span class="op">(</span>response_matrix <span class="op">=</span> <span class="va">response_matrix_lowmoi</span>,</span>
<span>grna_matrix <span class="op">=</span> <span class="va">grna_matrix_lowmoi</span>,</span>
<span>covariate_data_frame <span class="op">=</span> <span class="va">covariate_data_frame_lowmoi</span>,</span>
<span>grna_group_data_frame <span class="op">=</span> <span class="va">grna_group_data_frame_lowmoi</span>,</span>
<span>formula_object <span class="op">=</span> <span class="va">formula_object</span>,</span>
<span>response_grna_group_pairs <span class="op">=</span> <span class="va">response_grna_group_pairs</span>,</span>
<span>calibration_check <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># plot the calibration result to ensure adequate calibration of the p-values</span></span>
<span><span class="fu"><a href="../reference/plot_calibration_result.html">plot_calibration_result</a></span><span class="op">(</span><span class="va">calibration_result</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># run the discovery analysis (NOTE: `calibration_check` set to `FALSE`)</span></span>
<span><span class="va">discovery_result</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/run_sceptre_lowmoi.html">run_sceptre_lowmoi</a></span><span class="op">(</span>response_matrix <span class="op">=</span> <span class="va">response_matrix_lowmoi</span>,</span>
<span>grna_matrix <span class="op">=</span> <span class="va">grna_matrix_lowmoi</span>,</span>
<span>covariate_data_frame <span class="op">=</span> <span class="va">covariate_data_frame_lowmoi</span>,</span>
<span>grna_group_data_frame <span class="op">=</span> <span class="va">grna_group_data_frame_lowmoi</span>,</span>
<span>formula_object <span class="op">=</span> <span class="va">formula_object</span>,</span>
<span>response_grna_group_pairs <span class="op">=</span> <span class="va">response_grna_group_pairs</span>,</span>
<span>calibration_check <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># compare discovery p-values to the negative control p-values; make a volcano plot</span></span>
<span><span class="fu"><a href="../reference/compare_calibration_and_discovery_results.html">compare_calibration_and_discovery_results</a></span><span class="op">(</span><span class="va">calibration_result</span>, <span class="va">discovery_result</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/make_volcano_plot.html">make_volcano_plot</a></span><span class="op">(</span><span class="va">discovery_result</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># obtain the discovery set for downstream analysis</span></span>
<span><span class="va">discovery_set</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/obtain_discovery_set.html">obtain_discovery_set</a></span><span class="op">(</span><span class="va">discovery_result</span><span class="op">)</span></span></code></pre></div>
<p>All <code>sceptre</code> analyses follow this general template.
First, we load the data. Next, we construct the set of CRISPR
perturbation-gene pairs that we seek to test for association. We then
define a formula object, which specifies how <code>sceptre</code> is to
adjust for the cell-specific covariates (e.g., sequencing depth, batch,
etc.). Subsequently, we carry out the “calibration check,” which
involves applying <code>sceptre</code> to a set of
automatically-generated negative control CRISPR perturbation-gene pairs
(i.e., perturbation-gene pairs for which we know there is no regulatory
relationship). The calibration check enables us to verify that
<code>sceptre</code> controls the number of false discoveries on the
dataset under analysis, which ultimately ensures that the discovery set
that <code>sceptre</code> returns is high-quality. We then conduct the
“discovery analysis,” which entails applying <code>sceptre</code> to the
set of perturbation-gene pairs whose regulatory relationships we seek to
learn. Finally, we create a couple plots to visualize the results and
obtain the discovery set. Below, we describe each step of the analysis
procedure in greater detail.</p>
</div>
<div class="section level2">
<h2 id="part-i-setup">Part I: Setup<a class="anchor" aria-label="anchor" href="#part-i-setup"></a>
</h2>
<p>We begin by setting up the analysis. Setup consists of four steps:
preparing the data to pass to <code>sceptre</code>, determining the set
of CRISPR perturbation-gene pairs to test for association, defining a
formula object that specifies how to adjust for the covariates, and
selecting a pairwise quality control (QC) threshold.</p>
<div class="section level3">
<h3 id="step-1-prepare-the-data-objects">Step 1: Prepare the data objects<a class="anchor" aria-label="anchor" href="#step-1-prepare-the-data-objects"></a>
</h3>
<p>A single-cell CRISPR screen analysis involves four fundamental data
objects: (i) the response-by-cell expression matrix, (ii) the
gRNA-by-cell expression matrix, (iii) the cell covariate matrix, and
(iv) the gRNA group information table. (We use the term “response” to
refer to the outcome variable in the single-cell CRISPR screen
experiment; typically, the responses are genes or cell surface
proteins.) We describe each of these objects, using the Papalexi data as
a running example.</p>
<div class="figure" style="text-align: center">
<img src="data_objects.png" alt="The four fundamental data objects in a single-cell CRISPR screen analysis." width="500px"><p class="caption">
The four fundamental data objects in a single-cell CRISPR screen
analysis.
</p>
</div>
<p>i. The <strong>response-by-cell matrix</strong> is a matrix that
records the expression of each response in each cell. The example
response-by-cell matrix included in the <code>sceptre</code> package is
stored in the variable <code>response_matrix_lowmoi</code>.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">response_matrix_lowmoi</span><span class="op">)</span></span></code></pre></div>
<p><code>response_matrix_lowmoi</code> is an expression matrix
consisting of 290 genes and 20,729 cells. This matrix was constructed by
sampling 290 genes from the original expression matrix. The responses
(i.e., the genes) are in the rows, and the cells are in the columns. We
print a few entries of <code>response_matrix_lowmoi</code> to get a
sense for what these data look like.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">response_matrix_lowmoi</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">5</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">5</span><span class="op">]</span></span>
<span><span class="co">#&gt; 5 x 5 sparse Matrix of class "dgTMatrix"</span></span>
<span><span class="co">#&gt;                   </span></span>
<span><span class="co">#&gt; PCBP3    . . . . .</span></span>
<span><span class="co">#&gt; ITGB1BP1 3 2 . 1 3</span></span>
<span><span class="co">#&gt; SERTAD1  3 . 1 . 1</span></span>
<span><span class="co">#&gt; CHEK2    . . . . .</span></span>
<span><span class="co">#&gt; DUSP15   . . . . .</span></span></code></pre></div>
<p>The gene IDs, which uniquely identify each gene, are stored as row
names.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">response_matrix_lowmoi</span><span class="op">)</span> <span class="op">|&gt;</span> <span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "PCBP3"    "ITGB1BP1" "SERTAD1"  "CHEK2"    "DUSP15"   "IFT46"</span></span></code></pre></div>
<p><code>response_matrix_lowmoi</code> is a sparse R matrix. In general
the response-by-cell matrix should be a standard (i.e., dense) R matrix
or a sparse R matrix (of type <code>dgCMatrix</code>,
<code>dgTMatrix</code>, or <code>dgRMatrix</code>). Note that the
response-by-cell matrix is an <em>un-normalized</em> (i.e., raw) matrix
of counts.</p>
<p>ii. The <strong>gRNA-by-cell matrix</strong> records the expression
of each gRNA in each cell. The example gRNA-by-cell matrix is stored in
the variable <code>grna_matrix_lowmoi</code>.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">grna_matrix_lowmoi</span><span class="op">)</span></span></code></pre></div>
<p><code>grna_matrix_lowmoi</code> consists of 110 gRNAs and 20,729
cells. The gRNAs are stored in the rows and the cells in the columns. We
print a few entries of this matrix to get a sense for what it looks
like.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">grna_matrix_lowmoi</span><span class="op">[</span><span class="fl">25</span><span class="op">:</span><span class="fl">30</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">5</span><span class="op">]</span></span>
<span><span class="co">#&gt; 6 x 5 sparse Matrix of class "dgTMatrix"</span></span>
<span><span class="co">#&gt;                    </span></span>
<span><span class="co">#&gt; CAV1g2 .  . .   . .</span></span>
<span><span class="co">#&gt; CAV1g3 .  . .   . .</span></span>
<span><span class="co">#&gt; CAV1g4 . 66 .   . .</span></span>
<span><span class="co">#&gt; CD86g1 .  . . 119 .</span></span>
<span><span class="co">#&gt; CD86g2 .  . .   . .</span></span>
<span><span class="co">#&gt; CD86g3 .  . .   . .</span></span></code></pre></div>
<p>The gRNA IDs, which uniquely identify each individual gRNA, are
stored as row names.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">grna_matrix_lowmoi</span><span class="op">)</span> <span class="op">|&gt;</span> <span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "CUL3g1"  "CUL3g2"  "CUL3g3"  "CMTM6g1" "CMTM6g2" "CMTM6g3"</span></span></code></pre></div>
<p>Like the gene expression matrix, the gRNA-by-cell matrix either
should be a standard R matrix or a sparse R matrix. Users optionally can
pass a logical (i.e., <code>TRUE</code>/<code>FALSE</code>) matrix of
cell-to-gRNA assignments; see the documentation of
<code>run_sceptre_lowmoi</code> via (<code><a href="../reference/run_sceptre_lowmoi.html">?run_sceptre_lowmoi</a></code>)
for more details.</p>
<p>iii. The <strong>cell covariate matrix</strong> records the value of
each cell-specific covariate (e.g., sequencing batch, sequencing depth)
in each cell. The example cell covariate matrix is stored in the
variable <code>covariate_data_frame_lowmoi</code>.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">covariate_data_frame_lowmoi</span><span class="op">)</span></span></code></pre></div>
<p>The cells are in the rows and the covariates in the columns. We print
the first few rows of this data frame below.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">covariate_data_frame_lowmoi</span><span class="op">)</span></span>
<span><span class="co">#&gt;   response_n_umis response_n_nonzero bio_rep     p_mito</span></span>
<span><span class="co">#&gt; 1           17207               3942   rep_1 0.02295577</span></span>
<span><span class="co">#&gt; 2            9506               2948   rep_1 0.04512939</span></span>
<span><span class="co">#&gt; 3           15256               4258   rep_1 0.04116413</span></span>
<span><span class="co">#&gt; 4            5135               1780   rep_1 0.05491723</span></span>
<span><span class="co">#&gt; 5            9673               2671   rep_1 0.03359868</span></span>
<span><span class="co">#&gt; 6           14941               3918   rep_1 0.03379961</span></span></code></pre></div>
<p>There are four covariates: <code>response_n_umis</code> (i.e., the
total number of gene UMIs sequenced in the cell),
<code>response_n_nonzero</code> (i.e., the total number of genes
expressed in the cell), <code>bio_rep</code> (i.e., the biological
replicate in which the cell was sequenced, one of <code>rep_1</code>,
<code>rep_2</code>, and <code>rep_3</code>) and <code>p_mito</code>
(i.e., the proportion of sequenced gene transcripts that map to
mitochondrial genes). We strongly recommend that users include the
variables <code>response_n_umis</code>, <code>response_n_nonzero</code>,
and <code>p_mito</code> in their cell covariate matrix. When applicable,
sequencing batch and/or biological replicate also should be included.
Categorical variables (e.g., <code>bio_rep</code>) should be stored as
factors or strings rather than integers.</p>
<details><summary><strong>How to compute the cell covariate matrix within
<code>sceptre</code></strong>
</summary><p>Users can compute the variables <code>response_n_umis</code>,
<code>response_n_nonzero</code>, and <code>p_mito</code> for an input
matrix using the <code>sceptre</code> function
<code>compute_cell_covariates</code>. An demonstration of this function
is below:</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">covariate_data_frame_reduced</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/compute_cell_covariates.html">compute_cell_covariates</a></span><span class="op">(</span><span class="va">response_matrix_lowmoi</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">covariate_data_frame_reduced</span><span class="op">)</span></span>
<span><span class="co">#&gt;   n_nonzero n_umis    p_mito</span></span>
<span><span class="co">#&gt; 1        72    339 0.4631268</span></span>
<span><span class="co">#&gt; 2        65    328 0.5731707</span></span>
<span><span class="co">#&gt; 3        85    520 0.6115385</span></span>
<span><span class="co">#&gt; 4        36    195 0.7589744</span></span>
<span><span class="co">#&gt; 5        58    242 0.5495868</span></span>
<span><span class="co">#&gt; 6        70    466 0.5128755</span></span></code></pre></div>
<p>Users manually can append batch and/or biological replicate
information via the dollar sign (<code>$</code>) operator.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">covariate_data_frame_reduced</span><span class="op">$</span><span class="va">bio_rep</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="st">"rep_1"</span>, <span class="fl">9428</span><span class="op">)</span>,</span>
<span>                                          <span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="st">"rep_2"</span>, <span class="fl">6001</span><span class="op">)</span>,</span>
<span>                                          <span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="st">"rep_3"</span>, <span class="fl">5300</span><span class="op">)</span><span class="op">)</span> <span class="op">|&gt;</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">factor</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
We will use <code>covariate_data_frame_lowmoi</code> instead of the
above-computed <code>covariate_data_frame_reduced</code> throughout the
rest of this tutorial, as the former was computed using the entire set
of genes, while the latter was computed using the reduced set of 290
genes.
</details><p>iv. The <strong>gRNA group table</strong> records the “gRNA group” to
which each individual gRNA belongs. Typically, gRNAs that target the
same genomic location are assigned to the same gRNA group. The example
gRNA group table included in the <code>sceptre</code> package is stored
in the variable <code>grna_group_data_frame_lowmoi</code>.</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">grna_group_data_frame_lowmoi</span><span class="op">)</span></span></code></pre></div>
<p><code>grna_group_data_frame_lowmoi</code> is a data frame containing
two columns: <code>grna_id</code> and <code>grna_group</code>. In the
example data, gRNAs are grouped according the gene that they target. For
example, gRNAs <code>ATF2g1</code>, <code>ATF2g2</code>,
<code>ATF2g3</code>, and <code>ATF2g4</code> target gene
<code>ATF2</code>.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">grna_group_data_frame_lowmoi</span>, <span class="fl">8</span><span class="op">)</span></span>
<span><span class="co">#&gt;   grna_id grna_group</span></span>
<span><span class="co">#&gt; 1  ATF2g1       ATF2</span></span>
<span><span class="co">#&gt; 2  ATF2g2       ATF2</span></span>
<span><span class="co">#&gt; 3  ATF2g3       ATF2</span></span>
<span><span class="co">#&gt; 4  ATF2g4       ATF2</span></span>
<span><span class="co">#&gt; 5  BRD4g1       BRD4</span></span>
<span><span class="co">#&gt; 6  BRD4g2       BRD4</span></span>
<span><span class="co">#&gt; 7  BRD4g3       BRD4</span></span>
<span><span class="co">#&gt; 8  BRD4g4       BRD4</span></span></code></pre></div>
<p>Importantly, all non-targeting (NT) gRNAs (in this case,
<code>NTg1</code>-<code>NTg10</code>) are assigned a
<code>grna_group</code> label of “non-targeting.”</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span> <span class="va">grna_group_data_frame_lowmoi</span> <span class="op">|&gt;</span> <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">grna_group</span> <span class="op">==</span> <span class="st">"non-targeting"</span><span class="op">)</span></span>
<span><span class="co">#&gt;   grna_id    grna_group</span></span>
<span><span class="co">#&gt; 1    NTg1 non-targeting</span></span>
<span><span class="co">#&gt; 2    NTg2 non-targeting</span></span>
<span><span class="co">#&gt; 3    NTg3 non-targeting</span></span>
<span><span class="co">#&gt; 4    NTg4 non-targeting</span></span>
<span><span class="co">#&gt; 5    NTg5 non-targeting</span></span>
<span><span class="co">#&gt; 6    NTg7 non-targeting</span></span>
<span><span class="co">#&gt; 7    NTg8 non-targeting</span></span>
<span><span class="co">#&gt; 8    NTg9 non-targeting</span></span>
<span><span class="co">#&gt; 9   NTg10 non-targeting</span></span></code></pre></div>
<p><code>sceptre</code> tests for association between responses and
targeting <em>gRNA groups</em> (as opposed to <em>individual
gRNAs</em>). <code>sceptre</code> tests a given response-gRNA group pair
for association by comparing the expression level of the response across
cells that have been infected by (any gRNA within) the given gRNA group
against cells that have been infected by a non-targeting gRNA.</p>
</div>
<div class="section level3">
<h3 id="step-2-determine-the-set-of-response-grna-group-pairs-to-analyze-">Step 2: Determine the set of response-gRNA group pairs to
analyze.<a class="anchor" aria-label="anchor" href="#step-2-determine-the-set-of-response-grna-group-pairs-to-analyze-"></a>
</h3>
<p>After preparing the data objects, we determine the set of
response-gRNA group pairs to test for association. This determination
typically is made on the basis of the biological question under
investigation. For example, in an experiment that seeks to map enhancers
of unknown function to target genes, we might pair a given
enhancer-targeting gRNA group to the set of genes in close physical
proximity to the targeted enhancer. In the absence of a specific
biological question, a reasonable strategy is to conduct an unbiased
screen of all response-gRNA group pairs. The function
<code>generate_all_pairs</code> couples each response to each gRNA
group, taking as input the response matrix and the gRNA group table.</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">response_grna_group_pairs</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/generate_all_pairs.html">generate_all_pairs</a></span><span class="op">(</span><span class="va">response_matrix_lowmoi</span>,</span>
<span>                                                <span class="va">grna_group_data_frame_lowmoi</span><span class="op">)</span></span></code></pre></div>
<p>The outputted data frame, <code>response_grna_group_pairs</code>, has
two columns: <code>response_id</code> and <code>grna_group</code>.</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">response_grna_group_pairs</span><span class="op">)</span> <span class="co"># each gRNA group is mapped to each gene</span></span>
<span><span class="co">#&gt;   response_id grna_group</span></span>
<span><span class="co">#&gt; 1       PCBP3       ATF2</span></span>
<span><span class="co">#&gt; 2    ITGB1BP1       ATF2</span></span>
<span><span class="co">#&gt; 3     SERTAD1       ATF2</span></span>
<span><span class="co">#&gt; 4       CHEK2       ATF2</span></span>
<span><span class="co">#&gt; 5      DUSP15       ATF2</span></span>
<span><span class="co">#&gt; 6       IFT46       ATF2</span></span></code></pre></div>
<p>In general the <code>response_grna_group_pairs</code> data frame
should specify the set of response-gRNA group pairs to test for
association.</p>
</div>
<div class="section level3">
<h3 id="step-3-set-the-formula-object-">Step 3: Set the formula object.<a class="anchor" aria-label="anchor" href="#step-3-set-the-formula-object-"></a>
</h3>
<p>The third step is to construct a formula object that specifies how to
adjust for the cell-specific covariates. We recommend including the
following covariates in the formula object:
<code>response_n_umis</code>, <code>response_n_nonzero</code>,
<code>p_mito</code>, and (if applicable) sequencing batch and/or
biological replicate. It is best practice to log-transform integer count
variables (e.g., <code>response_n_umis</code>,
<code>response_n_nonzero</code>), although this is not strictly speaking
necessary.</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">formula_object</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/formula.html" class="external-link">formula</a></span><span class="op">(</span><span class="op">~</span><span class="fu"><a href="https://rdrr.io/r/base/Log.html" class="external-link">log</a></span><span class="op">(</span><span class="va">response_n_umis</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://rdrr.io/r/base/Log.html" class="external-link">log</a></span><span class="op">(</span><span class="va">response_n_nonzero</span><span class="op">)</span> <span class="op">+</span></span>
<span>                            <span class="va">bio_rep</span> <span class="op">+</span> <span class="va">p_mito</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="step-4-select-the-pairwise-quality-control-qc-threshold-">Step 4: Select the pairwise quality control (QC) threshold.<a class="anchor" aria-label="anchor" href="#step-4-select-the-pairwise-quality-control-qc-threshold-"></a>
</h3>
<p>The fourth and final step of the setup is to select a pairwise QC
threshold. We digress briefly to discuss QC in the context of
single-cell CRISPR screens. Three types of QC that are common in
single-cell CRISPR screen analysis are response-wise QC, gRNA-wise QC,
and cell-wise QC, which aim to filter out low-quality responses,
low-quality gRNAs, and low-quality cells, respectively. A fourth type of
QC, <em>pairwise QC</em>, aims to filter out low-quality response-gRNA
group pairs. Pairwise QC, though not currently widespread in the
single-cell CRISPR screen space, arguably is the most important type of
QC in single-cell CRISPR screen analysis. <code>sceptre</code> provides
support for pairwise QC. (<code>sceptre</code> does not currently
implement response-wise, gRNA-wise, or cell-wise QC; users can implement
these forms of QC on their own if they choose before passing their data
to <code>sceptre</code>.)</p>
<p>For a given response-gRNA group pair, we define the “number of
nonzero treatment cells” as the number of “treatment cells” (i.e., cells
perturbed by the given gRNA group) in which the response exhibits
nonzero expression. Similarly, we define the “number of nonzero control
cells” as the number of “control cells” (i.e., cells that have received
a non-targeting gRNA) in which the response exhibits nonzero expression.
The number of nonzero treatment cells and the number of nonzero control
cells are good metrics of pair quality. <code>sceptre</code> filters out
pairs for which the number of nonzero treatment cells or the number of
nonzero control cells falls below some threshold. By default, this
threshold is set to 7 for both metrics. (See the section “Pairwise QC”
in Barry 2023 for a justification of this threshold.) While a threshold
of 7 is a good starting point, users can experiment with different
choices for this threshold by setting the parameters
<code>n_nonzero_trt_thresh</code> and
<code>n_nonzero_cntrl_thresh</code> to different values in the
<code>run_sceptre_lowmoi</code> function. The choice of pairwise QC
threshold can impact the results, sometimes considerably. A simple
illustration of pairwise QC is below.</p>
<div class="figure" style="text-align: center">
<img src="qc_crop.png" alt="Pairwise QC using a default threshold of 7 nonzero treatment cells and 7 nonzero control cells." width="550px"><p class="caption">
Pairwise QC using a default threshold of 7 nonzero treatment cells and 7
nonzero control cells.
</p>
</div>
</div>
</div>
<div class="section level2">
<h2 id="part-ii-calibration-check">Part II: Calibration check<a class="anchor" aria-label="anchor" href="#part-ii-calibration-check"></a>
</h2>
<p>Now that we have completed setup, we are ready to begin the data
analysis.</p>
<div class="section level3">
<h3 id="step-5--run-the-calibration-check-">Step 5. Run the calibration check.<a class="anchor" aria-label="anchor" href="#step-5--run-the-calibration-check-"></a>
</h3>
<p>We begin by running the “calibration check.” The calibration check is
an analysis that verifies that <code>sceptre</code> is able to control
the rate of false discoveries on the dataset under analysis. The
calibration check consists of randomly grouping together NT gRNAs to
form “negative control” gRNA groups. These negative control gRNA groups
are then paired to a large set of randomly-selected responses, forming
“negative control” response-gRNA group pairs. <code>sceptre</code> is
applied to analyze these negative control response-gRNA group pairs. As
the negative control pairs are devoid of biological signal, the p-values
that <code>sceptre</code> produces on the negative control pairs should
be uniformly distributed. Moreover, after an appropriate multiple
testing correction, <code>sceptre</code> should make zero (or very few)
discoveries on the negative control pairs. Verifying that
<code>sceptre</code> satisfies these properties ensures that the
discovery set that <code>sceptre</code> ultimately produces is
uncontaminated by excess false positives.</p>
<p>We run the calibration check by calling
<code>run_sceptre_lowmoi</code>, passing as arguments (i) the response
matrix, (ii) the gRNA matrix, (iii) the covariate data frame, (iv) the
gRNA group data frame, (v) the formula object, and (vi) the (discovery)
response-gRNA group pairs. We also set the
<code>calibration_check</code> argument to <code>TRUE</code>. The
example invocation takes about 30 seconds to run.</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">calibration_result</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/run_sceptre_lowmoi.html">run_sceptre_lowmoi</a></span><span class="op">(</span>response_matrix <span class="op">=</span> <span class="va">response_matrix_lowmoi</span>,</span>
<span>grna_matrix <span class="op">=</span> <span class="va">grna_matrix_lowmoi</span>,</span>
<span>covariate_data_frame <span class="op">=</span> <span class="va">covariate_data_frame_lowmoi</span>,</span>
<span>grna_group_data_frame <span class="op">=</span> <span class="va">grna_group_data_frame_lowmoi</span>,</span>
<span>formula_object <span class="op">=</span> <span class="va">formula_object</span>,</span>
<span>response_grna_group_pairs <span class="op">=</span> <span class="va">response_grna_group_pairs</span>,</span>
<span>calibration_check <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span> <span class="co"># calibration_check TRUE</span></span></code></pre></div>
<details><summary><strong>Details about the <code>calibration_result</code>
output</strong>
</summary><p>The output, <code>calibration_result</code>, is a data frame
containing 6,327 rows and six columns. Each row corresponds to a
negative control response-gRNA group pair. The columns are as
follows:</p>
<ul>
<li>
<code>response_id</code>: the response in the given negative control
response-gRNA group pair.</li>
<li>
<code>grna_group</code>: the gRNA group in the given negative
control response-gRNA group pair. The individual NT gRNAs that
constitute the gRNA group are listed, concatenated into a string via an
ampersand (“&amp;”) separator.</li>
<li>
<code>n_nonzero_trt</code>: the number of “treatment cells” (i.e.,
cells perturbed by the negative control gRNA group) with nonzero
response expression.</li>
<li>
<code>n_nonzero_cntrl</code>: the number of “control cells” (i.e.,
NT-receiving cells that did <em>not</em> receive the negative control
gRNA group) with nonzero response expression.</li>
<li>
<code>p_value</code>: the <code>sceptre</code> p-value for the test
of association between the gRNA group and the response.</li>
<li>
<code>log_2_fold_change</code>: the <code>sceptre</code>-estimated
log-2 fold change in expression of the response.</li>
</ul>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">calibration_result</span><span class="op">)</span></span>
<span><span class="co">#&gt;      response_id           grna_group n_nonzero_trt n_nonzero_cntrl</span></span>
<span><span class="co">#&gt; 1:        SLC8B1 NTg2&amp;NTg3&amp;NTg5&amp;NTg10           153             356</span></span>
<span><span class="co">#&gt; 2: RP11-708J19.1  NTg1&amp;NTg2&amp;NTg3&amp;NTg7            35              13</span></span>
<span><span class="co">#&gt; 3: RP11-708J19.1  NTg1&amp;NTg3&amp;NTg7&amp;NTg8            35              13</span></span>
<span><span class="co">#&gt; 4:          TIA1  NTg3&amp;NTg7&amp;NTg8&amp;NTg9           371             408</span></span>
<span><span class="co">#&gt; 5:        ANTXR1  NTg1&amp;NTg2&amp;NTg7&amp;NTg9           277             279</span></span>
<span><span class="co">#&gt; 6:        SLC8B1 NTg3&amp;NTg5&amp;NTg8&amp;NTg10           149             360</span></span>
<span><span class="co">#&gt;         p_value log_2_fold_change</span></span>
<span><span class="co">#&gt; 1: 0.0002350391        -0.3079779</span></span>
<span><span class="co">#&gt; 2: 0.0005058863         0.6106884</span></span>
<span><span class="co">#&gt; 3: 0.0005499262         0.6381855</span></span>
<span><span class="co">#&gt; 4: 0.0005822327         0.1751009</span></span>
<span><span class="co">#&gt; 5: 0.0006522506         0.2180436</span></span>
<span><span class="co">#&gt; 6: 0.0013039292        -0.2703491</span></span></code></pre></div>
</details><details><summary><strong>Details about how negative control pairs are
constructed</strong>
</summary><p>We describe how the negative control pairs are constructed. Recall
that our ultimate aim is to conduct a “discovery analysis” (step 7),
which entails applying <code>sceptre</code> to the discovery
response-gRNA group pairs specified in
<code>response_grna_group_pairs</code>. The calibration check is
“matched” to the discovery analysis in three important ways. First, the
number of gRNAs assigned to each negative control gRNA group is equal to
the median number of gRNAs per targeting gRNA group (four, in the case
of the example data). Second, the negative control pairs are subjected
to the same pairwise QC to which the discovery pairs are subjected. In
particular, the minimum number of treatment cells with nonzero
expression (<code>n_nonzero_trt</code>) and the minimum number of
control cells with nonzero expression (<code>n_nonzero_cntrl</code>) are
both greater than or equal to seven, the default QC threshold.</p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/Extremes.html" class="external-link">min</a></span><span class="op">(</span><span class="va">calibration_result</span><span class="op">$</span><span class="va">n_nonzero_trt</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] 7</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/Extremes.html" class="external-link">min</a></span><span class="op">(</span><span class="va">calibration_result</span><span class="op">$</span><span class="va">n_nonzero_cntrl</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] 7</span></span></code></pre></div>
<p>Finally, as we will confirm shortly, the number of negative control
pairs analyzed (i.e., 6,327) is equal to the number of discovery pairs
that passes pairwise QC. In summary, the calibration check closely
mirrors the discovery analysis, the difference being that the
calibration check is carried out on “matched” pairs devoid of biological
signal.</p>
</details>
</div>
<div class="section level3">
<h3 id="step-6-assess-calibration-">Step 6: Assess calibration.<a class="anchor" aria-label="anchor" href="#step-6-assess-calibration-"></a>
</h3>
<p>We next assess whether the negative control results are
well-calibrated. To do so, we call the function
<code><a href="../reference/plot_calibration_result.html">plot_calibration_result()</a></code>, which creates a visualization
that facilitates calibration assessment. The visualization consists of
four panels, which we describe below.</p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plot_calibration_result.html">plot_calibration_result</a></span><span class="op">(</span><span class="va">calibration_result</span><span class="op">)</span></span></code></pre></div>
<p><img src="calibration_result.png" width="500px" style="display: block; margin: auto;"></p>
<ul>
<li><p>The upper left panel is a Q-Q plot of the p-values on an
untransformed scale. The p-values should lie along the diagonal line,
indicating uniformity of the p-values in the <em>bulk</em> of the
distribution.</p></li>
<li><p>The upper right panel is a Q-Q plot of the p-values on a negative
log-10 transformed scale. The p-values again should lie along the
diagonal line (with the majority of the p-values falling within the gray
confidence band), indicating uniformity of the p-values in the
<em>tail</em> of the distribution.</p></li>
<li><p>The lower left panel is a histogram of the estimated log-2 fold
changes. The histogram should be roughly symmetric and centered around
zero.</p></li>
<li><p>Finally, the bottom right panel is a text box displaying (i) the
number of false discoveries that <code>sceptre</code> makes on the
negative control data (after a BH correction at level 0.1) and (ii) the
mean estimated log-fold change. The number of false discoveries should
be a small integer like zero, one, two, or three. (Zero is ideal.) The
mean estimated log-fold change, meanwhile, should be a numeric value
close to zero; a number in the range [-0.1, 0.1] is adequate.</p></li>
</ul>
<p><code>sceptre</code> may not exhibit good calibration initially. This
is OK. There are several strategies that users employ to improve the
calibration of <code>sceptre</code>. First, one can make the QC
threshold more stringent by increasing <code>n_nonzero_trt_thresh</code>
and <code>n_nonzero_cntrl_thresh</code>. Second, one can deactivate the
asymptotic inference module of <code>sceptre</code> by setting
<code>fit_skew_normal</code> to <code>FALSE</code>. Third, one can add
additional, potentially relevant covariates to the
<code>formula_object</code>. A complete article discussing strategies
for improving the calibration of <code>sceptre</code> is
forthcoming.</p>
</div>
</div>
<div class="section level2">
<h2 id="part-iii-discovery-analysis">Part III: Discovery analysis<a class="anchor" aria-label="anchor" href="#part-iii-discovery-analysis"></a>
</h2>
<div class="section level3">
<h3 id="step-7--run-the-discovery-analysis-">Step 7. Run the discovery analysis.<a class="anchor" aria-label="anchor" href="#step-7--run-the-discovery-analysis-"></a>
</h3>
<p>Once we have verified that <code>sceptre</code> is adequately
calibrated, we are ready to carry out the discovery analysis. We run the
discovery analysis by calling <code>run_sceptre_lowmoi</code>, passing
the same arguments that we passed for the calibration check but setting
<code>calibration_check</code> to <code>FALSE</code>.</p>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">discovery_result</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/run_sceptre_lowmoi.html">run_sceptre_lowmoi</a></span><span class="op">(</span>response_matrix <span class="op">=</span> <span class="va">response_matrix_lowmoi</span>,</span>
<span>grna_matrix <span class="op">=</span> <span class="va">grna_matrix_lowmoi</span>,</span>
<span>covariate_data_frame <span class="op">=</span> <span class="va">covariate_data_frame_lowmoi</span>,</span>
<span>grna_group_data_frame <span class="op">=</span> <span class="va">grna_group_data_frame_lowmoi</span>,</span>
<span>formula_object <span class="op">=</span> <span class="va">formula_object</span>,</span>
<span>response_grna_group_pairs <span class="op">=</span> <span class="va">response_grna_group_pairs</span>,</span>
<span>calibration_check <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span> <span class="co"># calibration check FALSE</span></span></code></pre></div>
<p><code>run_sceptre_lowmoi</code> (with <code>calibration_check</code>
set to <code>FALSE</code>) computes a p-value and estimates the log-fold
change for each response-gRNA group pair specified in
<code>response_grna_group_pairs</code> (that passes QC). The columns of
the output data frame, <code>discovery_result</code>, are the same as
those of <code>calibration_result</code>. <code>response_id</code> and
<code>grna_group</code> indicate the response and gRNA group that
constitute the given pair; <code>n_nonzero_trt</code> and
<code>n_nonzero_cntrl</code> are the number of nonzero treatment cells
and nonzero control cells, respectively; and <code>p_value</code> and
<code>log_2_fold_change</code> are the p-value and log-2 fold change for
the pair. Rows are sorted by <code>p_value</code>.</p>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">discovery_result</span><span class="op">)</span></span>
<span><span class="co">#&gt;    response_id grna_group n_nonzero_trt n_nonzero_cntrl       p_value</span></span>
<span><span class="co">#&gt; 1:       PSMB9     IFNGR2          1061            2350 5.037028e-173</span></span>
<span><span class="co">#&gt; 2:       PSMB9       JAK2           890            2350 4.325720e-159</span></span>
<span><span class="co">#&gt; 3:       PSMB9     IFNGR1          1184            2350 9.583334e-159</span></span>
<span><span class="co">#&gt; 4:       PSMB9      STAT1           341            2350 2.213593e-116</span></span>
<span><span class="co">#&gt; 5:      ANTXR1     IFNGR1           565             556  3.577309e-44</span></span>
<span><span class="co">#&gt; 6:      ANTXR1       JAK2           486             556  1.825734e-40</span></span>
<span><span class="co">#&gt;    log_2_fold_change</span></span>
<span><span class="co">#&gt; 1:         -1.234883</span></span>
<span><span class="co">#&gt; 2:         -1.410967</span></span>
<span><span class="co">#&gt; 3:         -1.203587</span></span>
<span><span class="co">#&gt; 4:         -1.596683</span></span>
<span><span class="co">#&gt; 5:          1.168111</span></span>
<span><span class="co">#&gt; 6:          1.295196</span></span></code></pre></div>
<p>Pairs that fail to pass QC are assigned a <code>p_value</code> and
<code>log_2_fold_change</code> of <code>NA</code> and appear in the
bottom rows of the <code>discovery_result</code> data frame.</p>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">tail</a></span><span class="op">(</span><span class="va">discovery_result</span><span class="op">)</span></span>
<span><span class="co">#&gt;    response_id grna_group n_nonzero_trt n_nonzero_cntrl p_value</span></span>
<span><span class="co">#&gt; 1:      ZNF699      CD274             0             141      NA</span></span>
<span><span class="co">#&gt; 2:      ZNF699        MYC             5             141      NA</span></span>
<span><span class="co">#&gt; 3:      ZNF699       SPI1             1             141      NA</span></span>
<span><span class="co">#&gt; 4:       ZNF74      CD274             0              67      NA</span></span>
<span><span class="co">#&gt; 5:       ZNF74        MYC             3              67      NA</span></span>
<span><span class="co">#&gt; 6:       ZNF74       SPI1             2              67      NA</span></span>
<span><span class="co">#&gt;    log_2_fold_change</span></span>
<span><span class="co">#&gt; 1:                NA</span></span>
<span><span class="co">#&gt; 2:                NA</span></span>
<span><span class="co">#&gt; 3:                NA</span></span>
<span><span class="co">#&gt; 4:                NA</span></span>
<span><span class="co">#&gt; 5:                NA</span></span>
<span><span class="co">#&gt; 6:                NA</span></span></code></pre></div>
<details><summary><strong>Details about the discovery pairs</strong>
</summary><p>We briefly examine the discovery pairs that pass pairwise QC by
filtering <code>discovery_result</code> for rows that do not contain
<code>NA</code>.</p>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">discovery_result_pass_qc</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/na.fail.html" class="external-link">na.omit</a></span><span class="op">(</span><span class="va">discovery_result</span><span class="op">)</span></span></code></pre></div>
<p>Each discovery pair that passes QC has at least seven treatment cells
and seven control cells with nonzero expression.</p>
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/Extremes.html" class="external-link">min</a></span><span class="op">(</span><span class="va">discovery_result_pass_qc</span><span class="op">$</span><span class="va">n_nonzero_trt</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] 7</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/Extremes.html" class="external-link">min</a></span><span class="op">(</span><span class="va">discovery_result_pass_qc</span><span class="op">$</span><span class="va">n_nonzero_cntrl</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] 10</span></span></code></pre></div>
<p>Moreover, the number of discovery pairs that passes QC is equal to
the number of negative control pairs.</p>
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">discovery_result_pass_qc</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] 6327</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">calibration_result</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] 6327</span></span></code></pre></div>
</details>
</div>
<div class="section level3">
<h3 id="step-8--visualize-the-discovery-analysis-results-">Step 8. Visualize the discovery analysis results.<a class="anchor" aria-label="anchor" href="#step-8--visualize-the-discovery-analysis-results-"></a>
</h3>
<p>Next, we create a couple plots to help us visualize the discovery
results. We plot the negative control p-values and the discovery
p-values on the same Q-Q plot via a call to
<code>compare_calibration_and_discovery_results</code>. The negative
control p-values are plotted in red and the discovery p-values in blue.
The horizontal dashed line indicates the multiple testing threshold
(which, by default, is the BH threshold at level 0.1). Blue points above
the horizontal dashed line are called as discoveries. The negative
control p-values should lie along the diagonal line and fall mostly
within the gray confidence band. The discovery p-values, by contrast,
should trend above the diagonal line, indicating the presence of signal
in the discovery set.</p>
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/compare_calibration_and_discovery_results.html">compare_calibration_and_discovery_results</a></span><span class="op">(</span></span>
<span>  calibration_result <span class="op">=</span> <span class="va">calibration_result</span>,</span>
<span>  discovery_result <span class="op">=</span> <span class="va">discovery_result</span><span class="op">)</span></span></code></pre></div>
<p><img src="comparison_plot.png" width="400px" style="display: block; margin: auto;"></p>
<p>Second, we create a volcano plot of the discovery results via a call
to <code>make_volcano_plot</code>. Each point in the plot corresponds to
a response-gRNA group pair; the estimated log-2 fold change of the pair
is plotted on the x-axis, and the (negative log-10 transformed) p-value
is plotted on the y-axis. The horizontal dashed line again indicates the
multiple testing threshold. Points above the dashed line (colored in
purple) are called as discoveries, while points below (colored in blue)
are called as insignificant.</p>
<div class="sourceCode" id="cb29"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/make_volcano_plot.html">make_volcano_plot</a></span><span class="op">(</span>discovery_result <span class="op">=</span> <span class="va">discovery_result</span><span class="op">)</span></span></code></pre></div>
<p><img src="volcano_plot.png" width="400px" style="display: block; margin: auto;"></p>
</div>
<div class="section level3">
<h3 id="step-9-obtain-the-discovery-set-for-downstream-analysis-">Step 9: Obtain the discovery set for downstream analysis.<a class="anchor" aria-label="anchor" href="#step-9-obtain-the-discovery-set-for-downstream-analysis-"></a>
</h3>
<p>Finally, we obtain the discovery set. We pass
<code>discovery_result</code> to <code>obtain_discovery_set</code>,
which returns the subset of response-gRNA group pairs that are called as
discoveries.</p>
<div class="sourceCode" id="cb30"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">discovery_set</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/obtain_discovery_set.html">obtain_discovery_set</a></span><span class="op">(</span><span class="va">discovery_result</span><span class="op">)</span></span></code></pre></div>
<p>We can use <code>discovery_set</code> for various downstream
analyses.</p>
<details><summary><strong>Applying different multiple testing correction
procedures</strong>
</summary>
We have used the BH correction at level 0.1 throughout this analysis.
One can change the multiple testing correction procedure and level by
setting the arguments <code>multiple_testing_correction</code> and
<code>alpha</code> to different values in the functions
<code>plot_calibration_result</code>,
<code>compare_calibration_and_discovery_results</code>,
<code>make_volcano_plot</code>, and <code>obtain_discovery_set</code>.
See the documentation of these functions for more information.
</details>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by <a href="https://timothy-barry.github.io/" class="external-link">Timothy Barry</a>, <a href="https://ekatsevi.github.io/" class="external-link">Eugene Katsevich</a>.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.5.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
