<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="sceptre">
<title>Get started with sceptre • sceptre</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.2.2/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.2.2/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.rawgit.com/afeld/bootstrap-toc/v1.0.1/dist/bootstrap-toc.min.js"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Get started with sceptre">
<meta property="og:description" content="sceptre">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">sceptre</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.9.1</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="active nav-item">
  <a class="nav-link" href="../articles/sceptre.html">
    <span class="fa fa-play"></span>
     
    Get started
  </a>
</li>
<li class="nav-item">
  <a class="external-link nav-link" href="https://timothy-barry.github.io/sceptre-book/">
    <span class="fa fa-book"></span>
     
    Manual
  </a>
</li>
<li class="nav-item">
  <a class="external-link nav-link" href="https://timothy-barry.github.io/sceptre-book/faq.html">
    <span class="fa fa-question-circle-o"></span>
     
    FAQ
  </a>
</li>
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">
    <span class="fa fa-code"></span>
     
    Functions
  </a>
</li>
<li class="nav-item">
  <a class="nav-link" href="../news/index.html">
    <span class="fa fa-history"></span>
     
    Changelog
  </a>
</li>
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/Katsevich-Lab/sceptre/discussions">
    <span class="fa fa-users"></span>
     
    Community
  </a>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav">
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/Katsevich-Lab/sceptre/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>

    
  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>Get started with sceptre</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/Katsevich-Lab/sceptre/blob/HEAD/vignettes/sceptre.Rmd" class="external-link"><code>vignettes/sceptre.Rmd</code></a></small>
      <div class="d-none name"><code>sceptre.Rmd</code></div>
    </div>

    
    
<p><code>sceptre</code> is an R package that facilitates statistically
rigorous, computationally efficient, and user-friendly single-cell
CRISPR screen data analysis. Users can install <code>sceptre</code> via
the following command.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"devtools"</span><span class="op">)</span></span>
<span><span class="fu">devtools</span><span class="fu">::</span><span class="fu"><a href="https://remotes.r-lib.org/reference/install_github.html" class="external-link">install_github</a></span><span class="op">(</span><span class="st">"katsevich-lab/sceptre"</span><span class="op">)</span></span></code></pre></div>
<p>(Please see <a href="install.html">this page</a> for instructions to
install older versions of <code>sceptre</code>, though we recommend
users switch to the latest version as soon as they can.) Users can load
<code>sceptre</code> via a call to <code><a href="https://rdrr.io/r/base/library.html" class="external-link">library()</a></code>.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/Katsevich-Lab/sceptre" class="external-link">sceptre</a></span><span class="op">)</span></span></code></pre></div>
<p>The standard pipeline involved in applying <code>sceptre</code> to
analyze a dataset consists of several steps, which we summarize in the
following schematic.</p>
<div class="figure" style="text-align: center">
<img src="pipeline_schematic.png" alt="Standard `sceptre` pipeline" width="650px"><p class="caption">
Standard <code>sceptre</code> pipeline
</p>
</div>
<p>This vignette illustrates application of <code>sceptre</code> to a
CRISPRi screen of candidate enhancers in K562 cells (Gasperini et al,
2019). The goal of the analysis is to confidently link enhancers to
genes by testing for changes in gene expression in response to the
CRISPR perturbations of the candidate enhancers.</p>
<p>Using <code>sceptre</code> is simple; carrying out an entire analysis
requires only a few lines of code. Below, we provide a minimal working
example of applying <code>sceptre</code> to analyze the example data.
First, we load the example data into R via the function
<code><a href="../reference/import_data_from_cellranger.html">import_data_from_cellranger()</a></code>, which creates a
<code>sceptre_object</code>, an object-based representation of the
single-cell CRISPR screen data. Next, we specify the gRNA-gene pairs
that we seek to test for association. Then, we call the pipeline
functions on the <code>sceptre_object</code> in order. Finally, we write
the outputs to the directory <code>~/sceptre_outputs</code>.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># load the data, creating a sceptre_object</span></span>
<span><span class="va">directories</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, package <span class="op">=</span> <span class="st">"sceptre"</span><span class="op">)</span>,</span>
<span>  <span class="st">"/highmoi_example/gem_group_"</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">2</span></span>
<span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">grna_target_data_frame_highmoi</span><span class="op">)</span></span>
<span><span class="va">sceptre_object</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/import_data_from_cellranger.html">import_data_from_cellranger</a></span><span class="op">(</span></span>
<span>  directories <span class="op">=</span> <span class="va">directories</span>,</span>
<span>  moi <span class="op">=</span> <span class="st">"high"</span>,</span>
<span>  grna_target_data_frame <span class="op">=</span> <span class="va">grna_target_data_frame_highmoi</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># construct the grna-gene pairs to analyze</span></span>
<span><span class="va">positive_control_pairs</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/construct_positive_control_pairs.html">construct_positive_control_pairs</a></span><span class="op">(</span><span class="va">sceptre_object</span><span class="op">)</span></span>
<span><span class="va">discovery_pairs</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/construct_cis_pairs.html">construct_cis_pairs</a></span><span class="op">(</span></span>
<span>  <span class="va">sceptre_object</span>, </span>
<span>  positive_control_pairs <span class="op">=</span> <span class="va">positive_control_pairs</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># apply the pipeline functions to the sceptre_object in order</span></span>
<span><span class="va">sceptre_object</span> <span class="op">&lt;-</span> <span class="va">sceptre_object</span> <span class="op">|&gt;</span> <span class="co"># |&gt; is R's base pipe, similar to %&gt;%</span></span>
<span>  <span class="fu"><a href="../reference/set_analysis_parameters.html">set_analysis_parameters</a></span><span class="op">(</span><span class="va">discovery_pairs</span>, <span class="va">positive_control_pairs</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="../reference/run_calibration_check.html">run_calibration_check</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="../reference/run_power_check.html">run_power_check</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="../reference/run_discovery_analysis.html">run_discovery_analysis</a></span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># write the results to disk</span></span>
<span><span class="fu"><a href="../reference/write_outputs_to_directory.html">write_outputs_to_directory</a></span><span class="op">(</span><span class="va">sceptre_object</span>, <span class="st">"~/sceptre_outputs"</span><span class="op">)</span></span></code></pre></div>
<p>That’s it. The directory <code>sceptre_outputs</code> now contains a
variety of results and plots from the analysis, including the set of
significant gRNA-gene pairs:</p>
<pre><code><span><span class="co">#&gt;         grna_target     response_id      p_value log_2_fold_change</span></span>
<span><span class="co">#&gt; 1: candidate_enh_19 ENSG00000147454 4.362186e-40        -0.7337873</span></span>
<span><span class="co">#&gt; 2: candidate_enh_23 ENSG00000130560 8.639233e-14        -0.1637574</span></span>
<span><span class="co">#&gt; 3:  candidate_enh_1 ENSG00000069275 3.708631e-05        -0.1629311</span></span>
<span><span class="co">#&gt; 4:  candidate_enh_3 ENSG00000126698 7.996432e-04         0.1087084</span></span>
<span><span class="co">#&gt; 5:  candidate_enh_2 ENSG00000066027 9.610934e-04        -0.2275959</span></span></code></pre>
<p>We describe each step of the pipeline in greater detail below.</p>
<div class="section level2">
<h2 id="import-data">1. Import data<a class="anchor" aria-label="anchor" href="#import-data"></a>
</h2>
<p>The first step is to import the data. <strong>Data can be imported
into <code>sceptre</code> from 10X Cell Ranger outputs or R matrices;
support for Parse Biosciences outputs is coming soon.</strong> The
simplest way to import the data is to read the output of one or more
calls to <code>cellranger_count()</code> into <code>sceptre</code> via
the function <code><a href="../reference/import_data_from_cellranger.html">import_data_from_cellranger()</a></code>.
<code><a href="../reference/import_data_from_cellranger.html">import_data_from_cellranger()</a></code> requires three arguments:
<code>directories</code>, <code>grna_target_data_frame</code>, and
<code>moi</code>.</p>
<ol style="list-style-type: decimal">
<li>
<p><code>directories</code> is a character vector specifying the
locations of the directories outputted by one or more calls to
<code>cellranger_count()</code>. Below, we set the variable
<code>directories</code> to the (machine-dependent) location of the
example CRISPRi data on disk.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">directories</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, package <span class="op">=</span> <span class="st">"sceptre"</span><span class="op">)</span>, </span>
<span>  <span class="st">"/highmoi_example/gem_group_"</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">2</span></span>
<span><span class="op">)</span></span>
<span><span class="va">directories</span> <span class="co"># file paths to the example data on your computer</span></span>
<span><span class="co">#&gt; [1] "/private/var/folders/h6/pt59hmwd7rqb3jqfzwh0hlth0000gn/T/Rtmp031ihx/temp_libpath751c3e636f4a/sceptre/extdata/highmoi_example/gem_group_1"</span></span>
<span><span class="co">#&gt; [2] "/private/var/folders/h6/pt59hmwd7rqb3jqfzwh0hlth0000gn/T/Rtmp031ihx/temp_libpath751c3e636f4a/sceptre/extdata/highmoi_example/gem_group_2"</span></span></code></pre></div>
<p><code>directories</code> points to two directories, both of which
store the expression data in matrix market format and contain the files
<code>barcodes.tsv.gz</code>, <code>features.tsv.gz</code>, and
<code>matrix.mtx.gz</code>.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/list.files.html" class="external-link">list.files</a></span><span class="op">(</span><span class="va">directories</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "barcodes.tsv.gz" "features.tsv.gz" "matrix.mtx.gz"</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/list.files.html" class="external-link">list.files</a></span><span class="op">(</span><span class="va">directories</span><span class="op">[</span><span class="fl">2</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "barcodes.tsv.gz" "features.tsv.gz" "matrix.mtx.gz"</span></span></code></pre></div>
</li>
<li>
<p><code>grna_target_data_frame</code> is a data frame mapping each
individual gRNA to the genomic element that the gRNA targets.
<code>grna_target_data_frame</code> contains two required columns:
<code>grna_id</code> and <code>grna_target</code>. <code>grna_id</code>
is the ID of an individual gRNA, while <code>grna_target</code> is a
label specifying the genomic element that the gRNA targets. (Typically,
multiple gRNAs are designed to target a given genomic element in a
single-cell CRISPR screen.) Non-targeting (NT) gRNAs are assigned a gRNA
target label of “non-targeting”. <code>grna_target_data_frame</code>
optionally contains the columns <code>chr</code>, <code>start</code>,
and <code>end</code>, which give the chromosome, start coordinate, and
end coordinate, respectively, of the genomic region that each gRNA
targets. We load and examine the <code>grna_target_data_frame</code>
corresponding to the example data.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">grna_target_data_frame_highmoi</span><span class="op">)</span></span>
<span><span class="va">grna_target_data_frame_highmoi</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fl">4</span>, <span class="fl">21</span><span class="op">:</span><span class="fl">24</span>, <span class="fl">80</span><span class="op">:</span><span class="fl">83</span><span class="op">)</span>,<span class="op">]</span></span>
<span><span class="co">#&gt;         grna_id      grna_target   chr     start       end</span></span>
<span><span class="co">#&gt; 1  grna_CCGGGCG  ENSG00000069482 chr11  68451943  68451958</span></span>
<span><span class="co">#&gt; 2  grna_TGGCGGC  ENSG00000069482 chr11  68451958  68451974</span></span>
<span><span class="co">#&gt; 3  grna_AAGGCCG  ENSG00000100316 chr22  39715775  39715790</span></span>
<span><span class="co">#&gt; 4  grna_GACGCCG  ENSG00000100316 chr22  39715790  39715806</span></span>
<span><span class="co">#&gt; 21 grna_TCTTGAC  candidate_enh_1  chr1 205720419 205720623</span></span>
<span><span class="co">#&gt; 22 grna_TGATCGC  candidate_enh_1  chr1 205720623 205720828</span></span>
<span><span class="co">#&gt; 23 grna_ATAGGAG candidate_enh_10 chr19  19475680  19475878</span></span>
<span><span class="co">#&gt; 24 grna_GCAGGCA candidate_enh_10 chr19  19475878  19476077</span></span>
<span><span class="co">#&gt; 80 grna_TATTCGT    non-targeting  &lt;NA&gt;        NA        NA</span></span>
<span><span class="co">#&gt; 81 grna_GACCTCC    non-targeting  &lt;NA&gt;        NA        NA</span></span>
<span><span class="co">#&gt; 82 grna_TTTCTCT    non-targeting  &lt;NA&gt;        NA        NA</span></span>
<span><span class="co">#&gt; 83 grna_AATGAGG    non-targeting  &lt;NA&gt;        NA        NA</span></span></code></pre></div>
<p>Some gRNAs (e.g., <code>grna_CCGGGCG</code> and
<code>grna_TGGCGGC</code>) target gene transcription start sites and
serve as positive controls; other gRNAs (e.g., <code>grna_TCTTGAC</code>
and <code>grna_TGATCGC</code>) target candidate enhancers, while others
still (e.g., <code>grna_TATTCGT</code> and <code>grna_GACCTCC</code>)
are non-targeting. Each gene and candidate enhancer in this dataset is
targeted by exactly two gRNAs.</p>
</li>
<li>
<p><code>moi</code> is a string specifying the
multiplicity-of-infection (MOI) of the data, taking values “high” or
“low”. A high-MOI (respectively, low-MOI) dataset is one in which the
experimenter has aimed to insert multiple gRNAs (respectively, a single
gRNA) into each cell. (If a given cell is determined to contain multiple
gRNAs in a low-MOI screen, that cell is removed as part of the quality
control step, as discussed below.) The example dataset is a high MOI
dataset, and so we set <code>moi</code> to “high”.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">moi</span> <span class="op">&lt;-</span> <span class="st">"high"</span></span></code></pre></div>
</li>
</ol>
<p>Finally, we call the function
<code><a href="../reference/import_data_from_cellranger.html">import_data_from_cellranger()</a></code>, passing
<code>directories</code>, <code>grna_target_data_frame</code>, and
<code>moi</code> as arguments.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sceptre_object</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/import_data_from_cellranger.html">import_data_from_cellranger</a></span><span class="op">(</span></span>
<span>  directories <span class="op">=</span> <span class="va">directories</span>,</span>
<span>  grna_target_data_frame <span class="op">=</span> <span class="va">grna_target_data_frame_highmoi</span>,</span>
<span>  moi <span class="op">=</span> <span class="va">moi</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p><code><a href="../reference/import_data_from_cellranger.html">import_data_from_cellranger()</a></code> returns a
<code>sceptre_object</code>, which is an object-based representation of
the single-cell CRISPR screen data. Evaluating
<code>sceptre_object</code> in the console prints a helpful summary of
the data.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sceptre_object</span></span>
<span><span class="co">#&gt; An object of class <span style="color: #0000BB;">sceptre_object</span>.</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Attributes of the data:</span></span>
<span><span class="co">#&gt;  • <span style="color: #0000BB;">45919</span> cells</span></span>
<span><span class="co">#&gt;  • <span style="color: #0000BB;">526</span> responses</span></span>
<span><span class="co">#&gt;  • <span style="color: #0000BB;">High</span> multiplicity-of-infection </span></span>
<span><span class="co">#&gt;  • <span style="color: #0000BB;">70</span> targeting gRNAs (distributed across <span style="color: #0000BB;">35</span> targets) </span></span>
<span><span class="co">#&gt;  • <span style="color: #0000BB;">25</span> non-targeting gRNAs </span></span>
<span><span class="co">#&gt;  • <span style="color: #0000BB;">6</span> covariates (batch, grna_n_nonzero, grna_n_umis, response_n_nonzero, response_n_umis, response_p_mito)</span></span></code></pre></div>
<p>Several metrics are displayed, including the number of cells, the
number of genes (or “responses”), and the number of gRNAs present in the
data. <code>sceptre</code> also automatically computes the following
cell-specific covariates: <code>grna_n_nonzero</code> (i.e., the number
of gRNAs expressed in the cell), <code>grna_n_umis</code> (i.e., the
number of gRNA UMIs sequenced in the cell),
<code>response_n_nonzero</code> (i.e., the number of responses expressed
in the cell), <code>response_n_umis</code> (i.e., the number of response
UMIs sequenced in the cell), <code>response_p_mito</code> (i.e., the
fraction of transcripts mapping to mitochondrial genes), and
<code>batch</code>. (Cells loaded from different directories are assumed
to come from different batches.)</p>
<p>See <a href="https://timothy-barry.github.io/sceptre-book/import-data.html" class="external-link">Chapter
1 of the manual</a> for more details about data import.</p>
</div>
<div class="section level2">
<h2 id="set-analysis-parameters">2. Set analysis parameters<a class="anchor" aria-label="anchor" href="#set-analysis-parameters"></a>
</h2>
<p>The second step is to set the analysis parameters. The most important
analysis parameters are the discovery pairs, positive control pairs,
sidedness, and gRNA grouping strategy.</p>
<ol style="list-style-type: decimal">
<li>
<p><strong>Discovery pairs and positive control pairs</strong>. The
primary goal of <code>sceptre</code> is to determine whether
perturbation of a gRNA target (such as an enhancer) leads to a change in
expression of a response (such as gene). We use the term
<em>target-response pair</em> to refer to a given gRNA target and
response that we seek to test for association (upon perturbation of the
gRNA target). A <em>discovery target-response pair</em> is a
target-response pair whose association status we do not know but would
like to learn. For example, in an experiment in which we aim to link
putative enhancers to genes, the discovery target-response pairs might
consist of the set of putative enhancers and genes in close physical
proximity to one another.</p>
<p>A <em>positive control</em> (resp., <em>negative control</em>)
<em>target-response pair</em> is a target-response pair for which we
know that there <em>is</em> (resp., is <em>not</em>) a relationship
between the target and the response. Positive control target-response
pairs often are formed by coupling a transcription start site to the
gene known to be regulated by that transcription start site. Negative
control target-response pairs, meanwhile, typically are constructed by
pairing negative control gRNAs to one or more responses. (We defer a
detailed discussion of negative control pairs to a later section of this
vignette.) Discovery pairs are of primary scientific interest, while
positive control and negative control pairs serve a mainly technial
purpose, helping us verify that the biological assay and statistical
methodology are in working order.</p>
<p><code>sceptre</code> offers several helper functions to facilitate
the construction of positive control and discovery pairs. The function
<code><a href="../reference/construct_positive_control_pairs.html">construct_positive_control_pairs()</a></code> takes as argument a
<code>sceptre_object</code> and outputs the set of positive control
pairs formed by matching gRNA targets (as contained in the
<code>grna_target_data_frame</code>) to response IDs. Positive control
pairs are optional and need not be computed.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">positive_control_pairs</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/construct_positive_control_pairs.html">construct_positive_control_pairs</a></span><span class="op">(</span><span class="va">sceptre_object</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">positive_control_pairs</span><span class="op">)</span></span>
<span><span class="co">#&gt;       grna_target     response_id</span></span>
<span><span class="co">#&gt; 1 ENSG00000069482 ENSG00000069482</span></span>
<span><span class="co">#&gt; 2 ENSG00000100316 ENSG00000100316</span></span>
<span><span class="co">#&gt; 3 ENSG00000104131 ENSG00000104131</span></span>
<span><span class="co">#&gt; 4 ENSG00000122026 ENSG00000122026</span></span>
<span><span class="co">#&gt; 5 ENSG00000135821 ENSG00000135821</span></span>
<span><span class="co">#&gt; 6 ENSG00000147669 ENSG00000147669</span></span></code></pre></div>
<p>Next, the functions <code><a href="../reference/construct_cis_pairs.html">construct_cis_pairs()</a></code> and
<code><a href="../reference/construct_trans_pairs.html">construct_trans_pairs()</a></code> facilitate the construction of
<em>cis</em> and <em>trans</em> discovery sets, respectively.
<code><a href="../reference/construct_cis_pairs.html">construct_cis_pairs()</a></code> takes as arguments a
<code>sceptre_object</code> and an integer
<code>distance_threshold</code> and returns the set of response-target
pairs located on the same chromosome within
<code>distance_threshold</code> bases of one another.
<code>positive_control_pairs</code> optionally can be passed to this
function, in which case positive control gRNA targets are excluded from
the <em>cis</em> pairs. (Note that <code><a href="../reference/construct_cis_pairs.html">construct_cis_pairs()</a></code>
assumes that the responses are genes rather than, say, proteins or
chromatin-derived features.)</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">discovery_pairs</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/construct_cis_pairs.html">construct_cis_pairs</a></span><span class="op">(</span></span>
<span>  sceptre_object <span class="op">=</span> <span class="va">sceptre_object</span>,</span>
<span>  positive_control_pairs <span class="op">=</span> <span class="va">positive_control_pairs</span>,</span>
<span>  distance_threshold <span class="op">=</span> <span class="fl">5e6</span></span>
<span><span class="op">)</span></span>
<span><span class="va">discovery_pairs</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fl">4</span>, <span class="fl">101</span><span class="op">:</span><span class="fl">104</span><span class="op">)</span>,<span class="op">]</span></span>
<span><span class="co">#&gt;          grna_target     response_id</span></span>
<span><span class="co">#&gt; 1    candidate_enh_1 ENSG00000174529</span></span>
<span><span class="co">#&gt; 2    candidate_enh_1 ENSG00000117222</span></span>
<span><span class="co">#&gt; 3    candidate_enh_1 ENSG00000133059</span></span>
<span><span class="co">#&gt; 4    candidate_enh_1 ENSG00000133069</span></span>
<span><span class="co">#&gt; 101 candidate_enh_10 ENSG00000064490</span></span>
<span><span class="co">#&gt; 102 candidate_enh_10 ENSG00000254901</span></span>
<span><span class="co">#&gt; 103 candidate_enh_10 ENSG00000184162</span></span>
<span><span class="co">#&gt; 104 candidate_enh_10 ENSG00000105705</span></span></code></pre></div>
<p><code><a href="../reference/construct_trans_pairs.html">construct_trans_pairs()</a></code> constructs the entire set of
possible target-response pairs. See <a href="https://timothy-barry.github.io/sceptre-book/set-analysis-parameters.html" class="external-link">Chapter
2 of the manual</a> for more information about these functions.</p>
</li>
<li>
<p><strong>Sidedness</strong>. The parameter <code>side</code>
controls whether to run a left-tailed (<code>"left"</code>),
right-tailed (<code>"right"</code>), or two-tailed (<code>"both"</code>;
default) test. A left-tailed (resp., right-tailed) test is appropriate
when testing for a decrease (resp., increase) in expression; a
two-tailed test, by contrast, is appropriate when testing for an
increase <em>or</em> decrease in expression. A left-tailed test is the
most appropriate choice for a CRISPRi screen of enhancers, and so we set
<code>side</code> to <code>"left"</code>.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">side</span> <span class="op">&lt;-</span> <span class="st">"left"</span></span></code></pre></div>
</li>
<li><p><strong>gRNA integration strategy</strong>. Typically, multiple
gRNAs are designed to target a given genomic element. The parameter
<code>grna_integration_strategy</code> controls if and how gRNAs that
target the same genomic element are integrated. The default option,
<code>"union"</code>, combines gRNAs that target the same element into a
single “grouped gRNA;” this “grouped gRNA” is tested for association
against the responses to which the element is paired.
<code>grna_integration_strategy</code> also can be set to “singleton,”
in which case each gRNA targeting a given element is tested individually
against the responses paired to that element. In our analysis we use the
default “union” strategy.</p></li>
</ol>
<p>Finally, we set the analysis parameters by calling the function
<code><a href="../reference/set_analysis_parameters.html">set_analysis_parameters()</a></code>, passing
<code>sceptre_object</code>, <code>discovery_pairs</code>,
<code>positive_control_pairs</code>, and <code>side</code> as arguments.
Note that <code>sceptre_object</code> and <code>discovery_pairs</code>
are the only required arguments to this function.</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sceptre_object</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/set_analysis_parameters.html">set_analysis_parameters</a></span><span class="op">(</span></span>
<span>  sceptre_object <span class="op">=</span> <span class="va">sceptre_object</span>,</span>
<span>  discovery_pairs <span class="op">=</span> <span class="va">discovery_pairs</span>,</span>
<span>  positive_control_pairs <span class="op">=</span> <span class="va">positive_control_pairs</span>,</span>
<span>  side <span class="op">=</span> <span class="va">side</span></span>
<span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="va">sceptre_object</span><span class="op">)</span> <span class="co"># output suppressed for brevity</span></span></code></pre></div>
<p>See <a href="https://timothy-barry.github.io/sceptre-book/set-analysis-parameters.html" class="external-link">Chapter
2 of the manual</a> for more information about setting the analysis
parameters.</p>
</div>
<div class="section level2">
<h2 id="sec-sceptre_assign_grnas">3. Assign gRNAs to cells (optional)<a class="anchor" aria-label="anchor" href="#sec-sceptre_assign_grnas"></a>
</h2>
<p>The third step is to assign gRNAs to cells. This step can be skipped,
in which case gRNAs are assigned to cells automatically using default
options. The gRNA assignment step involves using the gRNA UMI counts to
determine which cells contain which gRNAs. We begin by plotting the UMI
count distribution of several randomly selected gRNAs via a call to the
function <code><a href="../reference/plot_grna_count_distributions.html">plot_grna_count_distributions()</a></code>.</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plot_grna_count_distributions.html">plot_grna_count_distributions</a></span><span class="op">(</span><span class="va">sceptre_object</span><span class="op">)</span></span></code></pre></div>
<div class="figure" style="text-align: center">
<img src="grna_count_distribution.png" alt="Histograms of the gRNA count distributions" width="650px"><p class="caption">
Histograms of the gRNA count distributions
</p>
</div>
<p>The gRNAs display fairly bimodal count distributions. Consider, for
example, <code>gRNA_CCTGACA</code> (top left corner). This gRNA exhibits
a UMI count of <span class="math inline">\(\leq 1\)</span> or <span class="math inline">\(\geq 5\)</span> in most cells and a UMI count of
2, 3, or 4 in only a handful of cells. The vast majority of cells with a
UMI count of 1 likely do not actually contain <code>gRNA_CCTGACA</code>.
This is an example of “background contamination,” the phenomenon by
which gRNA transcripts sometimes map to cells that do not contain the
corresponding gRNA.</p>
<p><strong><code>sceptre</code> provides three methods for assigning
gRNAs to cells (the “mixture method,” the “maximum method,” and the
“thresholding method”)</strong>, all of which account for background
contamination. The default method for high-MOI data is the “mixture
method.” The gRNA counts are regressed onto the (unobserved) gRNA
presence/absence indicator and the cell-specific covariates (e.g.,
<code>grna_n_umis</code>, <code>batch</code>) via a latent variable
Poisson GLM. The fitted model yields the probability that each cell
contains the gRNA, and these probabilities are thresholded to assign the
gRNA to cells. The default method in low-MOI is the simpler “maximum”
approach: the gRNA that accounts for the greatest number of UMIs in a
given cell is assigned to that cell. A backup option in both low- and
high-MOI is the “thresholding” approach: a given gRNA is assigned to a
given cell if the UMI count of that gRNA in that cell exceeds some
integer threshold.</p>
<p>We carry out the gRNA assignment step via a call to the function
<code><a href="../reference/assign_grnas.html">assign_grnas()</a></code>. <code><a href="../reference/assign_grnas.html">assign_grnas()</a></code> takes arguments
<code>sceptre_object</code> (required) and <code>method</code>
(optional); the latter argument can be set to <code>"mixture"</code>,
<code>"maximum"</code>, or <code>"thresholding"</code>. We parallelize
execution of <code><a href="../reference/assign_grnas.html">assign_grnas()</a></code> by setting
<code>parallel</code> to <code>TRUE</code>. (Windows users should remove
the <code>parallel = TRUE</code> argument, as parallel execution is not
yet configured for Windows.)</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sceptre_object</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/assign_grnas.html">assign_grnas</a></span><span class="op">(</span>sceptre_object <span class="op">=</span> <span class="va">sceptre_object</span>, parallel <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="va">sceptre_object</span><span class="op">)</span> <span class="co"># output suppressed for brevity</span></span></code></pre></div>
<p>We can call <code><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot()</a></code> on the resulting
<code>sceptre_object</code> to render a plot summarizing the output of
the gRNA-to-cell assignment step.</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">sceptre_object</span><span class="op">)</span></span></code></pre></div>
<div class="figure" style="text-align: center">
<img src="grna_count_assignments.png" alt="gRNA-to-cell assignments" width="650px"><p class="caption">
gRNA-to-cell assignments
</p>
</div>
<p>The top panel plots the gRNA-to-cell assignments of three randomly
selected gRNAs. In each plot the points represent cells; the vertical
axis indicates the UMI count of the gRNA in a given cell, and the
horizontal axis indicates whether the cell has been classified as
“perturbed” (i.e., it <em>contains</em> the gRNA) or unperturbed (i.e.,
it does <em>not contain</em> the gRNA). Perturbed (resp., unperturbed)
cells are shown in the left (resp., right) column. The bottom left panel
is a barplot of the number of cells to which each gRNA has been mapped.
Finally, the bottom right panel is a histogram of the number of gRNAs
contained in each cell. The mean number of gRNAs per cell — i.e., the
MOI — is displayed in purple text.</p>
<p>See <a href="https://timothy-barry.github.io/sceptre-book/assign-grnas.html" class="external-link">Chapter
3 of the manual</a> for more information about assigning gRNAs to
cells.</p>
</div>
<div class="section level2">
<h2 id="sec-sceptre_qc">4. Run quality control (optional)<a class="anchor" aria-label="anchor" href="#sec-sceptre_qc"></a>
</h2>
<p>The fourth step is to run quality control (QC). This step likewise
can be skipped, in which case QC is applied automatically using default
options. <strong><code>sceptre</code> implements two kinds of QC:
cellwise QC and pairwise QC. The former aims to remove low-quality
cells, while the latter aims to remove low-quality target-response
pairs.</strong></p>
<p>The cellwise QC that <code>sceptre</code> implements is standard in
single-cell analysis. Cells for which <code>response_n_nonzero</code>
(i.e., the number of expressed responses) or
<code>response_n_umis</code> (i.e., the number of response UMIs) is
extremely high or extremely low are removed. Likewise, cells for which
<code>response_p_mito</code> (i.e., the fraction of UMIs mapping to
mitochondrial genes) is excessively high are removed. Additionally, in
low-MOI, cells that contain multiple gRNAs (as determined during the
RNA-to-cell assignment step) are removed. Finally, users optionally can
provide a list of additional cells to remove.</p>
<p><code>sceptre</code> also implements QC at the level of the
target-response pair. For a given pair we define the “treatment cells”
as those that contain a gRNA targeting the given target. Next, we define
the “control cells” as the cells against which the treatment cells are
compared to carry out the differential expression test. We define the
“number of nonzero treatment cells” (<code>n_nonzero_trt</code>) as the
number of <em>treatment</em> cells with nonzero expression of the
response; similarly, we define the “number of nonzero control cells”
(<code>n_nonzero_cntrl</code>) as the number of <em>control</em> cells
with nonzero expression of the response. <code>sceptre</code> filters
out pairs for which <code>n_nonzero_trt</code> or
<code>n_nonzero_cntrl</code> falls below some threshold (by default
7).</p>
<p>We call the function <code><a href="../reference/run_qc.html">run_qc()</a></code> on the
<code>sceptre_object</code> to carry out cellwise and pairwise QC.
<code><a href="../reference/run_qc.html">run_qc()</a></code> has several optional arguments that control the
stringency of the various QC thresholds. For example, we set
<code>p_mito_threshold = 0.075</code>, which filters out cells whose
<code>response_p_mito</code> value exceeds 0.075. (The optional
arguments are set to reasonable defaults; the default for
<code>p_mito_threshold</code> is 0.2, for instance).</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sceptre_object</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/run_qc.html">run_qc</a></span><span class="op">(</span><span class="va">sceptre_object</span>, p_mito_threshold <span class="op">=</span> <span class="fl">0.075</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="va">sceptre_object</span><span class="op">)</span> <span class="co"># output suppressed for brevity</span></span></code></pre></div>
<p>We can visualize the output of the QC step by calling
<code><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot()</a></code> on the updated <code>sceptre_object</code>.</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">sceptre_object</span><span class="op">)</span></span></code></pre></div>
<div class="figure" style="text-align: center">
<img src="qc.png" alt="Cellwise and pairwise quality control" width="450px"><p class="caption">
Cellwise and pairwise quality control
</p>
</div>
<p>The top panel depicts the outcome of the cellwise QC. The various
cellwise QC filters (e.g., “N nonzero responses,” “N response UMIs,”
“Percent mito”, etc.) are shown on the horizontal axis, and the
percentage of cells removed due application of a given QC filter is
shown on the vertical axis. Note that a cell can be flagged by multiple
QC filters; for example, a cell might have an extremely high
<code>response_n_umi</code> value <em>and</em> an extremely high
<code>response_n_nonzero</code> value. Thus, the height of the “any
filter” bar (which indicates the percentage of cells removed due to
application of <em>any</em> filter) need not be equal to the sum of the
heights of the other bars. The bottom panel depicts the outcome of the
pairwise QC. Each point corresponds to a target-response pair; the
vertical axis (resp., horizontal axis) indicates the
<code>n_nonzero_trt</code> (resp., <code>n_nonzero_cntrl</code>) value
of that pair. Pairs for which <code>n_nonzero_trt</code> or
<code>n_nonzero_cntrl</code> fall below the threshold are removed (red),
while the remaining pairs are retained (green).</p>
<p>See <a href="https://timothy-barry.github.io/sceptre-book/run-qc.html" class="external-link">Chapter
4 of the manual</a> for more information about QC.</p>
</div>
<div class="section level2">
<h2 id="sec-sceptre_calibration_check">5. Run calibration check<a class="anchor" aria-label="anchor" href="#sec-sceptre_calibration_check"></a>
</h2>
<p>The fifth step is to run the calibration check. <strong>The
calibration check is an analysis that verifies that <code>sceptre</code>
controls the rate of false discoveries on the dataset under
analysis.</strong> The calibration check proceeds as follows. First,
negative control target-response pairs are constructed (automatically)
by coupling subsets of NT gRNAs to randomly selected responses.
Importantly, the negative control pairs are constructed in such a way
that they are similar to the discovery pairs, the difference being that
the negative control pairs are devoid of biological signal. Next,
<code>sceptre</code> is applied to analyze the negative control pairs.
Given that the negative control pairs are absent of signal,
<code>sceptre</code> should produce approximately uniformly distributed
p-values on the negative control pairs. Moreover, after an appropriate
multiple testing correction, <code>sceptre</code> should make zero (or
very few) discoveries on the negative control pairs. Verifying
calibration via the calibration check increases our confidence that the
discovery set that <code>sceptre</code> ultimately produces is
uncontaminated by excess false positives.</p>
<p>We run the calibration check by calling the function
<code><a href="../reference/run_calibration_check.html">run_calibration_check()</a></code> on the
<code>sceptre_object</code>.</p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sceptre_object</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/run_calibration_check.html">run_calibration_check</a></span><span class="op">(</span><span class="va">sceptre_object</span>, parallel <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="va">sceptre_object</span><span class="op">)</span> <span class="co"># output suppressed for brevity</span></span></code></pre></div>
<p>We can assess the outcome of the calibration check by calling
<code><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot()</a></code> on the resulting <code>sceptre_object</code>.</p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">sceptre_object</span><span class="op">)</span></span></code></pre></div>
<div class="figure" style="text-align: center">
<img src="calibration_check.png" alt="Calibration check results" width="650px"><p class="caption">
Calibration check results
</p>
</div>
<p>The visualization consists of four panels, which we describe
below.</p>
<ul>
<li><p>The upper left panel is a QQ plot of the p-values plotted on an
untransformed scale. The p-values should lie along the diagonal line,
indicating uniformity of the p-values in the <em>bulk</em> of the
distribution.</p></li>
<li><p>The upper right panel is a QQ plot of the p-values plotted on a
negative log-10 transformed scale. The p-values again should lie along
the diagonal line (with the majority of the p-values falling within the
gray confidence band), indicating uniformity of the p-values in the
<em>tail</em> of the distribution.</p></li>
<li><p>The lower left panel is a histogram of the estimated log-2 fold
changes. The histogram should be roughly symmetric and centered around
zero.</p></li>
<li><p>Finally, the bottom right panel is a text box displaying (i) the
number of false discoveries that <code>sceptre</code> has made on the
negative control data and (ii) the mean estimated log-fold change. The
number of false discoveries should be a small integer like zero, one,
two, or three, with zero being ideal. The mean estimated log-fold
change, meanwhile, should be a numeric value close to zero; a number in
the range [-0.1, 0.1] is adequate.</p></li>
</ul>
<p><code>sceptre</code> may not exhibit good calibration initially,
which is OK. See <a href="https://timothy-barry.github.io/sceptre-book/run-calibration-check.html" class="external-link">Chapter
5 of the manual</a> for more information about the calibration check
analysis, including concrete strategies for improving calibration.</p>
</div>
<div class="section level2">
<h2 id="sec-sceptre_run_power_check">6. Run power check (optional)<a class="anchor" aria-label="anchor" href="#sec-sceptre_run_power_check"></a>
</h2>
<p>The sixth step — which is optional — is to run the power check. The
power check involves applying <code>sceptre</code> to analyze the
positive control pairs. Given that the positive control pairs are known
to contain signal, <code>sceptre</code> should produce significant
(i.e., small) p-values on the positive control pairs. <strong>The power
check enables us to assess <code>sceptre</code>’s power (i.e., its
ability to detect true associations) on the dataset under
analysis.</strong> We run the power check by calling the function
<code><a href="../reference/run_power_check.html">run_power_check()</a></code> on the <code>sceptre_object</code>.</p>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sceptre_object</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/run_power_check.html">run_power_check</a></span><span class="op">(</span><span class="va">sceptre_object</span>, parallel <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="va">sceptre_object</span><span class="op">)</span> <span class="co"># output suppressed for brevity</span></span></code></pre></div>
<p>We can visualize the outcome of the power check by calling
<code><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot()</a></code> on the resulting <code>sceptre_object</code>.</p>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">sceptre_object</span><span class="op">)</span></span></code></pre></div>
<div class="figure" style="text-align: center">
<img src="power_check.png" alt="Power check results" width="550px"><p class="caption">
Power check results
</p>
</div>
<p>Each point in the plot corresponds to a target-response pair, with
positive control pairs in the left column and negative control pairs in
the right column. The vertical axis indicates the p-value of a given
pair; smaller (i.e., more significant) p-values are positioned higher
along this axis (p-values truncated at <span class="math inline">\(10^{-20}\)</span> for visualization). The positive
control p-values should be small, and in particular, smaller than the
negative control p-values. See <a href="https://timothy-barry.github.io/sceptre-book/run-power-check-and-discovery-analysis.html" class="external-link">Chapter
6 of the manual</a> for more information about the power check.</p>
</div>
<div class="section level2">
<h2 id="sec-sceptre_run_discovery_analysis">7. Run discovery analysis<a class="anchor" aria-label="anchor" href="#sec-sceptre_run_discovery_analysis"></a>
</h2>
<p>The seventh and penultimate step is to run the discovery analysis.
The discovery analysis entails applying <code>sceptre</code> to analyze
the discovery pairs. We run the discovery analysis by calling the
function <code><a href="../reference/run_discovery_analysis.html">run_discovery_analysis()</a></code>.</p>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sceptre_object</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/run_discovery_analysis.html">run_discovery_analysis</a></span><span class="op">(</span><span class="va">sceptre_object</span>, parallel <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="va">sceptre_object</span><span class="op">)</span> <span class="co"># output suppressed for brevity</span></span></code></pre></div>
<p>We can visualize the outcome of the discovery analysis by calling
<code><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot()</a></code> on the resulting <code>sceptre_object</code>.</p>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">sceptre_object</span><span class="op">)</span></span></code></pre></div>
<div class="figure" style="text-align: center">
<img src="discovery_analysis.png" alt="Discovery analysis results" width="650px"><p class="caption">
Discovery analysis results
</p>
</div>
<p>The visualization consists of four panels.</p>
<ul>
<li><p>The upper left plot superimposes the discovery p-values (blue) on
top of the negative control p-values (red) on an untransformed
scale.</p></li>
<li><p>The upper right plot is the same as the upper left plot, but the
scale is negative log-10 transformed. The discovery p-values should
trend above the diagonal line, indicating the presence of signal in the
discovery set. The horizontal dashed line indicates the multiple testing
threshold; discovery pairs whose p-value falls above this line are
called as significant.</p></li>
<li><p>The bottom left panel is a volcano plot of the p-values and log
fold changes of the discovery pairs. Each point corresponds to a pair;
the estimated log-2 fold change of the pair is plotted on the horizontal
axis, and the (negative log-10 transformed) p-value is plotted on the
vertical axis. The horizontal dashed line again indicates the multiple
testing threshold. Points above the dashed line (colored in purple) are
called as discoveries, while points below (colored in blue) are called
as insignificant.</p></li>
<li><p>The bottom right panel is a text box displaying the number of
discovery pairs called as significant.</p></li>
</ul>
<p>See <a href="https://timothy-barry.github.io/sceptre-book/run-power-check-and-discovery-analysis.html" class="external-link">Chapter
6 of the manual</a> for more information about the discovery
analysis.</p>
</div>
<div class="section level2">
<h2 id="sec-sceptre_write_outputs_to_directory">8. Write outputs to directory<a class="anchor" aria-label="anchor" href="#sec-sceptre_write_outputs_to_directory"></a>
</h2>
<p>The eighth and final step is to write the outputs of the analysis to
a directory on disk. We call the function
<code><a href="../reference/write_outputs_to_directory.html">write_outputs_to_directory()</a></code>, which takes as arguments a
<code>sceptre_object</code> and <code>directory</code>;
<code>directory</code> is a string indicating the location of the
directory in which to write the results contained within the
<code>sceptre_object</code>.</p>
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/write_outputs_to_directory.html">write_outputs_to_directory</a></span><span class="op">(</span></span>
<span>  sceptre_object <span class="op">=</span> <span class="va">sceptre_object</span>, </span>
<span>  directory <span class="op">=</span> <span class="st">"~/sceptre_outputs"</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p><code><a href="../reference/write_outputs_to_directory.html">write_outputs_to_directory()</a></code> writes several files to the
specified directory: a text-based summary of the analysis
(“analysis_summary.txt”), the various plots (“*.png”), and the
calibration check, power check, and discovery analysis results
(“results_run_calibration_check.rds,” “results_run_power_check.rds,” and
“results_run_discovery_analysis.rds,” respectively).</p>
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/list.files.html" class="external-link">list.files</a></span><span class="op">(</span><span class="st">"~/sceptre_outputs"</span><span class="op">)</span></span>
<span><span class="co">#&gt;  [1] "analysis_summary.txt"               "plot_assign_grnas.png"             </span></span>
<span><span class="co">#&gt;  [3] "plot_grna_count_distributions.png"  "plot_run_calibration_check.png"    </span></span>
<span><span class="co">#&gt;  [5] "plot_run_discovery_analysis.png"    "plot_run_power_check.png"          </span></span>
<span><span class="co">#&gt;  [7] "plot_run_qc.png"                    "results_run_calibration_check.rds" </span></span>
<span><span class="co">#&gt;  [9] "results_run_discovery_analysis.rds" "results_run_power_check.rds"</span></span></code></pre></div>
<p>We also can obtain the calibration check, power check, and discovery
analysis results in R via a call to the function
<code><a href="../reference/get_result.html">get_result()</a></code>, passing as arguments
<code>sceptre_object</code> and <code>analysis</code>, where the latter
is a string indicating the function whose results we are querying.</p>
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">result</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/get_result.html">get_result</a></span><span class="op">(</span><span class="va">sceptre_object</span>, analysis <span class="op">=</span> <span class="st">"run_discovery_analysis"</span><span class="op">)</span></span></code></pre></div>
<p>The variable <code>result</code> is a data frame, the rows of which
correspond to target-response pairs, and the columns of which are as
follows: <code>response_id</code>, <code>grna_target</code>,
<code>n_nonzero_trt</code>, <code>n_nonzero_cntrl</code>,
<code>pass_qc</code> (a <code>TRUE</code>/<code>FALSE</code> value
indicating whether the pair passes pairwise QC), <code>p_value</code>,
<code>log_2_fold_change</code>, and <code>significant</code> (a
<code>TRUE</code>/<code>FALSE</code> value indicating whether the pair
is called as significant).</p>
<div class="sourceCode" id="cb29"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">result</span><span class="op">)</span></span>
<span><span class="co">#&gt;        response_id      grna_target n_nonzero_trt n_nonzero_cntrl pass_qc</span></span>
<span><span class="co">#&gt; 1: ENSG00000147454 candidate_enh_19           987           38763    TRUE</span></span>
<span><span class="co">#&gt; 2: ENSG00000136938 candidate_enh_20           829           43056    TRUE</span></span>
<span><span class="co">#&gt; 3: ENSG00000165702 candidate_enh_23           421           17129    TRUE</span></span>
<span><span class="co">#&gt; 4: ENSG00000136929 candidate_enh_20           507           30997    TRUE</span></span>
<span><span class="co">#&gt; 5: ENSG00000135046 candidate_enh_25           438           32903    TRUE</span></span>
<span><span class="co">#&gt; 6: ENSG00000106992 candidate_enh_22           693           30172    TRUE</span></span>
<span><span class="co">#&gt;         p_value log_2_fold_change significant</span></span>
<span><span class="co">#&gt; 1: 4.132597e-39        -0.7299781        TRUE</span></span>
<span><span class="co">#&gt; 2: 3.674875e-37        -0.3377841        TRUE</span></span>
<span><span class="co">#&gt; 3: 2.743547e-31        -0.7788056        TRUE</span></span>
<span><span class="co">#&gt; 4: 7.498748e-24        -0.6062542        TRUE</span></span>
<span><span class="co">#&gt; 5: 1.112902e-18        -0.8483076        TRUE</span></span>
<span><span class="co">#&gt; 6: 1.142874e-14        -0.3438875        TRUE</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="further-reading">Further reading<a class="anchor" aria-label="anchor" href="#further-reading"></a>
</h2>
<p>Each step of the pipeline has its own dedicated chapter in the
manual. We encourage interested users to read these additional chapters
one-by-one to learn more about the various options available at each
step of the pipeline, starting with <a href="https://timothy-barry.github.io/sceptre-book/import-data.html" class="external-link">Chapter
1 of the manual</a>.</p>
<p>Also see complete worked examples for the low-MOI CRISPRko and
high-MOI CRISPRi data via <code><a href="../reference/sceptre.html">?sceptre</a></code>.</p>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>Developed by Timothy Barry, Eugene Katsevich.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.5.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
