<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Running sceptre on moderately sized data • sceptre</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/lumen/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Running sceptre on moderately sized data">
<meta property="og:description" content="sceptre">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">sceptre</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.0.1</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../reference/index.html">Function reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Tutorials
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/sceptre-at-scale.html">At-scale</a>
    </li>
    <li>
      <a href="../articles/sceptre-on-moderately-sized-data.html">Moderately-sized data</a>
    </li>
    <li>
      <a href="../articles/sceptre-small-example.html">Demo</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><script src="sceptre-on-moderately-sized-data_files/accessible-code-block-0.0.1/empty-anchor.js"></script><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Running sceptre on moderately sized data</h1>
            
      
      
      <div class="hidden name"><code>sceptre-on-moderately-sized-data.Rmd</code></div>

    </div>

    
    
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://dplyr.tidyverse.org">dplyr</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">sceptre</span><span class="op">)</span></code></pre></div>
<p>This tutorial shows how to run <code>sceptre</code> on a dataset of moderate size. The data should fit into memory and be small enough such that the analysis can run on a single machine (as opposed to a computer cluster). To keep the size of the package small, we generate synthetic data rather than use real data.</p>
<div id="obtain-or-generate-the-raw-data" class="section level2">
<h2 class="hasAnchor">
<a href="#obtain-or-generate-the-raw-data" class="anchor"></a>Obtain (or generate) the raw data</h2>
<p>First, we generate a gene-by-cell expression matrix and a binary gRNA-by-cell perturbation matrix. We use 40 genes, 50 gRNAs, and 5000 cells. Genes and gRNAs are put in rows, and cells are put in columns, as is standard in single-cell analysis.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/Random.html">set.seed</a></span><span class="op">(</span><span class="fl">4</span><span class="op">)</span>
<span class="co"># n genes, gRNAs, and cells</span>
<span class="va">n_genes</span> <span class="op">&lt;-</span> <span class="fl">40</span>
<span class="va">n_gRNAs</span> <span class="op">&lt;-</span> <span class="fl">50</span>
<span class="va">n_cells</span> <span class="op">&lt;-</span> <span class="fl">5000</span>
<span class="co"># gene expression matrix </span>
<span class="va">expression_matrix</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html">replicate</a></span><span class="op">(</span>n <span class="op">=</span> <span class="va">n_genes</span>, <span class="fu"><a href="https://rdrr.io/r/stats/Poisson.html">rpois</a></span><span class="op">(</span><span class="va">n_cells</span>, <span class="fl">1</span><span class="op">)</span><span class="op">)</span> <span class="op">%&gt;%</span> <span class="fu"><a href="https://rdrr.io/r/base/t.html">t</a></span><span class="op">(</span><span class="op">)</span>
<span class="co"># perturbation matrix</span>
<span class="va">perturbation_matrix</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html">replicate</a></span><span class="op">(</span>n <span class="op">=</span> <span class="va">n_gRNAs</span>, <span class="fu"><a href="https://rdrr.io/r/stats/Binomial.html">rbinom</a></span><span class="op">(</span><span class="va">n_cells</span>, <span class="fl">1</span>, <span class="fl">0.05</span><span class="op">)</span><span class="op">)</span> <span class="op">%&gt;%</span> <span class="fu"><a href="https://rdrr.io/r/base/t.html">t</a></span><span class="op">(</span><span class="op">)</span></code></pre></div>
<p>The “expression” and “perturbation” matrices must have row names corresponding to the genes and gRNAs, respectively. We assign simple names to each.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/row.names.html">row.names</a></span><span class="op">(</span><span class="va">expression_matrix</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="st">"gene"</span>, <span class="fu"><a href="https://rdrr.io/r/base/seq.html">seq</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fu"><a href="https://rdrr.io/r/base/nrow.html">nrow</a></span><span class="op">(</span><span class="va">expression_matrix</span><span class="op">)</span><span class="op">)</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/row.names.html">row.names</a></span><span class="op">(</span><span class="va">perturbation_matrix</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="st">"gRNA"</span>, <span class="fu"><a href="https://rdrr.io/r/base/seq.html">seq</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fu"><a href="https://rdrr.io/r/base/nrow.html">nrow</a></span><span class="op">(</span><span class="va">perturbation_matrix</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<p>Finally, for each cell, we simulate its sequencing batch: “batch 1” or “batch 2.”</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">batch</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/sample.html">sample</a></span><span class="op">(</span>x <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"batch1"</span>, <span class="st">"batch2"</span><span class="op">)</span>, size <span class="op">=</span> <span class="va">n_cells</span>, replace <span class="op">=</span> <span class="cn">TRUE</span>, prob <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">0.5</span>, <span class="fl">0.5</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<p>On a real dataset we would perform quality control at this stage. For example, we would remove cells with 0 perturbations. We recommend doing minimal quality control, as <code>sceptre</code> is robust and can handle unusual gene expression distributions and misspecified expression models.</p>
</div>
<div id="compute-the-cell-wise-covariate-matrix" class="section level2">
<h2 class="hasAnchor">
<a href="#compute-the-cell-wise-covariate-matrix" class="anchor"></a>Compute the cell-wise covariate matrix</h2>
<p>Next, we compute the cell-specific covariate matrix. The following covariates are reasonable to include:</p>
<ol style="list-style-type: decimal">
<li>sequencing batch</li>
<li>percentage of UMIs that map to mitochondrial genes</li>
<li>log-transformed mRNA UMI count</li>
<li>log-transformed number of genes expressed in cell</li>
<li>log-transformed number of perturbations detected in cell, or log-transfored gRNA UMI count</li>
</ol>
<p>In this simple tutorial we will use (1) and (3). First, we compute the log-transformed mRNA UMI counts.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">lg_mRNA_lib_size</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/Log.html">log</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/colSums.html">colSums</a></span><span class="op">(</span><span class="va">expression_matrix</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<p>Next, we create the cell-specific covariate matrix, putting cells in rows and covariates in columns.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">covariate_matrix</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span><span class="op">(</span>lg_mRNA_lib_size <span class="op">=</span> <span class="va">lg_mRNA_lib_size</span>,
                               batch <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html">factor</a></span><span class="op">(</span><span class="va">batch</span><span class="op">)</span><span class="op">)</span></code></pre></div>
</div>
<div id="select-the-genes-and-grnas-to-include-in-the-analysis" class="section level2">
<h2 class="hasAnchor">
<a href="#select-the-genes-and-grnas-to-include-in-the-analysis" class="anchor"></a>Select the genes and gRNAs to include in the analysis</h2>
<p>After this, we select the gene-gRNA pairs to analyze. If we were conducting a <em>cis</em> analysis, we might pair each gene with all gRNAs targeting within 1Mb of the gene’s transcription start site. For simplicity, we will randomly sample 90 gene-gRNA pairs. The data frame that we create, <code>gene_gRNA_pairs</code>, should have one column called <code>gene_id</code> and a second column called <code>gRNA_id</code>.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">gene_gRNA_pairs</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/expand.grid.html">expand.grid</a></span><span class="op">(</span>gene_id <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/row.names.html">row.names</a></span><span class="op">(</span><span class="va">expression_matrix</span><span class="op">)</span>,
                               gRNA_id <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/row.names.html">row.names</a></span><span class="op">(</span><span class="va">perturbation_matrix</span><span class="op">)</span><span class="op">)</span> <span class="op">%&gt;%</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/slice.html">slice_sample</a></span><span class="op">(</span>n <span class="op">=</span> <span class="fl">90</span><span class="op">)</span></code></pre></div>
</div>
<div id="run-sceptre" class="section level2">
<h2 class="hasAnchor">
<a href="#run-sceptre" class="anchor"></a>Run <code>sceptre</code>
</h2>
<p>The final step is to run <code>sceptre</code>. We use the function <code>run_sceptre_in_memory</code>, which, as the name suggests, runs <code>sceptre</code> on in-memory expression and perturbation matrices. Notice that we did not “normalize” the gene expression data at any step; <code>sceptre</code> operates on raw UMI count matrices.</p>
<p><code>run_sceptre_in_memory</code> takes several arguments, which we summarize below.</p>
<ul>
<li><p><code>storage_dir</code>: the name of a directory in which to store the intermediate computations and final results.</p></li>
<li><p><code>expression_matrix</code>: the gene-by-cell expression data, represented as an R matrix. The names of the rows should be the gene names.</p></li>
<li><p><code>perturbation_matrix</code>: the gRNA-by-cell perturbation data, represented as an R matrix. The names of the rows should be the gRNA names.</p></li>
<li><p><code>covariate_matrix</code>: the matrix of cell-specific covariates, represented as a data frame. The rows correspond to cells, and the columns to the covariates.</p></li>
<li><p><code>gene_gRNA_pairs</code>: a data frame storing the gene-gRNA pairs to analyze. The data frame must contain columns named <code>gene_id</code> and <code>gRNA_id</code>, with entries corresponding to the row names of <code>expression_matrix</code> and <code>perturbation_matrix</code>.</p></li>
<li><p><code>side</code>: the sidedness of the test: one of “left,” “right,” and “both.” “left” is appropriate when perturbations are expected to decrease the expression of genes.</p></li>
<li><p><code>pod_sizes</code>: an integer vector giving the size of the “gene”, “gRNA”, and “pair” pods. <code>run_sceptre_in_memory</code> groups genes, gRNAs, and gene-gRNA pairs into distinct “pods” and runs computations on these pods in parallel. Smaller pod sizes give rise to greater parallelization. <code>pod_sizes</code> should be an integer vector of length three, with values giving the pod sizes and names “gene,” “gRNA,” and “pair.”</p></li>
<li><p><code>regularization_amount</code>: the amount of regularization to apply to the estimated negative binomial size parameters, where 0 corresponds to no regularization at all. The default option is 0.1.</p></li>
<li><p><code>B</code>: the number of random samples to draw in the conditional randomization test. The default option is 500.</p></li>
<li><p><code>seed</code>: seed to pass to the random number generator. The default value is 4.</p></li>
</ul>
<p>We define <code>storage_dir</code>, <code>side</code>, and <code>pod_sizes</code>, which we have not yet set. We will use the default options for the remaining parameters.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">storage_dir</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/tempfile.html">tempdir</a></span><span class="op">(</span><span class="op">)</span>
<span class="va">side</span> <span class="op">&lt;-</span> <span class="st">"left"</span>
<span class="va">pod_sizes</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span>gene <span class="op">=</span> <span class="fl">10</span>, gRNA <span class="op">=</span> <span class="fl">10</span>, pair <span class="op">=</span> <span class="fl">15</span><span class="op">)</span></code></pre></div>
<p>We have set <code>storage_dir</code> to <code>tempdir</code> for the sake of portability; in general, you should choose a more meaningful directory to store the intermediate computations and final results.</p>
<p>Finally, we call the function.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">result</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/run_sceptre_in_memory.html">run_sceptre_in_memory</a></span><span class="op">(</span><span class="va">storage_dir</span>,
                     <span class="va">expression_matrix</span>,
                     <span class="va">perturbation_matrix</span>,
                     <span class="va">covariate_matrix</span>,
                     <span class="va">gene_gRNA_pairs</span>,
                     <span class="va">side</span>,
                     <span class="va">pod_sizes</span><span class="op">)</span>
<span class="co">#&gt; [1] "Note: check the `log` subdirectory of the storage directory /var/folders/h6/pt59hmwd7rqb3jqfzwh0hlth0000gn/T//RtmpHaKZqH for updates!"</span></code></pre></div>
<p>As the function runs, we can check the <code>logs</code> subdirectory of the storage directory to monitor progress.</p>
<pre><code># in terminal
cd /var/folders/h6/pt59hmwd7rqb3jqfzwh0hlth0000gn/T//RtmpUfO717 # update path
cd logs</code></pre>
<p>This function takes about 5 minutes to execute on a Macbook Pro laptop with four Intel Core i5 processors.</p>
</div>
<div id="interpreting-the-results" class="section level2">
<h2 class="hasAnchor">
<a href="#interpreting-the-results" class="anchor"></a>Interpreting the results</h2>
<p>Let’s examine the results (which also have been written to the <code>results</code> subdirectory of the storage directory).</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">result</span><span class="op">)</span>
<span class="co">#&gt;   gene_id gRNA_id   p_value skew_t_fit_success         xi     omega      alpha</span>
<span class="co">#&gt; 1   gene3  gRNA37 0.4122104               TRUE -0.4928713 0.9804951  0.5709883</span>
<span class="co">#&gt; 2  gene22   gRNA8 0.1459034               TRUE -0.2202602 0.8361726  0.1272744</span>
<span class="co">#&gt; 3  gene12   gRNA6 0.1554211               TRUE -0.2156716 0.8643593  0.2175973</span>
<span class="co">#&gt; 4  gene27  gRNA27 0.6319403               TRUE -0.3430067 0.8727603  0.3304092</span>
<span class="co">#&gt; 5  gene35  gRNA36 0.7990196               TRUE -0.4294733 0.9114994  0.4709964</span>
<span class="co">#&gt; 6  gene15  gRNA11 0.4710990               TRUE  0.2001623 0.8559170 -0.2887997</span>
<span class="co">#&gt;         nu     z_value n_successful_resamples</span>
<span class="co">#&gt; 1 6.333383 -0.30240352                    500</span>
<span class="co">#&gt; 2 5.118402 -1.09854699                    499</span>
<span class="co">#&gt; 3 5.293119 -1.00052175                    500</span>
<span class="co">#&gt; 4 5.241553  0.18927850                    500</span>
<span class="co">#&gt; 5 6.349938  0.68567762                    500</span>
<span class="co">#&gt; 6 4.840431 -0.06262772                    499</span></code></pre></div>
<p>The data frame contains the following columns:</p>
<ul>
<li><p><code>gene_id</code>: the ID of the gene</p></li>
<li><p><code>gRNA_id</code>: the ID of the gRNA</p></li>
<li><p><code>p_value</code>: the gRNA-to-gene <em>p</em>-value</p></li>
<li><p><code>skew_t_fit_success</code>: boolean indicating whether the skew-t fit was successful (if not, the empirical conditional randomization test <em>p</em>-value is returned)</p></li>
<li><p><code>xi</code>, <code>omega</code>, <code>alpha</code>, <code>nu</code>: parameters of the fitted skew-t distribution</p></li>
<li><p><code>z_value</code>: the z-score of the negative binomial regression on the observed (i.e., non-resampled) data. The sign <code>z_value</code> gives the direction of the effect.</p></li>
<li><p><code>n_successful_resamples</code>: the number of resamples in the conditional randomization test that were succesful</p></li>
</ul>
<p>We might want to apply a Benjamini-Hochberg correction to account for multiple hypothesis testing, but be sure to proceed with caution here.</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">result</span><span class="op">$</span><span class="va">p_value_adj</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/p.adjust.html">p.adjust</a></span><span class="op">(</span>p <span class="op">=</span> <span class="va">result</span><span class="op">$</span><span class="va">p_value</span>, method <span class="op">=</span> <span class="st">"BH"</span><span class="op">)</span></code></pre></div>
<p>At this point we have finished our analysis!</p>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p>Developed by <a href="https://timothy-barry.github.io/">Timothy Barry</a>, Eugene Katsevich.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
