<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="sceptre">
<title>Set analysis parameters • sceptre</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.2.2/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.2.2/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Set analysis parameters">
<meta property="og:description" content="sceptre">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">sceptre</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.0.4</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="../articles/sceptre.html">Get started</a>
</li>
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="active nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="../articles/assign-grnas.html">Assign gRNAs</a>
    <a class="dropdown-item" href="../articles/import-data.html">Import data</a>
    <a class="dropdown-item" href="../articles/run-calibration-check.html">Run calibration check</a>
    <a class="dropdown-item" href="../articles/run-qc.html">Run quality control</a>
    <a class="dropdown-item" href="../articles/set-analysis-parameters.html">Set analysis parameters</a>
    <a class="dropdown-item" href="../articles/troubleshooting-and-faq.html">Troubleshooting and frequently asked questions</a>
  </div>
</li>
<li class="nav-item">
  <a class="nav-link" href="../news/index.html">Changelog</a>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav">
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/Katsevich-Lab/sceptre/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>

    
  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>Set analysis parameters</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/Katsevich-Lab/sceptre/blob/HEAD/vignettes/set-analysis-parameters.Rmd" class="external-link"><code>vignettes/set-analysis-parameters.Rmd</code></a></small>
      <div class="d-none name"><code>set-analysis-parameters.Rmd</code></div>
    </div>

    
    
<p>This vignette explicates the “set analysis parameters” step of the
<code>sceptre</code> pipeline. The purpose of this step is to select the
parameters that govern how the statistical analysis is to be carried
out.</p>
<p><img src="pipeline_schematic_step_2.png" width="650px" style="display: block; margin: auto;"></p>
<p>We begin by loading the <code>sceptre</code> package.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/Katsevich-Lab/sceptre" class="external-link">sceptre</a></span><span class="op">)</span></span></code></pre></div>
<p>Throughout, we examine two example datasets. First, we examine the
example high-MOI CRISPRi screen of candidate enhancers in K562 cells.
This is the example dataset that we used in the Get Started
(<code><a href="../articles/sceptre.html">vignette("sceptre")</a></code>) and Import Data
(<code><a href="../articles/import-data.html">vignette("import-data")</a></code>) vignettes. We initialize a
<code>sceptre_object</code> called <code>sceptre_object_highmoi</code>
to represent these data.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">highmoi_example_data</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">grna_target_data_frame_highmoi</span><span class="op">)</span></span>
<span><span class="va">sceptre_object_highmoi</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/import_data.html">import_data</a></span><span class="op">(</span>response_matrix <span class="op">=</span> <span class="va">highmoi_example_data</span><span class="op">$</span><span class="va">response_matrix</span>,</span>
<span>                                      grna_matrix <span class="op">=</span> <span class="va">highmoi_example_data</span><span class="op">$</span><span class="va">grna_matrix</span>,</span>
<span>                                      grna_target_data_frame <span class="op">=</span> <span class="va">grna_target_data_frame_highmoi</span>,</span>
<span>                                      moi <span class="op">=</span> <span class="st">"high"</span>,</span>
<span>                                      extra_covariates <span class="op">=</span> <span class="va">highmoi_example_data</span><span class="op">$</span><span class="va">extra_covariates</span>,</span>
<span>                                      response_names <span class="op">=</span> <span class="va">highmoi_example_data</span><span class="op">$</span><span class="va">gene_names</span><span class="op">)</span></span></code></pre></div>
<p>We additionally examine an example low-MOI CRISPRko screen of gene
transcription start sites in immune cells. We initialize a
<code>sceptre_object</code> called <code>sceptre_object_lowmoi</code> to
store this latter dataset.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sceptre_object_lowmoi</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/import_data.html">import_data</a></span><span class="op">(</span>response_matrix <span class="op">=</span> <span class="va">lowmoi_example_data</span><span class="op">$</span><span class="va">response_matrix</span>,</span>
<span>                                     grna_matrix <span class="op">=</span> <span class="va">lowmoi_example_data</span><span class="op">$</span><span class="va">grna_matrix</span>,</span>
<span>                                     extra_covariates <span class="op">=</span> <span class="va">lowmoi_example_data</span><span class="op">$</span><span class="va">extra_covariates</span>,</span>
<span>                                     grna_target_data_frame <span class="op">=</span> <span class="va">lowmoi_example_data</span><span class="op">$</span><span class="va">grna_target_data_frame</span>,</span>
<span>                                     moi <span class="op">=</span> <span class="st">"low"</span><span class="op">)</span></span></code></pre></div>
<p>We print the details of the CRISPRko dataset below.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sceptre_object_lowmoi</span></span>
<span><span class="co">#&gt; An object of class <span style="color: #0000BB;">sceptre_object</span>.</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Attributes of the data:</span></span>
<span><span class="co">#&gt;  • <span style="color: #0000BB;">20729</span> cells</span></span>
<span><span class="co">#&gt;  • <span style="color: #0000BB;">299</span> responses</span></span>
<span><span class="co">#&gt;  • <span style="color: #0000BB;">Low</span> multiplicity-of-infection </span></span>
<span><span class="co">#&gt;  • <span style="color: #0000BB;">101</span> targeting gRNAs (distributed across <span style="color: #0000BB;">26</span> targets) </span></span>
<span><span class="co">#&gt;  • <span style="color: #0000BB;">9</span> non-targeting gRNAs </span></span>
<span><span class="co">#&gt;  • <span style="color: #0000BB;">6</span> covariates (bio_rep, grna_n_nonzero, grna_n_umis, response_n_nonzero, response_n_umis, response_p_mito)</span></span></code></pre></div>
<p>The CRISPR perturbations target gene transcription start sites. Each
gene is targeted by three or four gRNAs, and nine gRNAs in the dataset
are non-targeting. The “responses” are genes. We print a few rows of the
gRNA target data frame corresponding to this dataset below.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">lowmoi_example_data</span><span class="op">$</span><span class="va">grna_target_data_frame</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">8</span>,<span class="op">]</span></span>
<span><span class="co">#&gt;   grna_id grna_target</span></span>
<span><span class="co">#&gt; 1  ATF2g1        ATF2</span></span>
<span><span class="co">#&gt; 2  ATF2g2        ATF2</span></span>
<span><span class="co">#&gt; 3  ATF2g3        ATF2</span></span>
<span><span class="co">#&gt; 4  ATF2g4        ATF2</span></span>
<span><span class="co">#&gt; 5  BRD4g1        BRD4</span></span>
<span><span class="co">#&gt; 6  BRD4g2        BRD4</span></span>
<span><span class="co">#&gt; 7  BRD4g3        BRD4</span></span>
<span><span class="co">#&gt; 8  BRD4g4        BRD4</span></span></code></pre></div>
<p>Our goal in analyzing the CRISPRko dataset is to dissect gene
regulatory networks by linking perturbations of genes to changes in the
expression of other genes.</p>
<p>We set the analysis parameters by calling the function
<code><a href="../reference/set_analysis_parameters.html">set_analysis_parameters()</a></code> on the
<code>sceptre_object</code>. <code><a href="../reference/set_analysis_parameters.html">set_analysis_parameters()</a></code>
takes several arguments: <code>positive_control_pairs</code>,
<code>discovery_pairs</code>, <code>side</code>,
<code>grna_grouping_strategy</code>, <code>control_group</code>,
<code>resampling_mechanism</code>, <code>formula_object</code>,
<code>multiple_testing_method</code>,
<code>multiple_testing_alpha</code>, and
<code>fit_parametric_curve</code>. The only required argument among
these is <code>discovery_pairs</code>. We describe these arguments
below.</p>
<div class="section level2">
<h2 id="positive-control-pairs">Positive control pairs<a class="anchor" aria-label="anchor" href="#positive-control-pairs"></a>
</h2>
<p>Positive control pairs are target-response pairs for which we know
(or have strong reason to believe) that there is a regulatory
relationship between the target and the response. We can use positive
control pairs to verify that <code>sceptre</code> (or any association
testing method for that matter) is sensitive (i.e., capable of detecting
true associations) on the dataset under analysis. We use the function
<code><a href="../reference/construct_positive_control_pairs.html">construct_positive_control_pairs()</a></code> to construct the
positive control pairs. <code><a href="../reference/construct_positive_control_pairs.html">construct_positive_control_pairs()</a></code>
takes as an argument a <code>sceptre_object</code> and returns a data
frame with columns <code>grna_target</code> and
<code>response_id</code>, where gRNA targets and response IDs with
matching names are paired. We call
<code><a href="../reference/construct_positive_control_pairs.html">construct_positive_control_pairs()</a></code> on the high-MOI CRISPRi
dataset and the low-MOI CRISPRko dataset. In both cases the positive
control set consists of transcription start sites paired to the gene
regulated by those transcription start sites.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">positive_control_pairs_highmoi</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/construct_positive_control_pairs.html">construct_positive_control_pairs</a></span><span class="op">(</span><span class="va">sceptre_object_highmoi</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">positive_control_pairs_highmoi</span><span class="op">)</span> <span class="co"># high MOI CRISPRi dataset</span></span>
<span><span class="co">#&gt;       grna_target     response_id</span></span>
<span><span class="co">#&gt; 1 ENSG00000069482 ENSG00000069482</span></span>
<span><span class="co">#&gt; 2 ENSG00000100316 ENSG00000100316</span></span>
<span><span class="co">#&gt; 3 ENSG00000104131 ENSG00000104131</span></span>
<span><span class="co">#&gt; 4 ENSG00000122026 ENSG00000122026</span></span>
<span><span class="co">#&gt; 5 ENSG00000135821 ENSG00000135821</span></span>
<span><span class="co">#&gt; 6 ENSG00000147669 ENSG00000147669</span></span></code></pre></div>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">positive_control_pairs_lowmoi</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/construct_positive_control_pairs.html">construct_positive_control_pairs</a></span><span class="op">(</span><span class="va">sceptre_object_lowmoi</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">positive_control_pairs_lowmoi</span><span class="op">)</span> <span class="co"># low MOI CRISPRko dataset</span></span>
<span><span class="co">#&gt;   grna_target response_id</span></span>
<span><span class="co">#&gt; 1       CMTM6       CMTM6</span></span>
<span><span class="co">#&gt; 2      IFNGR2      IFNGR2</span></span>
<span><span class="co">#&gt; 3        JAK2        JAK2</span></span>
<span><span class="co">#&gt; 4      NFKBIA      NFKBIA</span></span>
<span><span class="co">#&gt; 5       STAT1       STAT1</span></span>
<span><span class="co">#&gt; 6       STAT2       STAT2</span></span></code></pre></div>
<p>Positive control pairs need not consist exclusively of gene
transcription start sites paired to target genes. For example,
enhancer-gene links that have been validated previously (through, e.g.,
arrayed CRISPR screens) also can serve as positive control pairs. Users
manually can append additional positive control pairs to the positive
control pair data frame via a call to <code><a href="https://rdrr.io/r/base/cbind.html" class="external-link">rbind()</a></code>. For example,
suppose we know on the high-MOI CRISPRi data that
<code>candidate_enh_1</code> regulates gene <code>ENSG00000069482</code>
and that <code>candidate_enh_10</code> regulates gene
<code>ENSG00000135821</code>. We can add these enhancer-gene links to
the positive control pair data frame as follows.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># construct data frame of additional positive control pairs</span></span>
<span><span class="va">additional_positive_control_pairs</span> <span class="op">&lt;-</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span>grna_target <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"candidate_enh_1"</span>, <span class="st">"candidate_enh_10"</span><span class="op">)</span>,</span>
<span>             response_id <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"ENSG00000069482"</span>, <span class="st">"ENSG00000135821"</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># append additional pairs to positive control data frame</span></span>
<span><span class="va">positive_control_pairs_highmoi_updated</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/cbind.html" class="external-link">rbind</a></span><span class="op">(</span><span class="va">positive_control_pairs_highmoi</span>,</span>
<span>                                                <span class="va">additional_positive_control_pairs</span><span class="op">)</span></span>
<span><span class="va">positive_control_pairs_highmoi_updated</span></span>
<span><span class="co">#&gt;         grna_target     response_id</span></span>
<span><span class="co">#&gt; 1   ENSG00000069482 ENSG00000069482</span></span>
<span><span class="co">#&gt; 2   ENSG00000100316 ENSG00000100316</span></span>
<span><span class="co">#&gt; 3   ENSG00000104131 ENSG00000104131</span></span>
<span><span class="co">#&gt; 4   ENSG00000122026 ENSG00000122026</span></span>
<span><span class="co">#&gt; 5   ENSG00000135821 ENSG00000135821</span></span>
<span><span class="co">#&gt; 6   ENSG00000147669 ENSG00000147669</span></span>
<span><span class="co">#&gt; 7   ENSG00000155380 ENSG00000155380</span></span>
<span><span class="co">#&gt; 8   ENSG00000176890 ENSG00000176890</span></span>
<span><span class="co">#&gt; 9   ENSG00000187840 ENSG00000187840</span></span>
<span><span class="co">#&gt; 10  ENSG00000196683 ENSG00000196683</span></span>
<span><span class="co">#&gt; 11  candidate_enh_1 ENSG00000069482</span></span>
<span><span class="co">#&gt; 12 candidate_enh_10 ENSG00000135821</span></span></code></pre></div>
<p>We then could supply
<code>positive_control_pairs_highmoi_updated</code> instead of
<code>positive_control_pairs_highmoi</code> to
<code><a href="../reference/set_analysis_parameters.html">set_analysis_parameters()</a></code>.</p>
</div>
<div class="section level2">
<h2 id="discovery-pairs">Discovery pairs<a class="anchor" aria-label="anchor" href="#discovery-pairs"></a>
</h2>
<p>Discovery pairs are target-response pairs whose association status we
do not know but seek to learn. Unlike positive control and negative
control pairs, which serve a mainly technical purpose, discovery pairs
are of primary scientific interest. Discovery pairs must be supplied to
<code><a href="../reference/set_analysis_parameters.html">set_analysis_parameters()</a></code>; all other arguments (aside from
<code>sceptre_object</code>) are optional. <code>sceptre</code> provides
two helper functions for constructing discovery pairs:
<code><a href="../reference/construct_cis_pairs.html">construct_cis_pairs()</a></code> and
<code><a href="../reference/construct_trans_pairs.html">construct_trans_pairs()</a></code>.</p>
<div class="section level4">
<h4 id="construct-cis-pairs">Construct <em>cis</em> pairs<a class="anchor" aria-label="anchor" href="#construct-cis-pairs"></a>
</h4>
<p><code><a href="../reference/construct_cis_pairs.html">construct_cis_pairs()</a></code> returns the set of target-response
pairs for which the target and response are located on the same
chromosome and in close physical proximity to one another.
<code><a href="../reference/construct_cis_pairs.html">construct_cis_pairs()</a></code> assumes that the columns
<code>chr</code>, <code>start</code>, and <code>stop</code> are present
within the <code>grna_target_data_frame</code>.
<code><a href="../reference/construct_cis_pairs.html">construct_cis_pairs()</a></code> also assumes that the responses are
genes as opposed to, say, proteins. <code><a href="../reference/construct_cis_pairs.html">construct_cis_pairs()</a></code>
is a useful pair constructor function for screens that aim to map
noncoding regulatory elements (e.g., enhancers, silencers, or noncoding
GWAS variants) to target genes in <em>cis</em>.
<code><a href="../reference/construct_cis_pairs.html">construct_cis_pairs()</a></code> takes several arguments:
<code>sceptre_object</code> (required) <code>distance_threshold</code>
(optional), <code>positive_control_pairs</code> (optional), and
<code>ref_genome</code> (optional). Each target is paired to the set of
genes within <code>distance_threshold</code> bases of that target. (The
default value of <code>distance_threshold</code> is 500,000 bases, or
half a megabase.) The <code>positive_control_pairs</code> data frame
optionally can be passed to <code><a href="../reference/construct_cis_pairs.html">construct_cis_pairs()</a></code>, in which
case the positive control targets (i.e., the entries within the
<code>grna_target</code> column of <code>positive_control_pairs</code>)
are excluded from the <em>cis</em> pairs. Finally,
<code>ref_genome</code> is a string indicating the reference genome that
<code>chr</code>, <code>start</code>, and <code>stop</code> are defined
with respect to. (The only reference genome currently available is
“10X_GRCh38_2020,” which is the GRCh38 reference genome that has shipped
with CellRanger since 2020.)</p>
<p>We use <code><a href="../reference/construct_cis_pairs.html">construct_cis_pairs()</a></code> to construct the discovery
pairs for the high-MOI CRISPRi dataset. We set
<code>distance_threshold</code> to <code>5e6</code> (i.e., 5 megabases)
so as to increase the number of pairs in the discovery set for
illustration purposes.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">discovery_pairs_highmoi</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/construct_cis_pairs.html">construct_cis_pairs</a></span><span class="op">(</span>sceptre_object <span class="op">=</span> <span class="va">sceptre_object_highmoi</span>,</span>
<span>                                               positive_control_pairs <span class="op">=</span> <span class="va">positive_control_pairs_highmoi</span>,</span>
<span>                                               distance_threshold <span class="op">=</span> <span class="fl">5e6</span><span class="op">)</span></span></code></pre></div>
<p><code>discovery_pairs_highmoi</code> is a data frame with columns
<code>grna_target</code> and <code>response_id</code>; each candidate
enhancer is mapped to the set of genes in close proximity to that
candidate enhancer.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">discovery_pairs_highmoi</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">10</span>, <span class="fl">20</span>, <span class="fl">30</span>, <span class="fl">40</span>, <span class="fl">50</span><span class="op">)</span>,<span class="op">]</span></span>
<span><span class="co">#&gt;        grna_target     response_id</span></span>
<span><span class="co">#&gt; 1  candidate_enh_1 ENSG00000174529</span></span>
<span><span class="co">#&gt; 10 candidate_enh_1 ENSG00000117280</span></span>
<span><span class="co">#&gt; 20 candidate_enh_2 ENSG00000143493</span></span>
<span><span class="co">#&gt; 30 candidate_enh_2 ENSG00000136643</span></span>
<span><span class="co">#&gt; 40 candidate_enh_3 ENSG00000142733</span></span>
<span><span class="co">#&gt; 50 candidate_enh_3 ENSG00000117748</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="construct-trans-pairs">Construct <em>trans</em> pairs<a class="anchor" aria-label="anchor" href="#construct-trans-pairs"></a>
</h4>
<p><code><a href="../reference/construct_trans_pairs.html">construct_trans_pairs()</a></code> returns the entire set of
possible target-response pairs. <code><a href="../reference/construct_trans_pairs.html">construct_trans_pairs()</a></code> is
a useful pair constructor function for analyses in which we seek to
conduct a <em>trans</em> analysis, testing each target against each
response. <code><a href="../reference/construct_trans_pairs.html">construct_trans_pairs()</a></code> takes as arguments
<code>sceptre_object</code> (required),
<code>positive_control_pairs</code> (optional),
<code>exclude_positive_control_pairs</code> (optional), and
<code>exclude_positive_control_targets</code> (optional). By default
<code><a href="../reference/construct_trans_pairs.html">construct_trans_pairs()</a></code> returns a data frame with columns
<code>grna_target</code> and <code>response_id</code>, where each gRNA
target is mapped to each response ID. If the
<code>positive_control_pairs</code> data frame is passed and
<code>exclude_positive_control_pairs</code> is set to <code>TRUE</code>,
then the positive control target-response pairs are excluded from the
<em>trans</em> pairs. Next, if <code>positive_control_pairs</code> is
passed and <code>exclude_positive_control_targets</code> is set to
<code>TRUE</code>, then <em>all</em> pairs containing a positive control
gRNA target are excluded from the <em>trans</em> pairs. (In this sense
setting <code>exclude_positive_control_targets</code> to
<code>TRUE</code> is stronger than setting
<code>exclude_positive_control_pairs</code> to <code>TRUE</code>.)</p>
<p>We use <code><a href="../reference/construct_trans_pairs.html">construct_trans_pairs()</a></code> to construct the
discovery set for the low-MOI CRISPRko dataset, excluding the positive
control pairs. The discovery set consists of the set of genes targeted
by one or more gRNAs paired to the set of <em>all other</em> genes.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">discovery_pairs_lowmoi</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/construct_trans_pairs.html">construct_trans_pairs</a></span><span class="op">(</span>sceptre_object <span class="op">=</span> <span class="va">sceptre_object_lowmoi</span>,</span>
<span>                                                positive_control_pairs <span class="op">=</span> <span class="va">positive_control_pairs_lowmoi</span>,</span>
<span>                                                exclude_positive_control_pairs <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">discovery_pairs_lowmoi</span><span class="op">)</span></span>
<span><span class="co">#&gt;   grna_target response_id</span></span>
<span><span class="co">#&gt; 1        ATF2       PCBP3</span></span>
<span><span class="co">#&gt; 2        BRD4       PCBP3</span></span>
<span><span class="co">#&gt; 3        CAV1       PCBP3</span></span>
<span><span class="co">#&gt; 4       CD274       PCBP3</span></span>
<span><span class="co">#&gt; 5        CD86       PCBP3</span></span>
<span><span class="co">#&gt; 6       CMTM6       PCBP3</span></span></code></pre></div>
<p>We also can use <code><a href="../reference/construct_trans_pairs.html">construct_trans_pairs()</a></code> to construct a
<em>trans</em> discovery set for the high-MOI, enhancer-targeting
CRISPRi dataset. To this end we call
<code><a href="../reference/construct_trans_pairs.html">construct_trans_pairs()</a></code>, setting
<code>exclude_positive_control_targets</code> to <code>TRUE</code> so as
to exclude all positive control gRNA targets from the <em>trans</em>
pairs.</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">discovery_pairs_highmoi_trans</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/construct_trans_pairs.html">construct_trans_pairs</a></span><span class="op">(</span>sceptre_object <span class="op">=</span> <span class="va">sceptre_object_highmoi</span>,</span>
<span>                                                       positive_control_pairs <span class="op">=</span> <span class="va">positive_control_pairs_highmoi</span>,</span>
<span>                                                       exclude_positive_control_targets <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<p>The resulting data frame, <code>discovery_pairs_highmoi_trans</code>,
maps each candidate enhancer to the entire set of genes.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">discovery_pairs_highmoi_trans</span><span class="op">)</span></span>
<span><span class="co">#&gt;        grna_target     response_id</span></span>
<span><span class="co">#&gt; 1  candidate_enh_1 ENSG00000069275</span></span>
<span><span class="co">#&gt; 2 candidate_enh_10 ENSG00000069275</span></span>
<span><span class="co">#&gt; 3 candidate_enh_11 ENSG00000069275</span></span>
<span><span class="co">#&gt; 4 candidate_enh_12 ENSG00000069275</span></span>
<span><span class="co">#&gt; 5 candidate_enh_13 ENSG00000069275</span></span>
<span><span class="co">#&gt; 6 candidate_enh_14 ENSG00000069275</span></span></code></pre></div>
<p>Some users may wish to run both <em>cis</em> and <em>trans</em>
analyses on their data. We recommend that such users carry out the
<code>sceptre</code> pipeline twice: once using a <em>cis</em> discovery
set and once using a <em>trans</em> discovery set.</p>
</div>
<div class="section level3">
<h3 id="reducing-the-multiplicity-burden">Reducing the multiplicity burden<a class="anchor" aria-label="anchor" href="#reducing-the-multiplicity-burden"></a>
</h3>
<p><code>sceptre</code> computes a p-value for each target-response pair
in the discovery set. These p-values are subjected to a multiple testing
correction procedure (by default, the Benjamini-Hochberg method) to
produce a discovery set that controls some notion of aggregate type-I
error (by default, the false discovery rate). In general it is good to
restrict one’s attention to pairs that are promising or scientifically
interesting. Including a large number of uninteresting or “null” pairs
(i.e., pairs for which there is no relationship between the target and
response) in the discovery set causes the power of the multiple testing
correction procedure to decrease, resulting in fewer discoveries.</p>
</div>
</div>
<div class="section level2">
<h2 id="side">Side<a class="anchor" aria-label="anchor" href="#side"></a>
</h2>
<p><code>sceptre</code> can run left-tailed, right-tailed, and
two-tailed tests of association. Left-tailed tests assess a
<em>decrease</em> in expression, right-tailed tests assess an
<em>increase</em> in expression, and two-tailed tests assess an increase
<em>or</em> decrease in expression. The parameter <code>side</code>
controls the sidedness of the test and can take values
<code>"left"</code>, <code>"right"</code>, or <code>"both"</code>.
(<code>"both"</code> indicates a two-tailed test.) Two-tailed tests
typically are the best choice for <em>trans</em> analyses, as the
direction of <em>trans</em> relationships generally is uncertain. For
<em>cis</em> analyses, on the other hand, the sidedness of the test
should be chosen on the basis of the type of genomic element targeted
and the CRISPR perturbation modality. The following table summarizes
whether a left- or right-tailed test is appropriate for a <em>cis</em>
analysis as a function of these variables.</p>
<table style="width:98%;" class="table">
<caption>Recommended sidedness of the test as a function of target
element and CRISPR modality for <em>cis</em> analyses.</caption>
<colgroup>
<col width="20%">
<col width="26%">
<col width="30%">
<col width="19%">
</colgroup>
<thead><tr class="header">
<th>Target element</th>
<th>CRISPR modality</th>
<th>Testing for</th>
<th>Sidedness</th>
</tr></thead>
<tbody>
<tr class="odd">
<td>Enhancer</td>
<td>CRISPRi or CRISPRko</td>
<td>Decrease in expression</td>
<td>Left</td>
</tr>
<tr class="even">
<td>Enhancer</td>
<td>CRISPRa</td>
<td>Increase in expression</td>
<td>Right</td>
</tr>
<tr class="odd">
<td>Silencer</td>
<td>CRISPRi or CRISPRko</td>
<td>Increase in expression</td>
<td>Right</td>
</tr>
<tr class="even">
<td>Silencer</td>
<td>CRISPRa</td>
<td>Decrease in expression</td>
<td>Left</td>
</tr>
</tbody>
</table>
</div>
<div class="section level2">
<h2 id="grna-grouping-strategy">gRNA grouping strategy<a class="anchor" aria-label="anchor" href="#grna-grouping-strategy"></a>
</h2>
<p>Multiple gRNAs typically are designed to target a given genomic
element. The parameter <code>grna_grouping_strategy</code> controls if
and how gRNAs that target the same genomic element are combined. The
default gRNA grouping strategy is <code>"union"</code>.
<code>"union"</code> constructs a “grouped gRNA” by combining all gRNAs
that target a given genomic element via a union operation; this “grouped
gRNA” is then tested for association against the responses to which the
element is paired. We illustrate this grouping strategy using an
example. Suppose that “gRNA 1” and “gRNA 2” target the same genomic
element. Suppose that “gRNA 1” is present in the cells indexed 3, 6, 9
and that “gRNA 2” is present in the cells indexed 1, 4, 6, 10. The
“grouped gRNA” formed combining “gRNA 1” and “gRNA 2” via the union
operation is defined to be present in the cells indexed 1, 3, 4, 6, 9,
10. (See schematic below.) This “grouped gRNA” is then tested against
responses as if it were a single gRNA.</p>
<div class="figure" style="text-align: center">
<img src="union_grouping.png" alt='Schematic of the `"union"` gRNA grouping strategy. Blue (resp., white) squares indicate cells in which the gRNA is present (resp., absent).' width="450px"><p class="caption">
Schematic of the <code>"union"</code> gRNA grouping strategy. Blue
(resp., white) squares indicate cells in which the gRNA is present
(resp., absent).
</p>
</div>
<p>The other available gRNA grouping strategy is
<code>"singleton"</code>. The <code>"singleton"</code> strategy involves
individually testing each gRNA targeting a given element against the
responses to which that element is paired. <code>"singleton"</code> may
be more appropriate than <code>"union"</code> for experiments in which
gRNA effectiveness is expected to vary considerably across gRNAs.
However, <code>"union"</code> is the default rather than
<code>"singleton"</code> because <code>"union"</code> tends to yield
more powerful results.</p>
</div>
<div class="section level2">
<h2 id="control-group">Control group<a class="anchor" aria-label="anchor" href="#control-group"></a>
</h2>
<p>The primary goal of <code>sceptre</code> is to test for association
between a given gRNA target and response. <code>control_group</code> is
a parameter that controls the set of cells used to carry out the test of
association. Suppose for simplicity that we are carrying out a singleton
analysis, i.e. suppose that we are testing for association between an
individual targeting gRNA and a response. We define the “treatment
group” as the set of cells that contain the given targeting gRNA, and we
define the “control group” as the set of cells against which the
treatment group is compared to conduct the association test. There are
two choices for the control group: the “complement set” and the
“non-targeting (NT) cells.” The complement set consists of the cells
that <em>do not</em> contain the given targeting gRNA; the NT cells, by
contrast, consist of the cells that contain a non-targeting gRNA. (See
schematic below.) <code>sceptre</code> tests for association between the
given targeting gRNA and the response by testing for differential
expression of the response across the treatment group and the control
group. Users can select the control group to use by passing either
<code>"complement"</code> or <code>"nt_cells"</code> to
<code>control_group</code>.</p>
<div class="figure" style="text-align: center">
<img src="control_vs_complement.png" alt='A schematic illustrating the difference between the "complement set" and "NT cells" control groups. The single-cell CRISPR screen experiment produces a pool of cells containing both targeting and non-targeting gRNAs (top). Suppose that we seek to test for association between a given targeting gRNA (e.g., gRNA $T_2$) and a given response. We first divide the cells into two groups: the treatment group (i.e., the cells containing gRNA $T_2$) and the control group. The control group either can consist of the cells not containing gRNA $T_2$ (the "complement set") or the cells containing a non-targeting gRNA (the "NT cells"). `sceptre` tests for differential expression of the given response across the treatment and control groups, yielding a p-value for the test of association between the given targeting gRNA and the given response.' width="760px"><p class="caption">
A schematic illustrating the difference between the “complement set” and
“NT cells” control groups. The single-cell CRISPR screen experiment
produces a pool of cells containing both targeting and non-targeting
gRNAs (top). Suppose that we seek to test for association between a
given targeting gRNA (e.g., gRNA <span class="math inline">\(T_2\)</span>) and a given response. We first
divide the cells into two groups: the treatment group (i.e., the cells
containing gRNA <span class="math inline">\(T_2\)</span>) and the
control group. The control group either can consist of the cells not
containing gRNA <span class="math inline">\(T_2\)</span> (the
“complement set”) or the cells containing a non-targeting gRNA (the “NT
cells”). <code>sceptre</code> tests for differential expression of the
given response across the treatment and control groups, yielding a
p-value for the test of association between the given targeting gRNA and
the given response.
</p>
</div>
<p>The choice of the control group should be made on the basis of the
MOI of the dataset and the type of genomic element being targeted. In
high-MOI screens each cell contains multiple gRNAs. Very few (if any)
cells contain exclusively NT gRNAs, and so the complement set is the
only choice for the control group. In low-MOI screens there is more
flexibility. The default choice for the control group is the NT cells,
as we seek to compare the effect of the targeting gRNA to that of a
“null” gRNA rather than to the average of the effects of all other gRNAs
introduced in the pooled screen. However, the complement set is a
reasonable choice for low-MOI enhancer-targeting screens, as gRNAs in
enhancer-targeting screens tend to exert small and local effects,
rendering the average effect over gRNAs in the complement set similar to
that of a “null” gRNA. Thus, we recommend trying both the complement set
and the NT cells for the control group in low-MOI enhancer-targeting
screens and selecting between these options on the basis of their
performance on the negative control and positive control data. The
decision tree below summarizes the control group that we recommend
selecting as a function of MOI and genomic element targeted.</p>
<div class="figure" style="text-align: center">
<img src="control_group_dec_tree.png" alt="A decision tree summarizing the control group to select as a function of the MOI of the data and the genomic element targeted. The only option in high-MOI is the complement set. In low-MOI the NT cells are the default option; however, we recommend trying both the complement set and the NT cells for low-MOI enhancer-targeting screens." width="350px"><p class="caption">
A decision tree summarizing the control group to select as a function of
the MOI of the data and the genomic element targeted. The only option in
high-MOI is the complement set. In low-MOI the NT cells are the default
option; however, we recommend trying both the complement set and the NT
cells for low-MOI enhancer-targeting screens.
</p>
</div>
</div>
<div class="section level2">
<h2 id="formula">Formula<a class="anchor" aria-label="anchor" href="#formula"></a>
</h2>
<p>The parameter <code>formula_object</code> specifies how
<code>sceptre</code> is to adjust for the cell-specific covariates.
<code>formula_object</code> is optional; if not provided,
<code><a href="../reference/set_analysis_parameters.html">set_analysis_parameters()</a></code> constructs a default
<code>formula_object</code>. The default <code>formula_object</code> is
formed by summing over all covariates and log-transforming the
count-based covariates. For example, the high-MOI CRISPRi data contain
the cell-specific covariates <code>response_n_nonzero</code>,
<code>response_n_umis</code>, <code>grna_n_nonzero</code>,
<code>grna_n_umis</code>, <code>response_p_mito</code>, and
<code>batch</code>. The first four of these covariates are count-based.
Thus, the default formula object for this dataset is as follows:</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/stats/formula.html" class="external-link">formula</a></span><span class="op">(</span><span class="op">~</span> <span class="fu"><a href="https://rdrr.io/r/base/Log.html" class="external-link">log</a></span><span class="op">(</span><span class="va">response_n_nonzero</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://rdrr.io/r/base/Log.html" class="external-link">log</a></span><span class="op">(</span><span class="va">response_n_umis</span><span class="op">)</span> <span class="op">+</span></span>
<span>          <span class="fu"><a href="https://rdrr.io/r/base/Log.html" class="external-link">log</a></span><span class="op">(</span><span class="va">grna_n_nonzero</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://rdrr.io/r/base/Log.html" class="external-link">log</a></span><span class="op">(</span><span class="va">grna_n_umis</span><span class="op">)</span> <span class="op">+</span> <span class="va">response_p_mito</span> <span class="op">+</span> <span class="va">batch</span><span class="op">)</span></span></code></pre></div>
<p>Similarly, the low-MOI CRISPRko data contain the covariates
<code>response_n_nonzero</code>, <code>response_n_umis</code>,
<code>grna_n_nonzero</code>, <code>grna_n_umis</code>,
<code>response_p_mito</code>, and <code>bio_rep</code>. The default
formula object for the CRISPRko dataset is as follows:</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/stats/formula.html" class="external-link">formula</a></span><span class="op">(</span><span class="op">~</span> <span class="fu"><a href="https://rdrr.io/r/base/Log.html" class="external-link">log</a></span><span class="op">(</span><span class="va">response_n_nonzero</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://rdrr.io/r/base/Log.html" class="external-link">log</a></span><span class="op">(</span><span class="va">response_n_umis</span><span class="op">)</span> <span class="op">+</span></span>
<span>          <span class="va">response_p_mito</span> <span class="op">+</span> <span class="va">batch</span><span class="op">)</span></span></code></pre></div>
<p>(The covariates <code>grna_n_nonzero</code> and
<code>grna_n_umis</code> by default are not included in the formula
object, as <code>grna_n_nonzero</code> and <code>grna_n_umis</code>
typically do not contain substantial information in low MOI after
quality control.)</p>
<p>Users can specify a custom formula by passing a formula object to the
argument <code>formula_object</code> in
<code><a href="../reference/set_analysis_parameters.html">set_analysis_parameters()</a></code>. User-specified formulas should
be defined in terms of the cell-specific covariates, which can be viewed
by evaluating the <code>sceptre_object</code> in the console.</p>
</div>
<div class="section level2">
<h2 id="resampling-mechanism">Resampling mechanism<a class="anchor" aria-label="anchor" href="#resampling-mechanism"></a>
</h2>
<p>The parameter <code>resampling_mechanism</code> controls the
inferential procedure that <code>sceptre</code> uses to compute a
p-value for a target-response pair. Suppose for simplicity that we are
carrying out a singleton gRNA analysis. There are three pieces of
information relevant to testing for association between a given gRNA and
response: (1) the vector of UMI counts of the response; (2) the “gRNA
indicator vector,” where a given entry of the vector is set to “1” if
the corresponding cell is part of the treatment group (i.e., it contains
the gRNA) and “0” if it is part of the control group; and (3) the matrix
of cell-specific covariates. First, <code>sceptre</code> regresses the
response vector onto the gRNA indicator vector and covariate matrix via
a negative binomial (NB) GLM. <code>sceptre</code> computes the z-score
<span class="math inline">\(z_\textrm{obs}\)</span> corresponding to a
test of the null hypothesis that the coefficient corresponding to the
gRNA indicator vector in the fitted GLM is zero. Next,
<code>sceptre</code> <em>resamples</em> the gRNA indicator vector
<em>B</em> times, where <em>B</em> is some large integer (e.g.,
<em>B</em> = 5000). (We discuss exactly how the gRNA indicator vector is
resampled below.) Finally, for each of the resampled gRNA indicator
vectors, <code>sceptre</code> recomputes the z-score, producing
<em>B</em> “null” z-scores <span class="math inline">\(\tilde{z_1},
\dots, \tilde{z_B}\)</span>. The observed z-score <span class="math inline">\(z_\textrm{obs}\)</span> is compared to the null
z-scores <span class="math inline">\(\tilde{z_1}, \dots,
\tilde{z_B}\)</span> to compute a p-value.</p>
<p><img src="sceptre_pt_1.png" width="650px" style="display: block; margin: auto;"></p>
<p><code>sceptre</code> provides two procedures for resampling the gRNA
indicator vector: permutations and conditional resampling (also known as
the conditional randomization test). The permutation procedure involves
randomly permuting the gRNA indicator vector <em>B</em> times. The
conditional resampling procedure is slightly more complex. Briefly, the
gRNA indicator vector is regressed onto the covariate matrix via
logistic regression, yielding an estimate for the probability that each
cell contains the gRNA as a function of the covariates. <em>B</em>
synthetic gRNA indicator vectors are then sampled according to these
estimated probabilities. The resampling mechanism can be specified by
passing either <code>permutations</code> or <code>crt</code> (for
conditional resampling) to the argument
<code>resampling_mechanism</code> in
<code><a href="../reference/set_analysis_parameters.html">set_analysis_parameters()</a></code>.</p>
<p>Permutations and conditional resampling offer distinct advantages and
disadvantages, although the differences can be subtle. The permutation
approach is faster and slightly better at handling sparsity, while the
conditional resampling approach is slightly better at handling
confounding due to cell-specific covariates (see table below). In
practice the two approaches often produce similar results; we recommend
that users try both, provided that they have the bandwidth and compute
to do so. The permutation (resp., conditional resampling) procedure is
the default option in low-MOI (resp., high-MOI). (Note that both options
are available in both MOI settings.)</p>
<table style="width:99%;" class="table">
<colgroup>
<col width="25%">
<col width="37%">
<col width="36%">
</colgroup>
<thead><tr class="header">
<th>Resampling procedure</th>
<th>Advantages</th>
<th>Disadvantages</th>
</tr></thead>
<tbody>
<tr class="odd">
<td>Permutations</td>
<td><ul>
<li><p>Faster</p></li>
<li><p>Better at handling sparsity</p></li>
</ul></td>
<td><ul>
<li>Worse at handling confounding</li>
</ul></td>
</tr>
<tr class="even">
<td>Conditional resampling</td>
<td><ul>
<li>Better at handling confounding</li>
</ul></td>
<td><ul>
<li><p>Slower</p></li>
<li><p>Worse at handling sparsity</p></li>
</ul></td>
</tr>
</tbody>
</table>
<p>Resampling forms the foundation of statistical inference within the
<code>sceptre</code> framework. In fact, “sceptre” is a loose acronym
for “<strong>s</strong>ingle <strong>c</strong>ell
<strong>p</strong>er<strong>t</strong>urbation analysis via
<strong>re</strong>sampling.” The resampling machinery of
<code>sceptre</code> is reasonably fast due to several algorithmic
advances, including the technique of resampling score statistics and a
novel sparsity-exploiting algorithm for computing GLM score tests.</p>
</div>
<div class="section level2">
<h2 id="fit-parametric-curve">Fit parametric curve<a class="anchor" aria-label="anchor" href="#fit-parametric-curve"></a>
</h2>
<p>As described above, for a given gRNA-response pair,
<code>sceptre</code> computes an “observed” z-score <span class="math inline">\(z_\textrm{obs}\)</span> and a set of “null”
z-scores <span class="math inline">\(\tilde{z_1}, \dots,
\tilde{z_B}\)</span> (regardless of whether
<code>resampling_mechanism</code> is set to <code>permutations</code> or
<code>crt</code>). By default, <code>sceptre</code> fits a parametric
density (specifically, a skew-normal density) to the distribution of
null z-scores and then computes a p-value by evaluating the tail
probability of the fitted density at the observed z-score <span class="math inline">\(z_\textrm{obs}\)</span> (see figure below). The
parametric density functionality can be deactivated by setting
<code>fit_parametric_curve</code> to <code>FALSE</code> in
<code><a href="../reference/set_analysis_parameters.html">set_analysis_parameters()</a></code>, in which case a p-value is
computed by directly comparing the observed z-score to the null
z-scores.</p>
<p><img src="fit_parametric_curve.png" width="650px" style="display: block; margin: auto;"></p>
<p>Most users need not to worry about the
<code>fit_parametric_curve</code> argument. <code>sceptre</code>
provides <code>fit_parametric_curve</code> as an option to facilitate
diagnosis of miscalibrated results on the negative control data.</p>
</div>
<div class="section level2">
<h2 id="multiple-testing-method-and-multiple-testing-alpha">Multiple testing method and multiple testing <span class="math inline">\(\alpha\)</span><a class="anchor" aria-label="anchor" href="#multiple-testing-method-and-multiple-testing-alpha"></a>
</h2>
<p>After <code>sceptre</code> is deployed to analyze the discovery pairs
and negative control pairs, the resulting p-values are subjected to a
multiplicity adjustment. The adjusted p-values are thresholded to call
each pair as “significant” or “not significant.” The parameter
<code>multiple_testing_method</code> controls the multiple testing
method that is used to adjust the p-values. The default option is
Benjamini-Hochberg (<code>"BH"</code>); other available methods can be
viewed via a call to <code>p.adjust.methods</code>.</p>
<p>The parameter <code>multiple_testing_alpha</code> (default value
<code>0.1</code>) controls the level of the multiple testing procedure.
If <code>multiple_testing_method</code> is set to a method that controls
that false discovery rate (e.g., <code>"BH"</code> or
<code>"BY"</code>), then <code>multiple_testing_alpha</code> is the
nominal false discovery rate. If, on the other hand,
<code>multiple_testing_method</code> is set to a method that controls
the family-wise error rate (e.g., <code>"bonferroni"</code> or
<code>"holm"</code>), then <code>multiple_testing_alpha</code> is the
nominal family-wise error rate.</p>
</div>
<div class="section level2">
<h2 id="setting-the-analysis-parameters">Setting the analysis parameters<a class="anchor" aria-label="anchor" href="#setting-the-analysis-parameters"></a>
</h2>
<p>We set the analysis parameters by calling
<code><a href="../reference/set_analysis_parameters.html">set_analysis_parameters()</a></code> on the
<code>sceptre_object</code>s, passing <code>discovery_pairs</code> and
<code>positive_control_pairs</code> (and, on the high-MOI CRISPRi data,
<code>side</code>). We allow all other arguments to take default
values.</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># high-MOI CRISPRi data</span></span>
<span><span class="va">sceptre_object_highmoi</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/set_analysis_parameters.html">set_analysis_parameters</a></span><span class="op">(</span>sceptre_object <span class="op">=</span> <span class="va">sceptre_object_highmoi</span>, </span>
<span>                                                  discovery_pairs <span class="op">=</span> <span class="va">discovery_pairs_highmoi</span>, </span>
<span>                                                  positive_control_pairs <span class="op">=</span> <span class="va">positive_control_pairs_highmoi</span>,</span>
<span>                                                  side <span class="op">=</span> <span class="st">"left"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># low-MOI CRISPRko data</span></span>
<span><span class="va">sceptre_object_lowmoi</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/set_analysis_parameters.html">set_analysis_parameters</a></span><span class="op">(</span>sceptre_object <span class="op">=</span> <span class="va">sceptre_object_lowmoi</span>,</span>
<span>                                                 discovery_pairs <span class="op">=</span> <span class="va">discovery_pairs_lowmoi</span>,</span>
<span>                                                 positive_control_pairs <span class="op">=</span> <span class="va">positive_control_pairs_lowmoi</span><span class="op">)</span></span></code></pre></div>
<p>We can call <code><a href="https://rdrr.io/r/base/print.html" class="external-link">print()</a></code> on the resulting
<code>sceptre_object</code>, which prints to the console a summary
tracking the status of the analysis. Notice that the “Analysis
parameters” field of the output contains information about the analysis
parameters that we have selected (or that have been set by default).</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="va">sceptre_object_highmoi</span><span class="op">)</span></span>
<span><span class="co">#&gt; An object of class <span style="color: #0000BB;">sceptre_object</span>.</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Attributes of the data:</span></span>
<span><span class="co">#&gt;  • <span style="color: #0000BB;">45919</span> cells</span></span>
<span><span class="co">#&gt;  • <span style="color: #0000BB;">526</span> responses</span></span>
<span><span class="co">#&gt;  • <span style="color: #0000BB;">High</span> multiplicity-of-infection </span></span>
<span><span class="co">#&gt;  • <span style="color: #0000BB;">70</span> targeting gRNAs (distributed across <span style="color: #0000BB;">35</span> targets) </span></span>
<span><span class="co">#&gt;  • <span style="color: #0000BB;">25</span> non-targeting gRNAs </span></span>
<span><span class="co">#&gt;  • <span style="color: #0000BB;">6</span> covariates (batch, grna_n_nonzero, grna_n_umis, response_n_nonzero, response_n_umis, response_p_mito)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Analysis status:</span></span>
<span><span class="co">#&gt;  <span style="color: #00BB00;">✓</span> import_data()</span></span>
<span><span class="co">#&gt;  <span style="color: #00BB00;">✓</span> set_analysis_parameters()</span></span>
<span><span class="co">#&gt;  <span style="color: #BB0000;">✗</span> assign_grnas()</span></span>
<span><span class="co">#&gt;  <span style="color: #BB0000;">✗</span> run_qc()</span></span>
<span><span class="co">#&gt;  <span style="color: #BB0000;">✗</span> run_calibration_check()</span></span>
<span><span class="co">#&gt;  <span style="color: #BB0000;">✗</span> run_power_check()</span></span>
<span><span class="co">#&gt;  <span style="color: #BB0000;">✗</span> run_discovery_analysis()</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Analysis parameters: </span></span>
<span><span class="co">#&gt;  • Discovery pairs: data frame with <span style="color: #0000BB;">610</span> pairs</span></span>
<span><span class="co">#&gt;  • Positive control pairs: data frame with <span style="color: #0000BB;">10</span> pairs</span></span>
<span><span class="co">#&gt;  • Sidedness of test: <span style="color: #0000BB;">left</span></span></span>
<span><span class="co">#&gt;  • N nonzero treatment cells threshold: not specified</span></span>
<span><span class="co">#&gt;  • N nonzero control cells threshold: not specified</span></span>
<span><span class="co">#&gt;  • Resampling mechanism: <span style="color: #0000BB;">conditional resampling</span></span></span>
<span><span class="co">#&gt;  • gRNA grouping strategy: <span style="color: #0000BB;">union</span></span></span>
<span><span class="co">#&gt;  • Formula object: <span style="color: #0000BB;">log(response_n_nonzero) + log(response_n_umis) + log(grna_n_nonzero) + log(grna_n_umis) + batch</span></span></span></code></pre></div>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>Developed by Timothy Barry, Eugene Katsevich.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
