<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Run SCEPTRE on high multiplicity-of-infection single-cell CRISPR screen data — run_sceptre_high_moi • sceptre</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/lumen/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Run SCEPTRE on high multiplicity-of-infection single-cell CRISPR screen data — run_sceptre_high_moi"><meta property="og:description" content="This function is the core function of the sceptre package. The function applies SCEPTRE to test for association between a set of gRNA groups and genes while controlling for technical confounders. The function returns a p-value for each pairwise test of association."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">sceptre</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.1.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../reference/index.html">Function reference</a>
</li>
<li>
  <a href="../articles/using_sceptre_v2.html">Tutorial</a>
</li>
<li>
  <a href="../news/index.html">News</a>
</li>
      </ul><ul class="nav navbar-nav navbar-right"></ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Run SCEPTRE on high multiplicity-of-infection single-cell CRISPR screen data</h1>
    
    <div class="hidden name"><code>run_sceptre_high_moi.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>This function is the core function of the <code>sceptre</code> package. The function applies SCEPTRE to test for association between a set of gRNA groups and genes while controlling for technical confounders. The function returns a p-value for each pairwise test of association.</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">run_sceptre_high_moi</span><span class="op">(</span></span>
<span>  <span class="va">gene_matrix</span>,</span>
<span>  <span class="va">combined_perturbation_matrix</span>,</span>
<span>  <span class="va">covariate_matrix</span>,</span>
<span>  <span class="va">gene_gRNA_group_pairs</span>,</span>
<span>  side <span class="op">=</span> <span class="st">"both"</span>,</span>
<span>  storage_dir <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/tempfile.html" class="external-link">tempdir</a></span><span class="op">(</span><span class="op">)</span>,</span>
<span>  regularization_amount <span class="op">=</span> <span class="fl">0</span>,</span>
<span>  B <span class="op">=</span> <span class="fl">1000</span>,</span>
<span>  full_output <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  parallel <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  seed <span class="op">=</span> <span class="fl">4</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>
    <dl><dt>gene_matrix</dt>
<dd><p>a gene-by-cell expression matrix; the rows (i.e., gene IDs) and columns (i.e., cell barcodes) should be named</p></dd>


<dt>combined_perturbation_matrix</dt>
<dd><p>a binary matrix of perturbations (i.e., gRNA group-to-cell assignments); the rows (i.e., gRNA groups) and columns (i.e., cell barcodes) should be named.</p></dd>


<dt>covariate_matrix</dt>
<dd><p>the cell-specific matrix of technical factors, ideally containing the following covariates: log-transformed gene library size (numeric), log-transformed gRNA library size (numeric), percent mitochondrial reads (numeric), and batch (factor). The rows (i.e., cell barcodes) should be named</p></dd>


<dt>gene_gRNA_group_pairs</dt>
<dd><p>a data frame specifying the gene-gRNA group pairs to test for association; the data frame should contain columns named <code>gene_id</code> and <code>gRNA_group</code>.</p></dd>


<dt>side</dt>
<dd><p>sidedness of the test; one of "both," "left," and "right". (default "both")</p></dd>


<dt>storage_dir</dt>
<dd><p>directory in which to store the intermediate computations (default tempdir)</p></dd>


<dt>regularization_amount</dt>
<dd><p>non-negative number specifying the amount of regularization to apply to the negative binomial dispersion parameter estimates (default 0)</p></dd>


<dt>B</dt>
<dd><p>number of resamples to draw for the conditional randomization test. (default 1000)</p></dd>


<dt>full_output</dt>
<dd><p>return the full output (TRUE) or a simplified, reduced output (FALSE)? (default FALSE)</p></dd>


<dt>parallel</dt>
<dd><p>parallelize execution? (default TRUE)</p></dd>


<dt>seed</dt>
<dd><p>seed to the random number generator (default 4)</p></dd>

</dl></div>
    <div id="value">
    <h2>Value</h2>
    

<p>the <code>gene_gRNA_group_pairs</code> data frame with new columns <code>p_value</code>, <code>z_value</code>, and <code>log_fold_change</code> appended. See "details" for a description of the output when <code>full_output</code> is set to TRUE.</p>
    </div>
    <div id="details">
    <h2>Details</h2>
    <p>Details are arranged from most to least important.</p><ul><li><p><code>gene_matrix</code> should be a <strong>raw</strong> (i.e., un-normalized) matrix of UMI (unique molecular identifier) counts.</p></li>
<li><p><code>combined_perturbation_matrix</code> should be a "combined perturbation matrix", which can be obtained by applying the functions <code>threshold_gRNA_matrix</code> and <code>combine_perturbations</code> (in that order) to a raw gRNA count matrix. <code>combined_perturbation_matrix</code> optionally can be a raw gRNA expression matrix or an uncombined perturbation matrix, in which case each gRNA is treated as its own group of one. See the tutorial for more details.</p></li>
<li><p>The gene IDs (respectively, gRNA groups) within <code>gene_gRNA_group_pairs</code> must be a subset of the row names of <code>gene_matrix</code> (respectively, <code>combined_perturbation_matrix</code>).</p></li>
<li><p>The <code>side</code> parameter controls the sidedness of the test. The arguments "left" and "right" are appropriate when testing for a decrease and increase in gene expression, respectively. The default argument -- "both" -- is appropriate when testing for an increase <em>or</em> decrease in gene expression.</p></li>
<li><p>The default value of <code>regularization_amount</code> is 0.0, meaning that zero regularization is applied to the estimated negative binomial size parameters. One can increase the value of this parameter to protect against overfitting, which can be useful when there are many genes.</p></li>
<li><p>When <code>full_output</code> is set to TRUE (as opposed to FALSE, the default), the output is a data frame with the following columns: <code>gene_id</code>, <code>gRNA_id</code>, <code>p_value</code>, <code>skew_t_fit_success</code> (if TRUE, <em>p</em>-value based on tail probability of fitted skew-t distribution returned; if FALSE, empirical <em>p</em>-value returned), <code>xi</code>, <code>omega</code>, <code>alpha</code>, <code>nu</code> (fitted parameters of the skew-t distribution; NA if fit failed), <code>z_value</code> (z-value obtained on "ground truth" data), and <code>z_null_1</code>, ..., <code>z_null_B</code> (z-values obtained from resampled datasets).</p></li>
</ul></div>

    <div id="ref-examples">
    <h2>Examples</h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="kw">if</span> <span class="op">(</span><span class="cn">FALSE</span><span class="op">)</span> <span class="op">{</span></span></span>
<span class="r-in"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://dplyr.tidyverse.org" class="external-link">dplyr</a></span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://magrittr.tidyverse.org" class="external-link">magrittr</a></span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="co"># 1. load the data</span></span></span>
<span class="r-in"><span><span class="fu"><a href="data.html">data</a></span><span class="op">(</span><span class="va">gene_matrix</span><span class="op">)</span> <span class="co"># i. gene expression matrix</span></span></span>
<span class="r-in"><span><span class="fu"><a href="data.html">data</a></span><span class="op">(</span><span class="va">gRNA_matrix</span><span class="op">)</span> <span class="co"># ii. gRNA expression matrix</span></span></span>
<span class="r-in"><span><span class="fu"><a href="data.html">data</a></span><span class="op">(</span><span class="va">covariate_matrix</span><span class="op">)</span> <span class="co"># iii. covariate matrix</span></span></span>
<span class="r-in"><span><span class="fu"><a href="data.html">data</a></span><span class="op">(</span><span class="va">gRNA_groups_table</span><span class="op">)</span> <span class="co"># iv. gRNAs grouped by target site</span></span></span>
<span class="r-in"><span><span class="fu"><a href="data.html">data</a></span><span class="op">(</span><span class="va">gene_gRNA_group_pairs</span><span class="op">)</span> <span class="co"># v. gene-gRNA group pairs to analyze</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># 2. threshold and combine gRNA matrix</span></span></span>
<span class="r-in"><span><span class="va">combined_perturbation_matrix</span> <span class="op">&lt;-</span> <span class="fu"><a href="threshold_gRNA_matrix.html">threshold_gRNA_matrix</a></span><span class="op">(</span><span class="va">gRNA_matrix</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span></span>
<span class="r-in"><span><span class="fu"><a href="combine_perturbations.html">combine_perturbations</a></span><span class="op">(</span><span class="va">gRNA_groups_table</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># 3. select the gene-gRNA group pairs to analyze</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">4</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">gene_gRNA_group_pairs</span> <span class="op">&lt;-</span> <span class="va">gene_gRNA_group_pairs</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/sample_n.html" class="external-link">sample_n</a></span><span class="op">(</span><span class="fl">25</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="co"># 3. run method (takes ~40s on an 8-core Macbook Pro)</span></span></span>
<span class="r-in"><span><span class="va">result</span> <span class="op">&lt;-</span> <span class="fu">run_sceptre_high_moi</span><span class="op">(</span>gene_matrix <span class="op">=</span> <span class="va">gene_matrix</span>,</span></span>
<span class="r-in"><span>combined_perturbation_matrix <span class="op">=</span> <span class="va">combined_perturbation_matrix</span>,</span></span>
<span class="r-in"><span>covariate_matrix <span class="op">=</span> <span class="va">covariate_matrix</span>,</span></span>
<span class="r-in"><span>gene_gRNA_group_pairs <span class="op">=</span> <span class="va">gene_gRNA_group_pairs</span>,</span></span>
<span class="r-in"><span>side <span class="op">=</span> <span class="st">"left"</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="op">}</span></span></span>
</code></pre></div>
    </div>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by <a href="https://timothy-barry.github.io/" class="external-link">Timothy Barry</a>, Eugene Katsevich.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.5.</p>
</div>

      </footer></div>

  


  

  </body></html>

