<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><meta name="description" content="The function applies SCEPTRE to test for association between a set of gRNA groups and genes in a high MOI experiment. The function returns a p-value and log fold change estimate for each pairwise test of association."><title>Run sceptre on high MOI data — run_sceptre_highmoi • sceptre</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.2.2/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.2.2/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- bootstrap-toc --><script src="https://cdn.rawgit.com/afeld/bootstrap-toc/v1.0.1/dist/bootstrap-toc.min.js"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Run sceptre on high MOI data — run_sceptre_highmoi"><meta property="og:description" content="The function applies SCEPTRE to test for association between a set of gRNA groups and genes in a high MOI experiment. The function returns a p-value and log fold change estimate for each pairwise test of association."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-default navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">sceptre</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="Released version">0.2.0</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="active nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="../articles/highmoi_tutorial.html">High MOI sceptre tutorial</a>
    <a class="dropdown-item" href="../articles/lowmoi_tutorial.html">Low MOI sceptre tutorial</a>
    <a class="dropdown-item" href="../articles/quality_control_in_low_moi.html">Low MOI Quality Control</a>
  </div>
</li>
<li class="nav-item">
  <a class="nav-link" href="../news/index.html">Changelog</a>
</li>
      </ul><form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off"></form>

      <ul class="navbar-nav"></ul></div>

    
  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>Run <code>sceptre</code> on high MOI data</h1>
      
      <div class="d-none name"><code>run_sceptre_highmoi.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>The function applies SCEPTRE to test for association between a set of gRNA groups and genes in a high MOI experiment. The function returns a p-value and log fold change estimate for each pairwise test of association.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">run_sceptre_highmoi</span><span class="op">(</span></span>
<span>  <span class="va">gene_matrix</span>,</span>
<span>  <span class="va">combined_perturbation_matrix</span>,</span>
<span>  <span class="va">covariate_matrix</span>,</span>
<span>  <span class="va">gene_gRNA_group_pairs</span>,</span>
<span>  side <span class="op">=</span> <span class="st">"both"</span>,</span>
<span>  storage_dir <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/tempfile.html" class="external-link">tempdir</a></span><span class="op">(</span><span class="op">)</span>,</span>
<span>  regularization_amount <span class="op">=</span> <span class="fl">0</span>,</span>
<span>  B <span class="op">=</span> <span class="fl">1000</span>,</span>
<span>  full_output <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  parallel <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  seed <span class="op">=</span> <span class="fl">4</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>
    <dl><dt>gene_matrix</dt>
<dd><p>a gene-by-cell expression matrix; the rows (i.e., gene IDs) and columns (i.e., cell barcodes) should be named</p></dd>


<dt>combined_perturbation_matrix</dt>
<dd><p>a binary matrix of perturbations (i.e., gRNA group-to-cell assignments); the rows (i.e., gRNA groups) and columns (i.e., cell barcodes) should be named.</p></dd>


<dt>covariate_matrix</dt>
<dd><p>the cell-specific matrix of technical factors, ideally containing the following covariates: log-transformed gene library size (numeric), log-transformed gRNA library size (numeric), percent mitochondrial reads (numeric), and batch (factor). The rows (i.e., cell barcodes) should be named</p></dd>


<dt>gene_gRNA_group_pairs</dt>
<dd><p>a data frame specifying the gene-gRNA group pairs to test for association; the data frame should contain columns named <code>gene_id</code> and <code>gRNA_group</code>.</p></dd>


<dt>side</dt>
<dd><p>sidedness of the test; one of "both," "left," and "right". (default "both")</p></dd>


<dt>storage_dir</dt>
<dd><p>directory in which to store the intermediate computations (default tempdir)</p></dd>


<dt>regularization_amount</dt>
<dd><p>non-negative number specifying the amount of regularization to apply to the negative binomial dispersion parameter estimates (default 0)</p></dd>


<dt>B</dt>
<dd><p>number of resamples to draw for the conditional randomization test. (default 1000)</p></dd>


<dt>full_output</dt>
<dd><p>return the full output (TRUE) or a simplified, reduced output (FALSE)? (default FALSE)</p></dd>


<dt>parallel</dt>
<dd><p>parallelize execution? (default TRUE)</p></dd>


<dt>seed</dt>
<dd><p>seed to the random number generator (default 4)</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    

<p>the <code>gene_gRNA_group_pairs</code> data frame with new columns <code>p_value</code>, <code>z_value</code>, and <code>log_fold_change</code> appended. See "details" for a description of the output when <code>full_output</code> is set to TRUE.</p>
    </div>
    <div class="section level2">
    <h2 id="details">Details<a class="anchor" aria-label="anchor" href="#details"></a></h2>
    <p>Details are arranged from most to least important.</p><ul><li><p><code>gene_matrix</code> should be a <strong>raw</strong> (i.e., un-normalized) matrix of UMI (unique molecular identifier) counts.</p></li>
<li><p><code>combined_perturbation_matrix</code> should be a "combined perturbation matrix", which can be obtained by applying the functions <code>threshold_gRNA_matrix</code> and <code>combine_perturbations</code> (in that order) to a raw gRNA count matrix. <code>combined_perturbation_matrix</code> optionally can be a raw gRNA expression matrix or an uncombined perturbation matrix, in which case each gRNA is treated as its own group of one. See the tutorial for more details.</p></li>
<li><p>The gene IDs (respectively, gRNA groups) within <code>gene_gRNA_group_pairs</code> must be a subset of the row names of <code>gene_matrix</code> (respectively, <code>combined_perturbation_matrix</code>).</p></li>
<li><p>The <code>side</code> parameter controls the sidedness of the test. The arguments "left" and "right" are appropriate when testing for a decrease and increase in gene expression, respectively. The default argument -- "both" -- is appropriate when testing for an increase <em>or</em> decrease in gene expression.</p></li>
<li><p>The default value of <code>regularization_amount</code> is 0.0, meaning that zero regularization is applied to the estimated negative binomial size parameters. One can increase the value of this parameter to protect against overfitting, which can be useful when there are many genes.</p></li>
<li><p>When <code>full_output</code> is set to TRUE (as opposed to FALSE, the default), the output is a data frame with the following columns: <code>gene_id</code>, <code>gRNA_id</code>, <code>p_value</code>, <code>skew_t_fit_success</code> (if TRUE, <em>p</em>-value based on tail probability of fitted skew-t distribution returned; if FALSE, empirical <em>p</em>-value returned), <code>xi</code>, <code>omega</code>, <code>alpha</code>, <code>nu</code> (fitted parameters of the skew-t distribution; NA if fit failed), <code>z_value</code> (z-value obtained on "ground truth" data), and <code>z_null_1</code>, ..., <code>z_null_B</code> (z-values obtained from resampled datasets).</p></li>
</ul></div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="kw">if</span> <span class="op">(</span><span class="cn">FALSE</span><span class="op">)</span> <span class="op">{</span></span></span>
<span class="r-in"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://dplyr.tidyverse.org" class="external-link">dplyr</a></span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="co"># 1. load the data</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">gene_matrix</span><span class="op">)</span> <span class="co"># i. gene expression matrix</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">gRNA_matrix</span><span class="op">)</span> <span class="co"># ii. gRNA expression matrix</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">covariate_matrix</span><span class="op">)</span> <span class="co"># iii. covariate matrix</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">gRNA_groups_table</span><span class="op">)</span> <span class="co"># iv. gRNAs grouped by target site</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">gene_gRNA_group_pairs</span><span class="op">)</span> <span class="co"># v. gene-gRNA group pairs to analyze</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># 2. threshold and combine gRNA matrix</span></span></span>
<span class="r-in"><span><span class="va">combined_perturbation_matrix</span> <span class="op">&lt;-</span> <span class="fu"><a href="threshold_gRNA_matrix.html">threshold_gRNA_matrix</a></span><span class="op">(</span><span class="va">gRNA_matrix</span><span class="op">)</span> <span class="op">|&gt;</span></span></span>
<span class="r-in"><span><span class="fu"><a href="combine_perturbations.html">combine_perturbations</a></span><span class="op">(</span><span class="va">gRNA_groups_table</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># 3. select the gene-gRNA group pairs to analyze</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">4</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">gene_gRNA_group_pairs</span> <span class="op">&lt;-</span> <span class="va">gene_gRNA_group_pairs</span> <span class="op">|&gt;</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/sample_n.html" class="external-link">sample_n</a></span><span class="op">(</span><span class="fl">25</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="co"># 3. run method (takes ~40s on an 8-core Macbook Pro)</span></span></span>
<span class="r-in"><span><span class="va">result</span> <span class="op">&lt;-</span> <span class="fu">run_sceptre_highmoi</span><span class="op">(</span>gene_matrix <span class="op">=</span> <span class="va">gene_matrix</span>,</span></span>
<span class="r-in"><span>combined_perturbation_matrix <span class="op">=</span> <span class="va">combined_perturbation_matrix</span>,</span></span>
<span class="r-in"><span>covariate_matrix <span class="op">=</span> <span class="va">covariate_matrix</span>,</span></span>
<span class="r-in"><span>gene_gRNA_group_pairs <span class="op">=</span> <span class="va">gene_gRNA_group_pairs</span>,</span></span>
<span class="r-in"><span>side <span class="op">=</span> <span class="st">"left"</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="op">}</span></span></span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p></p><p>Developed by <a href="https://timothy-barry.github.io/" class="external-link">Timothy Barry</a>, <a href="https://ekatsevi.github.io/" class="external-link">Eugene Katsevich</a>.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.5.</p>
</div>

    </footer></div>

  

  

  </body></html>

